{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Solve the Change Problem. The DPChange pseudocode is reproduced below for your convenience.\n",
    "'''\n",
    "    DPCHANGE(money, Coins)\n",
    "     MinNumCoins(0) = 0\n",
    "     for m = 1 to money\n",
    "        MinNumCoins(m) = inf\n",
    "            for i = 1 to |Coins|\n",
    "                if m less then or equal coini\n",
    "                    if MinNumCoins(m - coini) + 1 less then MinNumCoins(m)\n",
    "                        MinNumCoins(m) = MinNumCoins(m - coini) + 1\n",
    "    output MinNumCoins(money)\n",
    "'''\t\n",
    "def dpchange(money,coins):\n",
    "    minnumcoins = [0]\n",
    "    for m in range(1,money+1):\n",
    "        minnumcoins.append(float('Inf'))\n",
    "        for coin in coins:\n",
    "            if m >= coin:\n",
    "                if minnumcoins[m - coin] + 1 < minnumcoins[m]:\n",
    "                    minnumcoins[m] = minnumcoins[m - coin] + 1\n",
    "    return minnumcoins[money]\n",
    "\n",
    "def dynamicCoinChange(coins, change):\n",
    "    #initialize to zero\n",
    "    Optimal = [0 for i in range(0, change+1)]\n",
    "    for i in range(1, change+1):\n",
    "        smallest = float(\"inf\")\n",
    "        for j in range(0, len(coins)):\n",
    "            if (coins[j] <= i):\n",
    "                smallest = min(smallest, Optimal[i - coins[j]]) \n",
    "            Optimal[i] = 1 + smallest \n",
    "    return Optimal[change]\n",
    "\n",
    "dynamicCoinChange(coins, change)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "CODE CHALLENGE: Find the length of a longest path in the Manhattan Tourist Problem.\n",
    "Input: Integers n and m, followed by an n multiply (m plus 1) matrix Down and an (n plus 1)  multiply m matrix Right\n",
    "The two matrices are separated by the symbol\n",
    "Output: The length of a longest path from source (0, 0) to sink (n, m) in the n multiply m rectangular grid\n",
    "whose edges are defined by the matrices Down and Right.\n",
    "Sample Input:\n",
    "     4 4\n",
    "     1 0 2 4 3\n",
    "     4 6 5 2 1\n",
    "     4 4 5 2 1\n",
    "     5 6 8 5 3\n",
    "     -\n",
    "     3 2 4 0\n",
    "     3 2 4 2\n",
    "     0 7 3 3\n",
    "     3 3 0 2\n",
    "     1 3 2 2\n",
    "Sample Output:\n",
    "     34\n",
    "\"\"\"\n",
    "\"\"\" manhatten.txt --\n",
    "1 0 2 4 3\n",
    "4 6 5 2 1\n",
    "4 4 5 2 1\n",
    "5 6 8 5 3\n",
    "-\n",
    "3 2 4 0\n",
    "3 2 4 2\n",
    "0 7 3 3\n",
    "3 3 0 2\n",
    "1 3 2 2\n",
    "\"\"\"\n",
    "\n",
    "with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/manhatten.txt') as inputData:\n",
    "    downMatrix, rightMatrix = [[list(map(int, row.split())) for row in matricies.split('\\n')] for matricies in inputData.read().strip().split('\\n-\\n')]\n",
    "    \n",
    "n = 4\n",
    "m = 4   \n",
    "\n",
    "def manhattan_tourist(n, m, downMatrix, rightMatrix):\n",
    "    #initialize matrix to 0\n",
    "    S = [[0]*(m+1) for i in range(n+1)]\n",
    "    print(S)\n",
    "\n",
    "    for i in range(1, n + 1):\n",
    "        S[i][0] = S[i-1][0] + downMatrix[i-1][0]\n",
    "    for j in range(1, m + 1):\n",
    "        S[0][j] = S[0][j-1] + rightMatrix[0][j-1]\n",
    "\n",
    "    for i in range(1, n+1):\n",
    "        for j in range(1, m+1):\n",
    "            S[i][j] = max(S[i-1][j] + downMatrix[i-1][j], S[i][j-1] + rightMatrix[i][j-1])\n",
    "\n",
    "    return S[n][m]\n",
    "\n",
    "import numpy as np\n",
    "def dpmanhattantourist(n,m,down,right):\n",
    "    s = np.zeros((n+1,m+1), dtype=int)\n",
    "    for i in range(1,n+1):\n",
    "        s[i][0] = s[i-1][0] + down[i-1][0]\n",
    "    for j in range(1,m+1):\n",
    "        s[0][j] = s[0][j-1] + right[0][j-1]\n",
    "    for i in range(1,n+1):\n",
    "        for j in range(1,m+1):\n",
    "            s[i][j] = max(s[i-1][j]+down[i-1][j],s[i][j-1]+right[i][j-1])\n",
    "\n",
    "    return s[n][m]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "CODE CHALLENGE: Use OUTPUTLCS (reproduced below) to solve the Longest Common Subsequence Problem.\n",
    "     Input: Two strings s and t.\n",
    "     Output: A longest common subsequence of s and t. (Note: more than one solution may exist,\n",
    "     in which case you may output any one.)\n",
    "    OUTPUTLCS(backtrack, v, i, j)\n",
    "        if i equals 0 or j equals 0\n",
    "            return\n",
    "        if backtracki, j equals\"backward\"\n",
    "            OUTPUTLCS(backtrack, v, i minus 1, j)\n",
    "        else if backtracki, j equals \"forward\"\n",
    "            OUTPUTLCS(backtrack, v, i, j minus  1)\n",
    "        else\n",
    "            OUTPUTLCS(backtrack, v, i minus 1, j minus 1)\n",
    "            output vi\n",
    "Sample Input:\n",
    "     AACCTTGG\n",
    "     ACACTGTGA\n",
    "Sample Output:\n",
    "     AACTGG\n",
    "'''\n",
    "\n",
    "def lcsbacktrack(v,w):\n",
    "    n = len(v)\n",
    "    m = len(w)\n",
    "    s = np.zeros((n+1,m+1),dtype=int)\n",
    "    backtrack = []\n",
    "    #notation for backtrack\n",
    "    #'v' for vertical\n",
    "    #'d' for diagonal\n",
    "    #'h' for horizontal\n",
    "    for i in range(n):\n",
    "        temp = []\n",
    "        for j in range(m):\n",
    "            if v[i] == w[j]:\n",
    "                s[i+1][j+1] = s[i][j] + 1\n",
    "            else:\n",
    "                s[i+1][j+1] = max(s[i][j+1],s[i+1][j])\n",
    "            if s[i+1][j+1] == s[i][j] + 1 and v[i] == w[j]:\n",
    "                temp.append('d')\n",
    "            elif s[i+1][j+1] == s[i+1][j]:\n",
    "                temp.append('v')\n",
    "            elif s[i+1][j+1] == s[i][j+1]:\n",
    "                temp.append('h')\n",
    "        backtrack.append(temp)\n",
    "\n",
    "    longest_sseq = ''\n",
    "    i,j = len(v), len(w)\n",
    "    while i*j != 0:\n",
    "        if s[i][j] == s[i-1][j]:\n",
    "            i -= 1\n",
    "        elif s[i][j] == s[i][j-1]:\n",
    "            j -= 1\n",
    "        else:\n",
    "            longest_sseq = v[i-1] + longest_sseq\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "    return longest_sseq\n",
    "\n",
    "def longestCommonSubsequence(seq1, seq2):\n",
    "    '''Return a longest common subsequence of the input strings'''\n",
    "    lenSeq1 = len(seq1)\n",
    "    lenSeq2 = len(seq2)\n",
    "\n",
    "    S = [[0]*(lenSeq2+1) for k in range(lenSeq1 + 1)]\n",
    "\n",
    "    for i in range(lenSeq1):\n",
    "        for j in range(lenSeq2):\n",
    "\n",
    "            if seq1[i] == seq2[j]:\n",
    "                S[i+1][j+1] = S[i][j]+1\n",
    "            else:\n",
    "                S[i+1][j+1] = max(S[i+1][j], S[i][j+1])\n",
    "\n",
    "\n",
    "    longestSubSeq = ''\n",
    "\n",
    "    a = len(seq1)\n",
    "    b = len(seq2)\n",
    "    while a*b != 0:\n",
    "        if S[a][b] == S[a-1][b]:\n",
    "            a -= 1\n",
    "        elif S[a][b] == S[a][b - 1]:\n",
    "            b -= 1\n",
    "        else:\n",
    "            longestSubSeq = seq1[a-1] + longestSubSeq\n",
    "            a -= 1\n",
    "            b -= 1\n",
    "\n",
    "    return longestSubSeq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "CODE CHALLENGE: Solve the Longest Path in a DAG Problem.\n",
    "     Input: An integer representing the source node of a graph, followed by an integer representing the\n",
    "     sink node of the graph, followed by a list of edges in the graph. The edge notation 0->1:7 indicates\n",
    "     that an edge connects node 0 to node 1 with weight 7. \n",
    "     Output: The length of a longest path in the graph, followed by a longest path. (If multiple longest paths exist, you may return any one.)\n",
    "Sample Input:\n",
    "     0\n",
    "     4\n",
    "     0->1:7\n",
    "     0->2:4\n",
    "     2->3:2\n",
    "     1->4:1\n",
    "     3->4:3\n",
    "Sample Output:\n",
    "     9\n",
    "     0->2->3->4\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "Longest Path in a DAG Problem\n",
    "Find a longest path between two nodes in an edge-weighted DAG.\n",
    "Given: An integer representing the source node of a graph, followed by an integer \n",
    "representing the sink node of the graph, followed by an edge-weighted graph. \n",
    "The graph is represented by a modified adjacency list in which the notation \"0->1:7\" \n",
    "indicates that an edge connects node 0 to node 1 with weight 7.\n",
    "Return: The length of a longest path in the graph, followed by a longest path. \n",
    "(If multiple longest paths exist, you may return any one.)\n",
    "\"\"\"\n",
    "with open(filename) as f:\n",
    "    lines = f.read().splitlines()\n",
    "s = int(lines[0])\n",
    "v = int(lines[1])\n",
    "G = defaultdict(list)\n",
    "for line in lines[2:]:\n",
    "    nodei = int(line.split(\"->\")[0])\n",
    "    nodej = int(line.split(\"->\")[1].split(':')[0])\n",
    "    w = int(line.split(\"->\")[1].split(':')[1])\n",
    "    G[nodei].append((nodej,w))\n",
    "    \n",
    "from collections import defaultdict\n",
    "import sys\n",
    "\n",
    "def update_graph(graph):\n",
    "    nodes = list(graph.keys())\n",
    "    for node in graph.keys():\n",
    "        links = graph[node]\n",
    "        for link in links:\n",
    "            if link[0] not in nodes:\n",
    "                nodes.append(link[0])\n",
    "    ''' make graph complete with only in-going nodes'''\n",
    "    for node in nodes:\n",
    "        if node not in graph.keys():\n",
    "            graph[node] = []\n",
    "    return graph,nodes\n",
    "\n",
    "def acyclicLP(graph,s,v):\n",
    "    global postorder,marked,nodes,distTo,edgeTo\n",
    "    graph,nodes = update_graph(graph)\n",
    "    edgeTo = {node:None for node in nodes}\n",
    "    distTo = {node:float(\"-inf\") for node in nodes}\n",
    "    distTo[s] = 0\n",
    "    depthFirstOrder(graph)\n",
    "    topo_order = postorder[::-1]\n",
    "    for node in topo_order:\n",
    "        for edge in graph[node]:\n",
    "            relax(node,edge[0],edge[1])\n",
    "    pathsTOv = pathTo(graph,s,v)\n",
    "    return distTo[v],pathsTOv\n",
    "\n",
    "def relax(start,end,weight):\n",
    "    global distTo,edgeTo\n",
    "    if distTo[end] < distTo[start] + weight:\n",
    "        distTo[end] = distTo[start] + weight\n",
    "        edgeTo[end] = start\n",
    "\n",
    "def depthFirstOrder(graph):\n",
    "    global postorder, marked, nodes\n",
    "    ''' set up variables '''\n",
    "    postorder = []#queue FIFO\n",
    "    marked = {node:False for node in nodes}\n",
    "    for v in graph.keys():\n",
    "        if not marked[v]:\n",
    "            dfs(G,v)\n",
    "\n",
    "def dfs(graph,v):\n",
    "    global postorder, marked\n",
    "    marked[v] = True\n",
    "    for nodew in graph[v]:\n",
    "        w = nodew[0]\n",
    "        if not marked[w]:\n",
    "            dfs(graph,w)\n",
    "    postorder.append(v)\n",
    "\n",
    "def pathTo(graph,s,v):\n",
    "    global edgeTo\n",
    "    stack = []\n",
    "    w = edgeTo[v]\n",
    "    stack.append(v)\n",
    "    while w != s:\n",
    "        stack.append(w)\n",
    "        v = w\n",
    "        w = edgeTo[v]\n",
    "    stack.append(s)\n",
    "    return stack[::-1]\n",
    "\n",
    "longest, path = acyclicLP(G,s,v)\n",
    "print(longest)\n",
    "print('->'.join(map(str,path)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Here is the adjacency list of a graph with six nodes and ten edges:\n",
    "\n",
    "a -> b, c, d, e, f\n",
    "\n",
    "b -> c, f\n",
    "\n",
    "c -> d\n",
    "\n",
    "d ->\n",
    "\n",
    "e -> d, f\n",
    "\n",
    "f ->\n",
    "\n",
    "Which of the following are topological orderings of the nodes in this graph? (Select all that apply.)\n",
    "a = 0, b= 1, c =2, d= 3, e =4, f= 5\n",
    "\n",
    "'''\n",
    "# class to represent a graph object:\n",
    "class Graph:\n",
    "\n",
    "    # Constructor\n",
    "    def __init__(self, edges, N):\n",
    "\n",
    "        # A List of Lists to represent an adjacency list\n",
    "        self.adjList = [[] for _ in range(N)]\n",
    "\n",
    "        # stores in-degree of a vertex\n",
    "        # initialize in-degree of each vertex by 0\n",
    "        self.indegree = [0] * N\n",
    "\n",
    "        # add edges to the undirected graph\n",
    "        for (src, dest) in edges:\n",
    "\n",
    "            # add an edge from source to destination\n",
    "            self.adjList[src].append(dest)\n",
    "\n",
    "            # increment in-degree of destination vertex by 1\n",
    "            self.indegree[dest] = self.indegree[dest] + 1\n",
    "\n",
    "\n",
    "# Recursive function to find all topological orderings of a given DAG\n",
    "def findAllTopologicalOrders(graph, path, discovered, N):\n",
    "\n",
    "    # do for every vertex\n",
    "    for v in range(N):\n",
    "\n",
    "        # proceed only if in-degree of current node is 0 and\n",
    "        # current node is not processed yet\n",
    "        if graph.indegree[v] == 0 and not discovered[v]:\n",
    "\n",
    "            # for every adjacent vertex u of v, reduce in-degree of u by 1\n",
    "            for u in graph.adjList[v]:\n",
    "                graph.indegree[u] = graph.indegree[u] - 1\n",
    "\n",
    "            # include current node in the path and mark it as discovered\n",
    "            path.append(v)\n",
    "            discovered[v] = True\n",
    "\n",
    "            # recur\n",
    "            findAllTopologicalOrders(graph, path, discovered, N)\n",
    "\n",
    "            # backtrack: reset in-degree information for the current node\n",
    "            for u in graph.adjList[v]:\n",
    "                graph.indegree[u] = graph.indegree[u] + 1\n",
    "\n",
    "            # backtrack: remove current node from the path and\n",
    "            # mark it as undiscovered\n",
    "            path.pop()\n",
    "            discovered[v] = False\n",
    "\n",
    "    # print the topological order if all vertices are included in the path\n",
    "    if len(path) == N:\n",
    "        print(path)\n",
    "\n",
    "\n",
    "# Print all topological orderings of a given DAG\n",
    "def printAllTopologicalOrders(graph):\n",
    "\n",
    "    # get number of nodes in the graph\n",
    "    N = len(graph.adjList)\n",
    "\n",
    "    # create an auxiliary space to keep track of whether vertex is discovered\n",
    "    discovered = [False] * N\n",
    "\n",
    "    # list to store the topological order\n",
    "    path = []\n",
    "\n",
    "    # find all topological ordering and print them\n",
    "    findAllTopologicalOrders(graph, path, discovered, N)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# List of graph edges as per above diagram\n",
    "edges = [(0, 1), (0, 2), (0, 3), (0, 4), (0, 5), (1, 2), (1, 5), (2, 3), (4, 3), (4, 5)]\n",
    "\n",
    "# Number of nodes in the graph\n",
    "N = 6\n",
    "\n",
    "# create a graph from edges\n",
    "graph = Graph(edges, N)\n",
    "\n",
    "# print all topological ordering of the graph\n",
    "printAllTopologicalOrders(graph)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Global Alignment Problem\n",
    "Find the highest-scoring alignment between two strings using a scoring matrix.\n",
    "Given: Two protein strings.\n",
    "Return: The maximum alignment score of these strings followed by an \n",
    "alignment achieving this maximum score. \n",
    "https://github.com/ahmdeen/Bioinformatic-Algorthims/blob/7cae61bb45535911f220da290f9ecf609b964058/Ros5E_GlobalAlignment.py\n",
    "https://github.com/ngaude/sandbox/search?utf8=%E2%9C%93&q=+++++PLEASANTLY++++++MEANLY\n",
    "CODE CHALLENGE: Solve the Global Alignment Problem.\n",
    "     Input: Two protein strings written in the single-letter amino acid alphabet.\n",
    "     Output: The maximum alignment score of these strings followed by an alignment achieving this\n",
    "     maximum score. Use the BLOSUM62 scoring matrix and indel penalty σ = 5.\n",
    "Sample Input:\n",
    "     PLEASANTLY\n",
    "     MEANLY\n",
    "Sample Output:\n",
    "     8\n",
    "     PLEASANTLY\n",
    "     -MEA--N-LY\n",
    "\"\"\"\n",
    "\n",
    "def getBLOSUM62():\n",
    "    with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/BLOSUM62.txt', 'r') as f:\n",
    "        lines = f.read().splitlines()\n",
    "    alpha = lines[0].lstrip().split()\n",
    "    blosum62 = []\n",
    "    for i in range(1,len(lines)):\n",
    "        blosum62.append(list(map(int,lines[i][1:].lstrip().split())))\n",
    "    print(blosum62)\n",
    "    return blosum62,alpha\n",
    "\n",
    "def global_alignment(x,y,indel):\n",
    "    \"down:1, right2: diagnoal:3\"\n",
    "    blosum62,alphabet = getBLOSUM62()\n",
    "    # create the distance matrix\n",
    "    D = []\n",
    "    pointer = []\n",
    "    for i in range(len(x)+1):\n",
    "        D.append([0]*(len(y)+1))\n",
    "        pointer.append([None]*(len(y)+1))\n",
    "    # initialize the first column\n",
    "    for i in range(1,len(x)+1):\n",
    "        D[i][0] = D[i-1][0] - indel\n",
    "        pointer[i][0] = 1\n",
    "    # initialize the first row\n",
    "    for j in range(1,len(y)+1):\n",
    "        D[0][j] = D[0][j-1] - indel\n",
    "        pointer[0][j] = 2\n",
    "\n",
    "    # fill in the distance matrix and keep the pointer\n",
    "    for i in range(1,len(x)+1):\n",
    "        for j in range(1,len(y)+1):\n",
    "            distHor = D[i][j-1] - indel\n",
    "            distVer = D[i-1][j] - indel\n",
    "            distDiag = D[i-1][j-1] + blosum62[alphabet.index(x[i-1])][alphabet.index(y[j-1])]\n",
    "            candi = [distVer,distHor,distDiag]\n",
    "            D[i][j] = max(candi)\n",
    "            direction = candi.index(D[i][j])\n",
    "            '''preference: diag, down, right'''\n",
    "            if direction == 2:\n",
    "                pointer[i][j] = 3#diag\n",
    "            elif direction == 0:\n",
    "                pointer[i][j] = 1#down\n",
    "            else:\n",
    "                pointer[i][j] = 2#right\n",
    "    return D[-1][-1],pointer\n",
    "\n",
    "def back_track(pointer,x,y):\n",
    "    ret = [[],[]]\n",
    "    m = len(x)\n",
    "    n = len(y)\n",
    "    i = m\n",
    "    j = n\n",
    "    while i != 0 and j != 0:\n",
    "        if pointer[i][j] == 3:\n",
    "            ret[0].append(x[i-1])\n",
    "            ret[1].append(y[j-1])\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif pointer[i][j]== 1:\n",
    "            ret[0].append(x[i-1])\n",
    "            ret[1].append('-')\n",
    "            i -= 1\n",
    "        else:\n",
    "            ret[0].append('-')\n",
    "            ret[1].append(y[j-1])\n",
    "            j -=1\n",
    "    if i == 0:\n",
    "        ret[1].append(y[:j][::-1])\n",
    "        ret[0].append('-'*j)\n",
    "    elif j ==0:\n",
    "        ret[0].append(x[:i][::-1])\n",
    "        ret[1].append('-'*i)\n",
    "    return ''.join(ret[0][::-1]),''.join(ret[1][::-1])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "str1 = \"PLEASANTLY\"\n",
    "str2 = \"MEANLY\"\n",
    "indel = 5\n",
    "score,pointer = global_alignment(str1,str2,indel)\n",
    "aln1,aln2 = back_track(pointer,str1,str2)\n",
    "print('\\n'.join([str(score),aln1,aln2]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_no = {'A':0,'C':1,'D':2,'E':3,'F':4,'G':5,'H':6,'I':7,'K':8,'L':9,'M':10,'N':11,'P':12,'Q':13,'R':14,'S':15,'T':16,'V':17,'W':18,'Y':19}\n",
    "\n",
    "score_BLOSUM = [[4,0,-2,-1,-2,0,-2,-1,-1,-1,-1,-2,-1,-1,-1,1,0,0,-3,-2],\n",
    "                [0,9,-3,-4,-2,-3,-3,-1,-3,-1,-1,-3,-3,-3,-3,-1,-1,-1,-2,-2],\n",
    "                [-2,-3,6,2,-3,-1,-1,-3,-1,-4,-3,1,-1,0,-2,0,-1,-3,-4,-3],\n",
    "                [-1,-4,2,5,-3,-2,0,-3,1,-3,-2,0,-1,2,0,0,-1,-2,-3,-2],\n",
    "                [-2,-2,-3,-3,6,-3,-1,0,-3,0,0,-3,-4,-3,-3,-2,-2,-1,1,3],\n",
    "                [0,-3,-1,-2,-3,6,-2,-4,-2,-4,-3,0,-2,-2,-2,0,-2,-3,-2,-3],\n",
    "                [-2,-3,-1,0,-1,-2,8,-3,-1,-3,-2,1,-2,0,0,-1,-2,-3,-2,2],\n",
    "                [-1,-1,-3,-3,0,-4,-3,4,-3,2,1,-3,-3,-3,-3,-2,-1,3,-3,-1],\n",
    "                [-1,-3,-1,1,-3,-2,-1,-3,5,-2,-1,0,-1,1,2,0,-1,-2,-3,-2],\n",
    "                [-1,-1,-4,-3,0,-4,-3,2,-2,4,2,-3,-3,-2,-2,-2,-1,1,-2,-1],\n",
    "                [-1,-1,-3,-2,0,-3,-2,1,-1,2,5,-2,-2,0,-1,-1,-1,1,-1,-1],\n",
    "                [-2,-3,1,0,-3,0,1,-3,0,-3,-2,6,-2,0,0,1,0,-3,-4,-2],\n",
    "                [-1,-3,-1,-1,-4,-2,-2,-3,-1,-3,-2,-2,7,-1,-2,-1,-1,-2,-4,-3],\n",
    "                [-1,-3,0,2,-3,-2,0,-3,1,-2,0,0,-1,5,1,0,-1,-2,-2,-1],\n",
    "                [-1,-3,-2,0,-3,-2,0,-3,2,-2,-1,0,-2,1,5,-1,-1,-3,-3,-2],\n",
    "                [1,-1,0,0,-2,0,-1,-2,0,-2,-1,1,-1,0,-1,4,1,-2,-3,-2],\n",
    "                [0,-1,-1,-1,-2,-2,-2,-1,-1,-1,-1,0,-1,-1,-1,1,5,0,-2,-2],\n",
    "                [0,-1,-3,-2,-1,-3,-3,3,-2,1,1,-3,-2,-2,-3,-2,0,4,-3,-1],\n",
    "                [-3,-2,-4,-3,1,-2,-2,-3,-3,-2,-1,-4,-4,-2,-3,-3,-2,-3,11,2],\n",
    "                [-2,-2,-3,-2,3,-3,2,-1,-2,-1,-1,-2,-3,-1,-2,-2,-2,-1,2,7]]\n",
    "\n",
    "\n",
    "def getscore(AA1,AA2):\n",
    "    return score_BLOSUM[aa_no[AA1]][aa_no[AA2]]\n",
    "\n",
    "indel = -5\n",
    "\n",
    "def global_allignment(s1,s2,indel_penalty):\n",
    "    n = len(s1)\n",
    "    m = len(s2)\n",
    "    s = [[0 for repeat_j in range(len(s2)+1)] for repeat_i in range(len(s1)+1)]\n",
    "    backtrack = [['0' for repeat_j in range(len(s2)+1)] for repeat_i in range(len(s1)+1)]\n",
    "    for i in range(1,n+1):\n",
    "        s[i][0] = s[i-1][0] + indel_penalty\n",
    "        backtrack[i][0] = 'U'\n",
    "    for j in range(1,m+1):\n",
    "        s[0][j] = s[0][j-1] + indel_penalty\n",
    "        backtrack[0][j] = 'L'\n",
    "    for i in range(1,n+1):\n",
    "        for j in range(1,m+1):\n",
    "            up_score = s[i-1][j] + indel_penalty\n",
    "            left_score = s[i][j-1] + indel_penalty\n",
    "            diag_score = s[i-1][j-1] + getscore(s1[i-1],s2[j-1])\n",
    "            s[i][j] = max(left_score,up_score,diag_score)\n",
    "            if diag_score >= up_score:\n",
    "                if diag_score >= left_score:\n",
    "                    backtrack[i][j] = 'D'\n",
    "                elif diag_score < left_score:\n",
    "                    backtrack[i][j] = 'L'\n",
    "            elif diag_score < up_score:\n",
    "                if up_score >= left_score:\n",
    "                    backtrack[i][j] = 'U'\n",
    "                else:\n",
    "                    backtrack[i][j] = 'L'\n",
    "\n",
    "    return s[n][m],backtrack\n",
    "\n",
    "def backtrack_allignment(seq1,seq2,indel):\n",
    "    backtrack = global_allignment(seq1,seq2,indel)[1]\n",
    "    insert_indel = lambda word, i: word[:i] + '-' + word[i:]\n",
    "    seq1_aligned, seq2_aligned = seq1, seq2\n",
    "    i, j = len(seq1), len(seq2)\n",
    "    while 1:\n",
    "        if backtrack[i][j] == '0':\n",
    "            break\n",
    "        if backtrack[i][j] == 'U':\n",
    "            i -= 1\n",
    "            seq2_aligned = insert_indel(seq2_aligned, j)\n",
    "        elif backtrack[i][j] == 'L':\n",
    "            j -= 1\n",
    "            seq1_aligned = insert_indel(seq1_aligned, i)\n",
    "        else:\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "    for repeat in range(i):\n",
    "        seq2_aligned = insert_indel(seq2_aligned, 0)\n",
    "    for repeat in range(j):\n",
    "        seq1_aligned = insert_indel(seq1_aligned, 0)\n",
    "    return global_allignment(seq1,seq2,indel)[0],seq1_aligned,seq2_aligned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq1 = 'PLEASANTLY'\n",
    "seq2 = 'MEANLY'\n",
    "result =  backtrack_allignment(seq1,seq2,indel)\n",
    "for res in result:\n",
    "    print(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Local Alignment Problem\n",
    "Find the highest-scoring local alignment between two strings.\n",
    "Given: Two protein strings.\n",
    "Return: The maximum score of a local alignment of the strings, followed by a \n",
    "local alignment of these strings achieving the maximum score.\n",
    "https://github.com/ngaude/sandbox/blob/f009d5a50260ce26a69cd7b354f6d37b48937ee5/bioinformatics-002/bioinformatics_chapter5.py\n",
    "https://github.com/ahmdeen/Bioinformatic-Algorthims/blob/7cae61bb45535911f220da290f9ecf609b964058/Ros5F_LocalAlignment.py\n",
    "\"\"\"\n",
    "aa_no = {'A':0,'C':1,'D':2,'E':3,'F':4,'G':5,'H':6,'I':7,'K':8,'L':9,'M':10,'N':11,'P':12,'Q':13,'R':14,'S':15,'T':16,'V':17,'W':18,'Y':19}\n",
    "\n",
    "score_PAM250 = [[2,-2,0,0,-3,1,-1,-1,-1,-2,-1,0,1,0,-2,1,1,0,-6,-3],\n",
    "                [-2,12,-5,-5,-4,-3,-3,-2,-5,-6,-5,-4,-3,-5,-4,0,-2,-2,-8,0],\n",
    "                [0,-5,4,3,-6,1,1,-2,0,-4,-3,2,-1,2,-1,0,0,-2,-7,-4],\n",
    "                [0,-5,3,4,-5,0,1,-2,0,-3,-2,1,-1,2,-1,0,0,-2,-7,-4],\n",
    "                [-3,-4,-6,-5,9,-5,-2,1,-5,2,0,-3,-5,-5,-4,-3,-3,-1,0,7],\n",
    "                [1,-3,1,0,-5,5,-2,-3,-2,-4,-3,0,0,-1,-3,1,0,-1,-7,-5],\n",
    "                [-1,-3,1,1,-2,-2,6,-2,0,-2,-2,2,0,3,2,-1,-1,-2,-3,0],\n",
    "                [-1,-2,-2,-2,1,-3,-2,5,-2,2,2,-2,-2,-2,-2,-1,0,4,-5,-1],\n",
    "                [-1,-5,0,0,-5,-2,0,-2,5,-3,0,1,-1,1,3,0,0,-2,-3,-4],\n",
    "                [-2,-6,-4,-3,2,-4,-2,2,-3,6,4,-3,-3,-2,-3,-3,-2,2,-2,-1],\n",
    "                [-1,-5,-3,-2,0,-3,-2,2,0,4,6,-2,-2,-1,0,-2,-1,2,-4,-2],\n",
    "                [0,-4,2,1,-3,0,2,-2,1,-3,-2,2,0,1,0,1,0,-2,-4,-2],\n",
    "                [1,-3,-1,-1,-5,0,0,-2,-1,-3,-2,0,6,0,0,1,0,-1,-6,-5],\n",
    "                [0,-5,2,2,-5,-1,3,-2,1,-2,-1,1,0,4,1,-1,-1,-2,-5,-4],\n",
    "                [-2,-4,-1,-1,-4,-3,2,-2,3,-3,0,0,0,1,6,0,-1,-2,2,-4],\n",
    "                [1,0,0,0,-3,1,-1,-1,0,-3,-2,1,1,-1,0,2,1,-1,-2,-3],\n",
    "                [1,-2,0,0,-3,0,-1,0,0,-2,-1,0,0,-1,-1,1,3,0,-5,-3],\n",
    "                [0,-2,-2,-2,-1,-1,-2,4,-2,2,2,-2,-1,-2,-2,-1,0,4,-6,-2],\n",
    "                [-6,-8,-7,-7,0,-7,-3,-5,-3,-2,-4,-4,-6,-5,2,-2,-5,-6,17,0],\n",
    "                [-3,0,-4,-4,7,-5,0,-1,-4,-1,-2,-2,-5,-4,-4,-3,-3,-2,0,10]]\n",
    "\n",
    "\n",
    "def getscore(AA1,AA2):\n",
    "    return score_PAM250[aa_no[AA1]][aa_no[AA2]]\n",
    "\n",
    "indel = -5\n",
    "\n",
    "\n",
    "def local_allignment(s1,s2,indel_penalty):\n",
    "    n = len(s1)\n",
    "    m = len(s2)\n",
    "    s = [[0 for repeat_j in range(len(s2)+1)] for repeat_i in range(len(s1)+1)]\n",
    "    backtrack = [['0' for repeat_j in range(len(s2)+1)] for repeat_i in range(len(s1)+1)]\n",
    "    for i in range(1,n+1):\n",
    "        s[i][0] = s[i-1][0]\n",
    "        #backtrack[i][0] = 'Up'\n",
    "    for j in range(1,m+1):\n",
    "        s[0][j] = s[0][j-1]\n",
    "        #backtrack[0][j] = 'Left'\n",
    "    for i in range(1,n+1):\n",
    "        for j in range(1,m+1):\n",
    "            up_score = s[i-1][j] + indel_penalty\n",
    "            left_score = s[i][j-1] + indel_penalty\n",
    "            diag_score = s[i-1][j-1] + getscore(s1[i-1],s2[j-1])\n",
    "            s[i][j] = max(0,left_score,up_score,diag_score)\n",
    "            if s[i][j] == 0:\n",
    "                backtrack[i][j] = '0'\n",
    "            if s[i][j] == up_score:\n",
    "                backtrack[i][j] = 'U'\n",
    "            if s[i][j] == left_score:\n",
    "                backtrack[i][j] = 'L'\n",
    "            if s[i][j] == diag_score:\n",
    "                backtrack[i][j] = 'D'\n",
    "    max_score = 0\n",
    "    max_i = 0\n",
    "    max_j = 0\n",
    "    for i in range(1,n+1):\n",
    "        for j in range(1,m+1):\n",
    "            if s[i][j]>max_score:\n",
    "                max_score=s[i][j]\n",
    "                max_i = i\n",
    "                max_j = j\n",
    "\n",
    "    return max_score,max_i,max_j,backtrack\n",
    "\n",
    "def backtrack_local(s1,s2,indel_penalty):\n",
    "    score,maxi,maxj,backtrack = local_allignment(s1,s2,indel_penalty)\n",
    "    align1 = ''\n",
    "    align2 = ''\n",
    "    i,j = maxi,maxj\n",
    "    while backtrack[i][j] != '0':\n",
    "        if backtrack[i][j] == 'D':\n",
    "            align1 += s1[i-1]\n",
    "            align2 += s2[j-1]\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif backtrack[i][j] == 'L':\n",
    "            align1 += '-'\n",
    "            align2 += s2[j-1]\n",
    "            j -= 1\n",
    "        elif backtrack[i][j] == 'U':\n",
    "            align1 += s1[i-1]\n",
    "            align2 += '-'\n",
    "            i -= 1\n",
    "\n",
    "    align1 = align1[::-1]\n",
    "    align2 = align2[::-1]\n",
    "\n",
    "    return score,align1,align2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq1 = 'MEANLY'\n",
    "seq2 = 'PENALTY'\n",
    "ans =  backtrack_local(seq1,seq2,indel)\n",
    "for res in ans:\n",
    "    print(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getScore():\n",
    "    with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/PAM250.txt', 'r') as f:\n",
    "        lines = f.read().splitlines()\n",
    "    alphabet = lines[0].lstrip().split()\n",
    "    score = []\n",
    "    for i in range(1,len(lines)):\n",
    "        score.append(list(map(int,lines[i][1:].lstrip().split())))\n",
    "    return score,alphabet\n",
    "\n",
    "def local_alignment(x,y,indel):\n",
    "    \"down:1, right2: diagnoal:3\"\n",
    "    score,alphabet = getScore()\n",
    "    # create the distance matrix\n",
    "    D = []\n",
    "    pointer = []\n",
    "    for i in range(len(x)+1):\n",
    "        D.append([0]*(len(y)+1))\n",
    "        pointer.append([-1]*(len(y)+1))\n",
    "\n",
    "    # fill in the distance matrix and keep the pointer\n",
    "    for i in range(1,len(x)+1):\n",
    "        for j in range(1,len(y)+1):\n",
    "            distHor = D[i][j-1] - indel\n",
    "            distVer = D[i-1][j] - indel\n",
    "            distDiag = D[i-1][j-1] + score[alphabet.index(x[i-1])][alphabet.index(y[j-1])]\n",
    "            candi = [0,distVer,distHor,distDiag]\n",
    "            D[i][j] = max(candi)\n",
    "            direction = candi.index(D[i][j])\n",
    "            '''preference: diag, down, right'''\n",
    "            if direction == 0:\n",
    "                pointer[i][j] = -1 #STOP\n",
    "            elif direction == 1:\n",
    "                pointer[i][j] = 1#down\n",
    "            elif direction == 2:\n",
    "                pointer[i][j] = 2#right\n",
    "            elif direction == 3:\n",
    "                pointer[i][j] = 3#diag\n",
    "    return D,pointer\n",
    "\n",
    "def back_track(D,pointer,x,y):\n",
    "    ret = [[],[]]\n",
    "    maxrow = [max(row) for row in D]\n",
    "    maxcol = [max(col) for col in zip(*D)]\n",
    "    maxscore = max(maxcol)\n",
    "    i = maxrow.index(maxscore)\n",
    "    j = maxcol.index(maxscore)\n",
    "    while i != 0 and j != 0:\n",
    "        if pointer[i][j] == -1:\n",
    "            break\n",
    "        elif pointer[i][j] == 3:\n",
    "            ret[0].append(x[i-1])\n",
    "            ret[1].append(y[j-1])\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif pointer[i][j]== 1:\n",
    "            ret[0].append(x[i-1])\n",
    "            ret[1].append('-')\n",
    "            i -= 1\n",
    "        else:\n",
    "            ret[0].append('-')\n",
    "            ret[1].append(y[j-1])\n",
    "            j -=1\n",
    "    return maxscore,''.join(ret[0][::-1]),''.join(ret[1][::-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "str1 = \"MEANLY\"\n",
    "str2 = \"PENALTY\"\n",
    "indel = 5\n",
    "score,pointer = local_alignment(str1,str2,indel)\n",
    "maxscore,aln1,aln2 = back_track(score,pointer,str1,str2)\n",
    "print('\\n'.join([str(maxscore),aln1,aln2]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Edit Distance Problem\n",
    "Find the edit distance between two strings.\n",
    "Given: Two protein strings.\n",
    "Return: The edit distance between these strings.\n",
    "Levenshtein introduced edit distance but did not describe an algorithm for computing it, which we leave to you.\n",
    "Edit Distance Problem: Find the edit distance between two strings.\n",
    "     Input: Two strings.\n",
    "     Output: The edit distance between these strings.\n",
    "CODE CHALLENGE: Solve the Edit Distance Problem.\n",
    "Sample Input:\n",
    "     PLEASANTLY\n",
    "     MEANLY\n",
    "Sample Output:\n",
    "     5\n",
    "\"\"\"\n",
    "def edit_distance(s1,s2):\n",
    "    n = len(s1)\n",
    "    m = len(s2)\n",
    "    dp = np.zeros((n+1,m+1),dtype = int)\n",
    "    for i in range(1,n+1):\n",
    "        dp[i][0] = i\n",
    "    for j in range(1,m+1):\n",
    "        dp[0][j] = j\n",
    "    for i in range(1,n+1):\n",
    "        for j in range(1,m+1):\n",
    "            if s1[i-1] == s2[j-1]:\n",
    "                dp[i][j] = dp[i-1][j-1]\n",
    "            else:\n",
    "                dp[i][j] = 1 + min(dp[i-1][j-1],dp[i-1][j],dp[i][j-1])\n",
    "\n",
    "    return dp[n][m]\n",
    "\n",
    "print(edit_distance(seq1,seq2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Fitting Alignment Problem\n",
    "Construct a highest-scoring fitting alignment between two strings.\n",
    "Given: Two DNA strings v and w, where v has length at most 10000 and w has length at most 1000.\n",
    "Return: The maximum score of a fitting alignment of v and w, followed by a fitting alignment \n",
    "achieving this maximum score. Use the simple scoring method in which matches count +1 and both the \n",
    "mismatch and indel penalties are equal to 1. (If multiple fitting alignments achieving the maximum \n",
    "score exist, you may return any one.)\n",
    "CODE CHALLENGE: Solve the Fitting Alignment Problem.\n",
    "     Input: Two nucleotide strings v and w, where v has length at most 1000 and w has length at most 100.\n",
    "     Output: A highest-scoring fitting alignment between v and w. Use the simple scoring method in which\n",
    "     matches count +1 and both the mismatch and indel penalties are 1.\n",
    "Sample Input:\n",
    "     GTAGGCTTAAGGTTA\n",
    "     TAGATA\n",
    "Sample Output:\n",
    "     2\n",
    "     TAGGCTTA\n",
    "     TAGA--TA\n",
    "\"\"\"\n",
    "\n",
    "def fitAlignment(seq1, seq2):\n",
    "    '''Returns the maximum score of a fitting alignment of the input sequences, followed by the alignment'''\n",
    "\n",
    "\n",
    "    len1, len2 = len(seq1), len(seq2)\n",
    "    #Initialize Score and Backtrack Matricies as 0 matricies\n",
    "\n",
    "    S = [[0 for j in range(len2 + 1)] for i in range(len1 + 1)]\n",
    "    backtrack = [[0 for j in range(len2 + 1)] for i in range(len1 + 1)]\n",
    "\n",
    "    #Complete the entries in the matricies\n",
    "    for i in range(1, len1 + 1):\n",
    "        for j in range(1, len2 + 1):\n",
    "            scoreList = [S[i-1][j] - 1, S[i][j-1] - 1, S[i-1][j-1] + [-1, 1][seq1[i-1] == seq2[j-1]]] #*\n",
    "            S[i][j] = max(scoreList)\n",
    "            backtrack[i][j] = scoreList.index(S[i][j])\n",
    "\n",
    "    #Get the score of the end alignment of seq2, the shorter sequence\n",
    "\n",
    "    b = len2\n",
    "    a = max(enumerate([S[row][b] for row in range(len2, len1)]), key = lambda y:y[1])[0] + len2\n",
    "    maxScore = str(S[a][b])\n",
    "\n",
    "    #The initial alignment of seq1 to seq2\n",
    "    align1, align2 = seq1[:a], seq2[:b]\n",
    "\n",
    "    #-----\n",
    "    insertIndelFcn = lambda seq, i: seq[:i] + '-' + seq[i:]\n",
    "    #-----\n",
    "\n",
    "    #Backtrack to find fitting alignment\n",
    "\n",
    "    while a*b != 0:\n",
    "        if backtrack[a][b] == 0:\n",
    "            a -= 1\n",
    "            align2 = insertIndelFcn(align2, b)\n",
    "        elif backtrack[a][b] == 1:\n",
    "            b -= 1\n",
    "            align1 = insertIndelFcn(align1, a)\n",
    "        elif backtrack[a][b] == 2:\n",
    "            a -= 1\n",
    "            b -= 1\n",
    "\n",
    "    # Need to cut off seq1 at the ending point\n",
    "    align1 = align1[a:]\n",
    "\n",
    "    return maxScore, align1, align2\n",
    "\n",
    "answer = fitAlignment(seq1, seq2)\n",
    "print('\\n'.join(answer))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Overlap Alignment Problem\n",
    "Construct a highest-scoring overlap alignment between two strings.\n",
    "Given: Two protein strings s and t, each of length at most 1000.\n",
    "Return: The score of an optimal overlap alignment of v and w, followed by an alignment \n",
    "CODE CHALLENGE: Solve the Overlap Alignment Problem.\n",
    "     Input: Two strings v and w, each of length at most 1000.\n",
    "     Output: The score of an optimal overlap alignment of v and w, followed by an alignment of a suffix v' of\n",
    "     v and a prefix w' of w achieving this maximum score. Use an alignment score in which matches count\n",
    "     +1 and both the mismatch and indel penalties are 2.\n",
    "Sample Input:\n",
    "     PAWHEAE\n",
    "     HEAGAWGHEE\n",
    "Sample Output:\n",
    "     1\n",
    "     HEAE\n",
    "     HEAG\n",
    "\"\"\"\n",
    "\n",
    "def overlapAlignment(seq1, seq2):\n",
    "    '''Returns the score of the overlapAlignment as well as the alignment of the two input sequences'''\n",
    "\n",
    "    len1, len2 = len(seq1), len(seq2)\n",
    "\n",
    "    #Initialize Score and backtrack matricies\n",
    "    S = [[0 for j in range(len2 + 1)] for i in range(len1 + 1)]\n",
    "    backtrack = [[0 for j in range(len2 + 1)] for i in range(len1 +1)]\n",
    "\n",
    "    #Inital Max Score**\n",
    "\n",
    "    maxScore = -3*(len1 + len2)\n",
    "\n",
    "    #Complete the Score and backtrack matricies\n",
    "\n",
    "    for i in range(1, len1 + 1):\n",
    "        for j in range(1, len2 + 1):\n",
    "            #Match = 1, Mismatch/Indel = -2\n",
    "\n",
    "            scoreList = [S[i-1][j-1] + [-2, 1][seq1[i-1] == seq2[j-1]], S[i-1][j] - 2, S[i][j-1] - 2]\n",
    "            S[i][j] = max(scoreList)\n",
    "            backtrack[i][j] = scoreList.index(S[i][j])\n",
    "\n",
    "            #Check for maximums along final row/column\n",
    "            if i == len1 or j == len2:\n",
    "                if S[i][j] > maxScore:\n",
    "                    maxScore = S[i][j]\n",
    "                    maxIndex = (i,j)\n",
    "\n",
    "    a, b = maxIndex\n",
    "    align1, align2 = seq1[:a], seq2[:b]\n",
    "\n",
    "    #------------\n",
    "    insertIndelFcn = lambda seq, i:seq[:i] + '-' + seq[i:]\n",
    "\n",
    "    #Backtrack to find Alignment\n",
    "\n",
    "    while a*b != 0:\n",
    "        if backtrack[a][b] == 1:\n",
    "            a -= 1\n",
    "            align2 = insertIndelFcn(align2, b)\n",
    "        elif backtrack[a][b] == 2:\n",
    "            b -= 1\n",
    "            align1 = insertIndelFcn(align1, a)\n",
    "        else:\n",
    "            a -= 1\n",
    "            b -= 1\n",
    "\n",
    "    align1 = align1[a:]\n",
    "    align2 = align2[b:] \n",
    "\n",
    "    return str(maxScore), align1, align2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Alignment with Affine Gap Penalties Problem: Construct a highest-scoring global alignment between two strings (with affine gap penalties).\n",
    "\n",
    "Input: Two strings, a matrix score, and numbers σ and ε.\n",
    "Output: A highest scoring global alignment between these strings, as defined by the scoring matrix score and by the gap opening and extension penalties σ and ε.\n",
    "\n",
    "\n",
    "Code Challenge: Solve the Alignment with Affine Gap Penalties Problem.\n",
    "\n",
    "Input: Two amino acid strings v and w (each of length at most 100).\n",
    "Output: The maximum alignment score between v and w, followed by an alignment of v and w achieving this maximum score. Use the BLOSUM62 scoring matrix, a gap opening penalty of 11, and a gap extension penalty of 1.\n",
    "\n",
    "Sample Input:\n",
    "PRTEINS\n",
    "PRTWPSEIN\n",
    "Sample Output:\n",
    "8\n",
    "PRT---EINS\n",
    "PRTWPSEIN-\n",
    "\n",
    "\"\"\"\n",
    "aa_no = {'A':0,'C':1,'D':2,'E':3,'F':4,'G':5,'H':6,'I':7,'K':8,'L':9,'M':10,'N':11,'P':12,'Q':13,'R':14,'S':15,'T':16,'V':17,'W':18,'Y':19}\n",
    "\n",
    "score_BLOSUM = [[4,0,-2,-1,-2,0,-2,-1,-1,-1,-1,-2,-1,-1,-1,1,0,0,-3,-2],\n",
    "                [0,9,-3,-4,-2,-3,-3,-1,-3,-1,-1,-3,-3,-3,-3,-1,-1,-1,-2,-2],\n",
    "                [-2,-3,6,2,-3,-1,-1,-3,-1,-4,-3,1,-1,0,-2,0,-1,-3,-4,-3],\n",
    "                [-1,-4,2,5,-3,-2,0,-3,1,-3,-2,0,-1,2,0,0,-1,-2,-3,-2],\n",
    "                [-2,-2,-3,-3,6,-3,-1,0,-3,0,0,-3,-4,-3,-3,-2,-2,-1,1,3],\n",
    "                [0,-3,-1,-2,-3,6,-2,-4,-2,-4,-3,0,-2,-2,-2,0,-2,-3,-2,-3],\n",
    "                [-2,-3,-1,0,-1,-2,8,-3,-1,-3,-2,1,-2,0,0,-1,-2,-3,-2,2],\n",
    "                [-1,-1,-3,-3,0,-4,-3,4,-3,2,1,-3,-3,-3,-3,-2,-1,3,-3,-1],\n",
    "                [-1,-3,-1,1,-3,-2,-1,-3,5,-2,-1,0,-1,1,2,0,-1,-2,-3,-2],\n",
    "                [-1,-1,-4,-3,0,-4,-3,2,-2,4,2,-3,-3,-2,-2,-2,-1,1,-2,-1],\n",
    "                [-1,-1,-3,-2,0,-3,-2,1,-1,2,5,-2,-2,0,-1,-1,-1,1,-1,-1],\n",
    "                [-2,-3,1,0,-3,0,1,-3,0,-3,-2,6,-2,0,0,1,0,-3,-4,-2],\n",
    "                [-1,-3,-1,-1,-4,-2,-2,-3,-1,-3,-2,-2,7,-1,-2,-1,-1,-2,-4,-3],\n",
    "                [-1,-3,0,2,-3,-2,0,-3,1,-2,0,0,-1,5,1,0,-1,-2,-2,-1],\n",
    "                [-1,-3,-2,0,-3,-2,0,-3,2,-2,-1,0,-2,1,5,-1,-1,-3,-3,-2],\n",
    "                [1,-1,0,0,-2,0,-1,-2,0,-2,-1,1,-1,0,-1,4,1,-2,-3,-2],\n",
    "                [0,-1,-1,-1,-2,-2,-2,-1,-1,-1,-1,0,-1,-1,-1,1,5,0,-2,-2],\n",
    "                [0,-1,-3,-2,-1,-3,-3,3,-2,1,1,-3,-2,-2,-3,-2,0,4,-3,-1],\n",
    "                [-3,-2,-4,-3,1,-2,-2,-3,-3,-2,-1,-4,-4,-2,-3,-3,-2,-3,11,2],\n",
    "                [-2,-2,-3,-2,3,-3,2,-1,-2,-1,-1,-2,-3,-1,-2,-2,-2,-1,2,7]]\n",
    "\n",
    "\n",
    "def getscore(AA1,AA2):\n",
    "    return score_BLOSUM[aa_no[AA1]][aa_no[AA2]]\n",
    "\n",
    "sigma = -11\n",
    "epsilon = -1\n",
    "\n",
    "def global_allignment_affine_gaps(s1,s2,sigma,epsilon):\n",
    "    n = len(s1)\n",
    "    m = len(s2)\n",
    "\n",
    "    s = [[[0 for i in range(m+1)] for j in range(n+1)] for k in range(3)]\n",
    "    backtrack = [[[0 for i in range(m+1)] for j in range(n+1)] for k in range(3)]\n",
    "\n",
    "    for i in range(1,n+1):\n",
    "        s[0][i][0] = sigma + (i-1)*epsilon\n",
    "        s[1][i][0] = sigma + (i-1)*epsilon\n",
    "        s[2][i][0] = 10*sigma\n",
    "\n",
    "    for j in range(1,m+1):\n",
    "        s[0][0][j] = 10*sigma\n",
    "        s[1][0][j] = sigma + (j-1)*epsilon\n",
    "        s[2][0][j] = sigma + (j-1)*epsilon\n",
    "\n",
    "    for i in range(1,n+1):\n",
    "        for j in range(1,m+1):\n",
    "            low_scores = [s[0][i-1][j] + epsilon,s[1][i-1][j] + sigma]\n",
    "            s[0][i][j] = max(low_scores)\n",
    "            backtrack[0][i][j] = low_scores.index(s[0][i][j])\n",
    "\n",
    "            up_scores = [s[2][i][j-1] + epsilon,s[1][i][j-1] + sigma]\n",
    "            s[2][i][j] = max(up_scores)\n",
    "            backtrack[2][i][j] = up_scores.index(s[2][i][j])\n",
    "\n",
    "            mid_scores = [s[0][i][j],s[1][i-1][j-1] + getscore(s1[i-1],s2[j-1]),s[2][i][j]]\n",
    "            s[1][i][j] = max(mid_scores)\n",
    "            backtrack[1][i][j] = mid_scores.index(s[1][i][j])\n",
    "\n",
    "    max_scores = [s[0][i][j],s[1][i][j],s[2][i][j]]\n",
    "    max_score = max(max_scores)\n",
    "    backtrack_matrix = max_scores.index(max_score)\n",
    "\n",
    "    insert_indel = lambda word, i: word[:i] + '-' + word[i:]\n",
    "\n",
    "    s1_allign = s1\n",
    "    s2_allign = s2\n",
    "\n",
    "    while i*j != 0:\n",
    "        if backtrack_matrix == 0:\n",
    "            if backtrack[0][i][j] == 1:\n",
    "                backtrack_matrix = 1\n",
    "            i -= 1\n",
    "            s2_allign  = insert_indel(s2_allign,j)\n",
    "        elif backtrack_matrix == 1:\n",
    "            if backtrack[1][i][j] == 0:\n",
    "                backtrack_matrix = 0\n",
    "            elif backtrack[1][i][j] == 2:\n",
    "                backtrack_matrix = 2\n",
    "            else:\n",
    "                i -= 1\n",
    "                j -= 1\n",
    "        else:\n",
    "            if backtrack[2][i][j] == 1:\n",
    "                backtrack_matrix = 1\n",
    "            j -= 1\n",
    "            s1_allign = insert_indel(s1_allign,i)\n",
    "\n",
    "    for _ in range(i):\n",
    "        s2_allign = insert_indel(s2_allign,0)\n",
    "\n",
    "    for _ in range(j):\n",
    "        s1_allign = insert_indel(s1_allign,0) \n",
    "\n",
    "    return max_score,s1_allign,s2_allign\n",
    "\n",
    "seq1 = \"PRTEINS\"\n",
    "seq2 = \"PRTWPSEIN\"\n",
    "ans = global_allignment_affine_gaps(seq1,seq2,sigma,epsilon)\n",
    "for res in ans:\n",
    "    print(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Middle Edge in Linear Space Problem: Find a middle edge in the alignment graph in linear space.\n",
    "\n",
    "Input: Two strings and a matrix score.\n",
    "Output: A middle edge in the alignment graph of these strings (where the edge lengths are defined by score).\n",
    "\n",
    "Code Challenge: Solve the Middle Edge in Linear Space Problem (for protein strings).\n",
    "\n",
    "Input: Two amino acid strings.\n",
    "Output: A middle edge in the alignment graph in the form \"(i, j) (k, l)\", where (i, j) connects to (k, l). To compute scores, use the BLOSUM62 scoring matrix and a (linear) indel penalty equal to 5.\n",
    "\n",
    "i,j,k,l should be integers\n",
    "\n",
    "\"\"\"\n",
    "aa_no = {'A':0,'C':1,'D':2,'E':3,'F':4,'G':5,'H':6,'I':7,'K':8,'L':9,'M':10,'N':11,'P':12,'Q':13,'R':14,'S':15,'T':16,'V':17,'W':18,'Y':19}\n",
    "\n",
    "score_BLOSUM = [[4,0,-2,-1,-2,0,-2,-1,-1,-1,-1,-2,-1,-1,-1,1,0,0,-3,-2],\n",
    "                [0,9,-3,-4,-2,-3,-3,-1,-3,-1,-1,-3,-3,-3,-3,-1,-1,-1,-2,-2],\n",
    "                [-2,-3,6,2,-3,-1,-1,-3,-1,-4,-3,1,-1,0,-2,0,-1,-3,-4,-3],\n",
    "                [-1,-4,2,5,-3,-2,0,-3,1,-3,-2,0,-1,2,0,0,-1,-2,-3,-2],\n",
    "                [-2,-2,-3,-3,6,-3,-1,0,-3,0,0,-3,-4,-3,-3,-2,-2,-1,1,3],\n",
    "                [0,-3,-1,-2,-3,6,-2,-4,-2,-4,-3,0,-2,-2,-2,0,-2,-3,-2,-3],\n",
    "                [-2,-3,-1,0,-1,-2,8,-3,-1,-3,-2,1,-2,0,0,-1,-2,-3,-2,2],\n",
    "                [-1,-1,-3,-3,0,-4,-3,4,-3,2,1,-3,-3,-3,-3,-2,-1,3,-3,-1],\n",
    "                [-1,-3,-1,1,-3,-2,-1,-3,5,-2,-1,0,-1,1,2,0,-1,-2,-3,-2],\n",
    "                [-1,-1,-4,-3,0,-4,-3,2,-2,4,2,-3,-3,-2,-2,-2,-1,1,-2,-1],\n",
    "                [-1,-1,-3,-2,0,-3,-2,1,-1,2,5,-2,-2,0,-1,-1,-1,1,-1,-1],\n",
    "                [-2,-3,1,0,-3,0,1,-3,0,-3,-2,6,-2,0,0,1,0,-3,-4,-2],\n",
    "                [-1,-3,-1,-1,-4,-2,-2,-3,-1,-3,-2,-2,7,-1,-2,-1,-1,-2,-4,-3],\n",
    "                [-1,-3,0,2,-3,-2,0,-3,1,-2,0,0,-1,5,1,0,-1,-2,-2,-1],\n",
    "                [-1,-3,-2,0,-3,-2,0,-3,2,-2,-1,0,-2,1,5,-1,-1,-3,-3,-2],\n",
    "                [1,-1,0,0,-2,0,-1,-2,0,-2,-1,1,-1,0,-1,4,1,-2,-3,-2],\n",
    "                [0,-1,-1,-1,-2,-2,-2,-1,-1,-1,-1,0,-1,-1,-1,1,5,0,-2,-2],\n",
    "                [0,-1,-3,-2,-1,-3,-3,3,-2,1,1,-3,-2,-2,-3,-2,0,4,-3,-1],\n",
    "                [-3,-2,-4,-3,1,-2,-2,-3,-3,-2,-1,-4,-4,-2,-3,-3,-2,-3,11,2],\n",
    "                [-2,-2,-3,-2,3,-3,2,-1,-2,-1,-1,-2,-3,-1,-2,-2,-2,-1,2,7]]\n",
    "\n",
    "\n",
    "def getscore(AA1,AA2):\n",
    "    return score_BLOSUM[aa_no[AA1]][aa_no[AA2]]\n",
    "\n",
    "sigma = -5\n",
    "\n",
    "def middle_col_score(s1,s2,sigma):\n",
    "    n = len(s1)\n",
    "    m = len(s2)\n",
    "    s = [[i*j*(-sigma) for j in range(-1,1)] for i in range(n+1)]\n",
    "    s[0][1] = sigma\n",
    "    backtrack = [0]*(n+1)\n",
    "\n",
    "    for j in range(1,int(m/2) + 1):\n",
    "        for i in range(0,n+1):\n",
    "            if i == 0:\n",
    "                s[i][1] = j*sigma\n",
    "            else:\n",
    "                scores = [s[i-1][0] + getscore(s1[i-1],s2[j-1]),s[i][0] + sigma, s[i-1][1] + sigma]\n",
    "                s[i][1] = max(scores)\n",
    "                backtrack[i] = scores.index(s[i][1])\n",
    "\n",
    "        if j != m/2:\n",
    "            s = [[row[1]]*2 for row in s]\n",
    "\n",
    "    return [row[1] for row in s], backtrack\n",
    "\n",
    "def middle_edge(s1,s2,sigma):\n",
    "    source_to_middle = middle_col_score(seq1, seq2, sigma)[0]\n",
    "    middle_to_sink, backtrack = map(lambda l: l[::-1], middle_col_score(seq1[::-1], seq2[::-1]+['', '$'][len(seq2) % 2 == 1 and len(seq2) > 1], sigma))\n",
    "    scores = list(map(sum, zip(source_to_middle, middle_to_sink)))\n",
    "    max_middle = max(range(len(scores)), key=lambda i: scores[i])\n",
    "\n",
    "    if max_middle == len(scores) - 1:\n",
    "        next_node = (max_middle, len(seq2)/2 + 1)\n",
    "    else:\n",
    "        next_node = [(max_middle + 1, len(seq2)/2 + 1), (max_middle, len(seq2)/2 + 1), (max_middle + 1, len(seq2)/2),][backtrack[max_middle]]\n",
    "\n",
    "    return (max_middle, int(len(seq2)/2)), next_node\n",
    "\n",
    "seq1 = \"PLEASANTLY\"\n",
    "seq2 = \"MEASNLY\"\n",
    "ans =  middle_edge(seq1, seq2, sigma)\n",
    "for res in ans:\n",
    "    print(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Multiple Alignment Problem: Find the highest-scoring alignment between multiple strings under a given scoring matrix.\n",
    "\n",
    "Input: A collection of t strings and a t-dimensional matrix Score.\n",
    "Output: A multiple alignment of these strings whose score (as defined by the matrix Score) is maximized among all possible alignments of these strings.\n",
    "\n",
    "In the Multiple Longest Common Subsequence Problem, the score of a column of the alignment matrix is equal to 1 if all of the column's symbols are identical, and 0 if even one symbol disagrees.\n",
    "\n",
    "Code Challenge: Solve the Multiple Longest Common Subsequence Problem.\n",
    "\n",
    "Input: Three DNA strings of length at most 10.\n",
    "Output: The length of a longest common subsequence of these three strings, followed by a multiple alignment of the three strings corresponding to such an alignment.\n",
    "\n",
    "\"\"\"\n",
    "def multiple_allignment(s1,s2,s3):\n",
    "    n = len(s1)\n",
    "    m = len(s2)\n",
    "    o = len(s3)\n",
    "\n",
    "    s = [[[0 for _ in range(o+1)] for _ in range(m+1)] for _ in range(n+1)]\n",
    "    backtrack = [[[0 for _ in range(o+1)] for _ in range(m+1)] for _ in range(n+1)]\n",
    "\n",
    "    for i in range(1,n+1):\n",
    "        for j in range(1,m+1):\n",
    "            for k in range(1,o+1):\n",
    "                scores = [s[i-1][j-1][k-1] + int(s1[i-1] == s2[j-1] == s3[k-1]), s[i-1][j][k], s[i][j-1][k], s[i][j][k-1], s[i-1][j][k-1], s[i][j-1][k-1]]\n",
    "                backtrack[i][j][k], s[i][j][k] = max(enumerate(scores), key=lambda p: p[1])\n",
    "\n",
    "    insert_indel = lambda word, i: word[:i] +'-'+ word[i:]\n",
    "\n",
    "    s1_allign, s2_allign, s3_allign = s1, s2, s3\n",
    "\n",
    "    max_score = s[n][m][o]\n",
    "\n",
    "    i,j,k = n,m,o\n",
    "\n",
    "    while i*j*k != 0:\n",
    "        if backtrack[i][j][k] == 1:\n",
    "            i -= 1\n",
    "            s2_allign = insert_indel(s2_allign,j)\n",
    "            s3_allign = insert_indel(s3_allign,k)\n",
    "\n",
    "        elif backtrack[i][j][k] ==2:\n",
    "            j -= 1\n",
    "            s1_allign = insert_indel(s1_allign,i)\n",
    "            s3_allign = insert_indel(s3_allign,k)\n",
    "\n",
    "        elif backtrack[i][j][k] == 3:\n",
    "            k -= 1\n",
    "            s1_allign = insert_indel(s1_allign,i)\n",
    "            s2_allign = insert_indel(s3_allign,k)\n",
    "\n",
    "        elif backtrack[i][j][k] == 4:\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "            s3_allign = insert_indel(s3_allign,k)\n",
    "\n",
    "        elif backtrack[i][j][k] == 5:\n",
    "            i -= 1\n",
    "            k -= 1\n",
    "            s2_allign = insert_indel(s2_allign,j)\n",
    "\n",
    "        elif backtrack[i][j][k] == 6:\n",
    "            j -= 1\n",
    "            k -= 1\n",
    "            s1_allign = insert_indel(s1_allign,i)\n",
    "\n",
    "        else:\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "            k -= 1\n",
    "\n",
    "\n",
    "    while len(s1_allign) != max(len(s1_allign),len(s2_allign),len(s3_allign)):\n",
    "        s1_allign = insert_indel(s1_allign,0)\n",
    "    while len(s2_allign) != max(len(s1_allign),len(s2_allign),len(s3_allign)):\n",
    "        s2_allign = insert_indel(s2_allign,0)\n",
    "    while len(s3_allign) != max(len(s1_allign),len(s2_allign),len(s3_allign)):\n",
    "        s3_allign = insert_indel(s3_allign,0)\n",
    "\n",
    "    return str(max_score), s1_allign, s2_allign, s3_allign\n",
    "\n",
    "seq1 = \"TACAAAGTG\"\n",
    "seq2 = \"ACCCAACGTA\"\n",
    "seq3 = \"TAGACGCA\"\n",
    "ans = multiple_allignment(seq1,seq2,seq3)\n",
    "for res in ans:\n",
    "    print(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Implement GreedySorting.\n",
    "     Input: A permutation P.\n",
    "     Output: The sequence of permutations corresponding to applying GreedySorting to P, ending with the identity permutation.\n",
    "Sample Input:\n",
    "(-3 +4 +1 +5 -2)\n",
    "Sample Output:\n",
    "(-1 -4 +3 +5 -2)\n",
    "(+1 -4 +3 +5 -2)\n",
    "(+1 +2 -5 -3 +4)\n",
    "(+1 +2 +3 +5 +4)\n",
    "(+1 +2 +3 -4 -5)\n",
    "(+1 +2 +3 +4 -5)\n",
    "(+1 +2 +3 +4 +5)\n",
    "Pseudocode:\n",
    "    GreedySorting(P)\n",
    "        approxReversalDistance ← 0\n",
    "        for k = 1 to |P|\n",
    "            if element k is not sorted\n",
    "                apply the k-sorting reversal to P\n",
    "                approxReversalDistance ← approxReversalDistance + 1\n",
    "            if k-th element of P is −k\n",
    "                apply the k-sorting reversal to P\n",
    "                approxReversalDistance ← approxReversalDistance + 1\n",
    "        return approxReversalDistance\n",
    "'''\n",
    "\n",
    "def reverse(p):\n",
    "    x = p[::-1]\n",
    "    x = [-i for i in x]\n",
    "    return x\n",
    "\n",
    "\n",
    "def greedysort(p):\n",
    "    n = len(p)\n",
    "    pos = 0\n",
    "    perms = []\n",
    "    while pos < n:\n",
    "        nums = [abs(i) for i in p[pos:]]\n",
    "        min_pos = nums.index(min(nums)) + pos\n",
    "        p[pos:min_pos+1] = reverse(p[pos:min_pos+1])\n",
    "        #print '('+' '.join('+'+str(i) if i>0 else str(i) for i in p)+')'\n",
    "        perms.append('('+' '.join('+'+str(i) if i>0 else str(i) for i in p)+')')\n",
    "        if p[pos] < 0:\n",
    "            p[pos] = -p[pos]\n",
    "            #print '('+' '.join('+'+str(i) if i>0 else str(i) for i in p)+')'\n",
    "            perms.append('('+' '.join('+'+str(i) if i>0 else str(i) for i in p)+')')\n",
    "        pos += 1\n",
    "    return perms\n",
    "\n",
    "P = list(map(int,P[1:-1].split(' ')))\n",
    "#ans = greedysort(P)\n",
    "#print(\"\\n\".join(ans))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Compute the number of steps required by GreedySorting to sort the following permutation (i.e., to transform it into the identity permutation.)\n",
    "\n",
    "(+6 -12 -9 +17 +18 -4 +5 -3 +11 +19 +14 +10 +8 +15 -13 +20 +2 +7 -16 -1)\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "def reverse(p):\n",
    "    x = p[::-1]\n",
    "    x = [-i for i in x]\n",
    "    return x\n",
    "\n",
    "\n",
    "def greedysort(p):\n",
    "    n = len(p)\n",
    "    pos = 0\n",
    "    perms = []\n",
    "    while pos < n:\n",
    "        nums = [abs(i) for i in p[pos:]]\n",
    "        min_pos = nums.index(min(nums)) + pos\n",
    "        p[pos:min_pos+1] = reverse(p[pos:min_pos+1])\n",
    "        #print '('+' '.join('+'+str(i) if i>0 else str(i) for i in p)+')'\n",
    "        perms.append('('+' '.join('+'+str(i) if i>0 else str(i) for i in p)+')')\n",
    "        if p[pos] < 0:\n",
    "            p[pos] = -p[pos]\n",
    "            #print '('+' '.join('+'+str(i) if i>0 else str(i) for i in p)+')'\n",
    "            perms.append('('+' '.join('+'+str(i) if i>0 else str(i) for i in p)+')')\n",
    "        pos += 1\n",
    "    return perms\n",
    "\n",
    "P = '(+2 +6 -8 -17 +7 -14 +18 +3 -5 -16 -11 -19 -4 +10 +13 -20 -1 +9 -12 +15)'\n",
    "P = list(map(int,P[1:-1].split(' ')))\n",
    "print(greedysort(P))\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def greedy(P):\n",
    "    approxReversalDist = 0\n",
    "    perArray = []\n",
    "\n",
    "    for i in range(len(P)):\n",
    "        tempArray = []\n",
    "        if P[i] != i +1:\n",
    "            if i+1 in P:\n",
    "                indexOfI = P.index(abs(i+1))\n",
    "            else:\n",
    "                indexOfI = P.index(-abs(i+1))\n",
    "            for x in list(reversed(P[i:indexOfI+1])):\n",
    "                tempArray.append(-int(x))\n",
    "            P = P[:i] + tempArray + P[indexOfI+1:]\n",
    "            approxReversalDist +=1\n",
    "            tempArray2 = P[:]\n",
    "\n",
    "            for y in range(len(tempArray2)):\n",
    "                if tempArray2[y] > 0:\n",
    "                    tempArray2[y] = \"%+d\" % (tempArray2[y])\n",
    "\n",
    "            perArray.append(tempArray2[:])\n",
    "\n",
    "\n",
    "        if P[i] == -(i+1):\n",
    "            P[i] = i+1\n",
    "            tempArray3 = P[:]\n",
    "            for y in range(len(tempArray2)):\n",
    "                if tempArray3[y] > 0:\n",
    "                    tempArray3[y] = \"%+d\" % (tempArray3[y])\n",
    "\n",
    "            perArray.append(tempArray3[:])\n",
    "\n",
    "\n",
    "    return perArray\n",
    "\n",
    "filename = \"/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/dataset_286_4-2.txt\"\n",
    "\n",
    "file = open(filename, 'r')\n",
    "data = file.read().splitlines()\n",
    "array = []\n",
    "\n",
    "for line in data:\n",
    "    line = line.strip('(')\n",
    "    line = line.strip(')')\n",
    "    array.append([int(x) for x in line.split()])\n",
    "\n",
    "seqArray = []\n",
    "for element in array:\n",
    "    for element2 in element:\n",
    "        seqArray.append(element2)\n",
    "\n",
    "\n",
    "result = greedy(seqArray)\n",
    "\n",
    "for element in result:\n",
    "    print('('+(\" \".join(str(x) for x in element)) + ')')\n",
    " \n",
    "\n",
    "\"\"\"\n",
    "dataset_286_4-2.txt - -52 +5 +72 +71 +69 +64 +108 -36 +44 +15 -80 -99 +81 -130 +83 +85 +90 -14 +6 -42 -127 -31 +41 +13 -56 -94 -18 +110 -62 -11 -68 +51 -65 -4 +2 +102 +16 -97 +26 +37 -103 +10 -95 -93 +86 +101 -33 +73 -124 +54 -35 -113 +126 +132 -111 -25 -19 +74 +40 -58 -67 +77 +70 +121 +1 -88 -29 -98 +49 -17 -135 -55 -3 -118 +104 +96 -136 +66 +61 +46 -89 +119 -92 -107 +59 +114 +30 +9 -115 +129 -109 +32 +105 -84 -76 -27 -78 -112 -39 +23 +48 +117 -91 -133 +12 +45 -82 +43 -75 -57 -50 +8 +53 +131 +134 +7 +63 +24 -106 +28 +34 +20 +87 +22 -79 -21 +122 +123 +100 -47 +60 +125 -120 +38 +128 +116\n",
    "\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Number of Breakpoints Problem: Find the number of breakpoints in a permutation.\n",
    "\n",
    "Input: A permutation.\n",
    "Output: The number of breakpoints in this permutation.\n",
    "\n",
    "Sample Input:\n",
    "+3 +4 +5 -12 -8 -7 -6 +1 +2 +10 +9 -11 +13 +14\n",
    "Sample Output:\n",
    "8\n",
    "\n",
    "\"\"\"\n",
    "def find_break_points(p):\n",
    "    p = [0] + p + [len(p)+1]\n",
    "    num = 0\n",
    "    for i in range(len(p)-1):\n",
    "        if p[i+1]-p[i] != 1:\n",
    "            num += 1\n",
    "    return num\n",
    "\n",
    "P = '(+3 +4 +5 -12 -8 -7 -6 +1 +2 +10 +9 -11 +13 +14)'\n",
    "P = list(map(int,P[1:-1].split(' ')))\n",
    "print find_break_points(P)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "2-Break Distance Theorem: The 2-break distance between genomes P and Q is equal to Blocks(P, Q)− Cycles(P, Q).\n",
    "\n",
    "Proof: We call a transformation of P into Q by 2-breaks a sorting by 2-breaks﻿.  From before, we know that any sorting by 2-breaks must increase the number of alternating cycles by Cycles(Q, Q)− Cycles(P, Q), which is equal to Blocks(P, Q) − Cycles(P, Q) because Blocks(P, Q) = Cycles(Q, Q). The Cycle Theorem implies that each 2-break increases the number of cycles in the breakpoint graph by at most 1. This immediately implies in turn that d(P, Q) is at least Blocks(P, Q) − Cycles(P, Q). If P is not equal to Q, there must be a non-trivial cycle in BreakPointGraph(P, Q), i.e., a cycle with more than two edges. As shown in the figure from the cycle theorem (see middle figure below), any non-trivial cycle in the breakpoint graph can be split into two cycles by a 2-break, implying that we can always find a 2-break increasing the number of red-blue cycles by 1. Therefore, d(P, Q) is equal to Blocks(P, Q) − Cycles(P, Q).\n",
    "\n",
    "Code Challenge: Solve the 2-Break Distance Problem.\n",
    "\n",
    "Input: Genomes P and Q.\n",
    "Output: The 2-break distance d(P, Q).\n",
    "\n",
    "Sample Input:\n",
    "(+1 +2 +3 +4 +5 +6)\n",
    "(+1 -3 -6 -5)(+2 -4)\n",
    "Sample Output:\n",
    "3\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "from collections import defaultdict\n",
    "\n",
    "def two_break_dist(P, Q):\n",
    "    '''Returns the 2-Break Distance of Circular Chromosomes P and Q.'''\n",
    "\n",
    "    # Construct the break point graph of P and Q.\n",
    "    graph = defaultdict(list)\n",
    "    for perm_cycle in P+Q:\n",
    "        L = len(perm_cycle)\n",
    "        for i in range(len(perm_cycle)):\n",
    "            # Add the edge between consecutive items (both orders since the breakpoint graph is undirected).\n",
    "            # Note: Modulo L in the higher index for the edge between the last and first elements.\n",
    "            graph[perm_cycle[i]].append(-1*perm_cycle[(i+1) % L])\n",
    "            graph[-1*perm_cycle[(i+1) % L]].append(perm_cycle[i])\n",
    "\n",
    "    # BFS to find the number of connected components in the breakpoint graph.\n",
    "    component_count = 0\n",
    "    remaining = set(graph.keys())\n",
    "    while len(remaining) > 0:\n",
    "        component_count += 1\n",
    "        queue = [remaining.pop()]  # Components are cyclic, so starting point is unimportant.\n",
    "        while queue:\n",
    "            current = queue.pop(0)\n",
    "            queue += filter(lambda node: node in remaining, graph.get(current, []))\n",
    "            remaining -= set(queue)  # Overkill, but it's nice and concise!\n",
    "\n",
    "    # Theorem: d(P,Q) = blocks(P,Q) - cycles(P,Q)\n",
    "    return sum(map(len,P)) - component_count\n",
    "\n",
    "P = \"(+1 +2 +3 +4 +5 +6)\"\n",
    "P = P.strip().lstrip('(').rstrip(')').split(')(')\n",
    "P = [map(int, block.split()) for block in P]\n",
    "P = [list(a) for a in P]\n",
    "print(P)\n",
    "\n",
    "Q = \"(+1 -3 -6 -5)(+2 -4)\"\n",
    "Q = Q.strip().lstrip('(').rstrip(')').split(')(')\n",
    "Q = [map(int, block.split()) for block in Q]\n",
    "Q = [list(a) for a in Q]\n",
    "print(Q)\n",
    "\n",
    "dist = two_break_dist(P, Q)\n",
    "print(str(dist))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Shared k-mers Problem: Given two strings, find all their shared k-mers.\n",
    "     Input: An integer k and two strings.\n",
    "     Output: All k-mers shared by these strings, in the form of ordered pairs (x, y) corresponding to starting positions\n",
    "     of these k-mers in the respective strings.\n",
    "Sample Input:\n",
    "3\n",
    "AAACTCATC\n",
    "TTTCAAATC\n",
    "Sample Output:\n",
    "(0, 4)\n",
    "(0, 0)\n",
    "(4, 2)\n",
    "(6, 6)\n",
    "'''\n",
    "\n",
    "from collections import defaultdict\n",
    "def rev_comp(sequence):\n",
    "    rev_comp = ''\n",
    "    for nucleotide in sequence:\n",
    "        if nucleotide == 'A':\n",
    "            rev_comp = rev_comp + 'T'\n",
    "        elif nucleotide == 'T':\n",
    "            rev_comp = rev_comp + 'A'\n",
    "        elif nucleotide == 'G':\n",
    "            rev_comp = rev_comp + 'C'\n",
    "        elif nucleotide == 'C':\n",
    "            rev_comp = rev_comp + 'G'\n",
    "    rev_comp = rev_comp[::-1]\n",
    "    return rev_comp\n",
    "\n",
    "def sharedkmer(dna1,dna2):\n",
    "    dna1_dict = defaultdict(list)\n",
    "    for i in range(len(dna1) - k + 1):\n",
    "        dna1_dict[dna1[i:i+k]].append(i)\n",
    "    shared_kmer_indices = set()\n",
    "    for j in range(len(dna2) - k + 1):\n",
    "        shared_kmer_indices |= set(map(lambda x: (x,j), dna1_dict[dna2[j:j+k]]))\n",
    "        shared_kmer_indices |= set(map(lambda x: (x,j), dna1_dict[rev_comp(dna2[j:j+k])]))\n",
    "    return shared_kmer_indices\n",
    "\n",
    "p = \"AAACTCATC\"\n",
    "q = \"TTTCAAATC\"\n",
    "k = 3\n",
    "\n",
    "ans = sharedkmer(p,q)\n",
    "for res in ans:\n",
    "    print(res)\n",
    "    \n",
    "def shared(k, dna1, dna2):\n",
    "\n",
    "    dictionary = dict()\n",
    "\n",
    "    for i in range(len(dna1) - k +1):\n",
    "        pattern = dna1[i:i+k]\n",
    "\n",
    "        if pattern in dictionary:\n",
    "            dictionary[pattern].append(i)\n",
    "        else:\n",
    "            dictionary[pattern] = [i]\n",
    "\n",
    "\n",
    "    for i in range(len(dna2)-k+1):\n",
    "        pattern2 = dna2[i:i+k]\n",
    "        reverse = reverseComp(pattern2)\n",
    "\n",
    "        if pattern2 in dictionary:\n",
    "            for element in dictionary[pattern2]:\n",
    "                print('(' + str(element) + ', ' + str(i) + ')')\n",
    "\n",
    "        if reverse in dictionary:\n",
    "            for element in dictionary[reverse]:\n",
    "                print('(' + str(element) + ', ' + str(i) + ')')\n",
    "\n",
    "\n",
    "def reverseComp(text):\n",
    "    reverseStack = []\n",
    "    for i in range(len(text)):\n",
    "        if text[i] == \"A\":\n",
    "            ##text.replace(text[i], 'T')\n",
    "            reverseStack.append('T')\n",
    "        elif text[i] == 'T':\n",
    "            #text.replace(text[i], 'A')\n",
    "            reverseStack.append('A')\n",
    "\n",
    "        elif text[i] == 'C':\n",
    "            reverseStack.append('G')\n",
    "        elif text[i] == 'G':\n",
    "            reverseStack.append('C')\n",
    "\n",
    "    reverseComplement = []\n",
    "    for i in range(len(reverseStack)):\n",
    "        reverseComplement.append(reverseStack.pop())\n",
    "    return (\"\".join(str(x) for x in reverseComplement))\n",
    "\n",
    "print(shared(3,p,q))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "    CODE CHALLENGE: Solve the 2-Break Sorting Problem.     \n",
    "    2-Break Sorting Problem: Find a shortest transformation \n",
    "    of one genome into another via 2-breaks.\n",
    "    Input: Two genomes with circular chromosomes on the same \n",
    "    set of synteny blocks.\n",
    "    Output: The collection of genomes resulting from applying \n",
    "    a shortest sequence of 2-breaks transforming one genome into the other.\n",
    "    '''\n",
    "\n",
    "import random\n",
    "def chromosometocycle(chromosome):\n",
    "    nodes = []\n",
    "    for i in chromosome:\n",
    "        if i > 0:\n",
    "            nodes.append(2*i - 1)\n",
    "            nodes.append(2*i)\n",
    "        else:\n",
    "            nodes.append(-2*i)\n",
    "            nodes.append(-2*i - 1)\n",
    "    return nodes\n",
    "\n",
    "def cycletochromosome(nodes):\n",
    "    chromosome = []\n",
    "    for j in range(1,int(len(nodes)/2)+1):\n",
    "        if nodes[2*j - 2] < nodes[2*j - 1]:\n",
    "            chromosome.append(nodes[2*j - 1]/2)\n",
    "        else:\n",
    "            chromosome.append(-nodes[2*j - 2]/2)\n",
    "    return chromosome\n",
    "\n",
    "def colorededges(p):\n",
    "    q = []\n",
    "    for i in p:\n",
    "        if isinstance(i,float):\n",
    "            q.append(int(i)) \n",
    "    if len(q) != 0:\n",
    "        p = q\n",
    "    edges = []\n",
    "    if type(p[0]) == int: \n",
    "        if type(p[0]) == float:\n",
    "            chromosome = [int(z) for z in p]\n",
    "        else:\n",
    "            chromosome = p\n",
    "        chromosome.append(chromosome[0])\n",
    "        nodes = chromosometocycle(chromosome)\n",
    "        for j in range(1,len(chromosome)):\n",
    "            edges.append([nodes[2*j-1],nodes[2*j]])\n",
    "        chromosome.pop()\n",
    "        return edges\n",
    "    else:\n",
    "        for chromosome in p:\n",
    "            chromosome = [int(z) for z in chromosome]\n",
    "            temp = chromosome[0]\n",
    "            chromosome.append(temp)\n",
    "            nodes = chromosometocycle(chromosome)\n",
    "            for j in range(1,len(chromosome)):\n",
    "                edges.append([nodes[2*j-1],nodes[2*j]])\n",
    "            chromosome.pop()\n",
    "        return edges\n",
    "\n",
    "def get_blacked_edges(col_edges):\n",
    "    blacked_edges = []\n",
    "    for node in col_edges:\n",
    "        temp = []\n",
    "        for num in node:\n",
    "            if num%2 == 0:\n",
    "                temp.append(num/2)\n",
    "            else:\n",
    "                temp.append(-(num+1)/2)\n",
    "        blacked_edges.append(temp)\n",
    "    return blacked_edges\n",
    "\n",
    "def graphtogenome(col_edges):\n",
    "    blacked_edges = get_blacked_edges(col_edges)\n",
    "    chromosome_list = []\n",
    "    while blacked_edges != []:\n",
    "        start,end = blacked_edges[0]\n",
    "        blacked_edges.remove(blacked_edges[0])\n",
    "        loop_end = -1*start\n",
    "        chromosome = [start]\n",
    "        while end != loop_end:\n",
    "            for i in range(len(blacked_edges)):\n",
    "                if blacked_edges[i][0] == -1*end:\n",
    "                    start,end = blacked_edges[i]\n",
    "                    blacked_edges.remove(blacked_edges[i])\n",
    "                    break\n",
    "                elif blacked_edges[i][1] == -1*end:\n",
    "                    blacked_edges[i] = blacked_edges[i][::-1] \n",
    "                    start,end = blacked_edges[i]\n",
    "                    blacked_edges.remove(blacked_edges[i])\n",
    "                    break\n",
    "\n",
    "            chromosome.append(start)\n",
    "        chromosome_list.append(chromosome)\n",
    "    return chromosome_list\n",
    "\n",
    "def two_breakongenomegraph(col_edges,i,j,k,l):\n",
    "    if [i,j] in col_edges:\n",
    "        col_edges.remove([i,j])\n",
    "    else:\n",
    "        col_edges.remove([j,i])\n",
    "    if [k,l] in col_edges:\n",
    "        col_edges.remove([k,l])\n",
    "    else:\n",
    "        col_edges.remove([l,k])\n",
    "    col_edges.append([i,k])\n",
    "    col_edges.append([j,l])\n",
    "    return col_edges\n",
    "\n",
    "def two_breakongenome(p,i,j,k,l):\n",
    "    colored_edges = colorededges(p)\n",
    "    genomegraph = two_breakongenomegraph(colored_edges,i,j,k,l)\n",
    "    ans = graphtogenome(genomegraph)\n",
    "    return ans\n",
    "\n",
    "def shortestrearrangement(p,q):\n",
    "    #print p\n",
    "    print('('+' '.join('+'+str(i) if i>0 else str(i) for i in p)+')')\n",
    "    red_edges = colorededges(p)\n",
    "    #print(\"red_edges\", red_edges)\n",
    "    blue_edges = colorededges(q)\n",
    "    #print(\"blue_edges\",blue_edges)\n",
    "    breakpointgraph = red_edges + blue_edges\n",
    "    while True:\n",
    "        red_edges = colorededges(p)\n",
    "        for i in range(len(red_edges)):\n",
    "            if red_edges[i][::-1] in blue_edges:\n",
    "                red_edges[i] = red_edges[i][::-1]\n",
    "        if sorted(red_edges) == sorted(blue_edges):\n",
    "            break\n",
    "        temp_list = []\n",
    "        for blue_edge in blue_edges:\n",
    "            if blue_edge not in red_edges:\n",
    "                temp_list.append(blue_edge)\n",
    "        node = random.choice(temp_list)\n",
    "        #node = temp_list[0]\n",
    "        #from_edge,to_edge = 0,0\n",
    "        for red_edge in red_edges:\n",
    "            if node[0] in red_edge:\n",
    "                to_edge = red_edge\n",
    "            elif node[1] in red_edge:\n",
    "                from_edge = red_edge\n",
    "        if (to_edge[1]==node[0]) & (from_edge[0]==node[1]):\n",
    "            red_edges.remove(from_edge)\n",
    "            red_edges.remove(to_edge)\n",
    "            red_edges.append([to_edge[1],from_edge[0]])\n",
    "            red_edges.append([to_edge[0],from_edge[1]])\n",
    "        elif (to_edge[0]==node[0]) & (from_edge[0]==node[1]):\n",
    "            red_edges.remove(from_edge)\n",
    "            red_edges.remove(to_edge)\n",
    "            to_edge = to_edge[::-1]\n",
    "            red_edges.append([to_edge[1],from_edge[0]])\n",
    "            red_edges.append([to_edge[0],from_edge[1]])\n",
    "        elif (to_edge[1]==node[0]) & (from_edge[1]==node[1]):\n",
    "            red_edges.remove(from_edge)\n",
    "            red_edges.remove(to_edge)\n",
    "            from_edge = from_edge[::-1]\n",
    "            red_edges.append([to_edge[1],from_edge[0]])\n",
    "            red_edges.append([to_edge[0],from_edge[1]])\n",
    "        elif (to_edge[0]==node[0]) & (from_edge[1]==node[1]):\n",
    "            red_edges.remove(from_edge)\n",
    "            red_edges.remove(to_edge)\n",
    "            from_edge = from_edge[::-1]\n",
    "            to_edge = to_edge[::-1]\n",
    "            red_edges.append([to_edge[1],from_edge[0]])\n",
    "            red_edges.append([to_edge[0],from_edge[1]])\n",
    "        breakpointgraph = red_edges + blue_edges\n",
    "        p = two_breakongenome(p,to_edge[0],to_edge[1],from_edge[1],from_edge[0])\n",
    "        str_p = ''\n",
    "        for res in p:\n",
    "            str_p += '('+' '.join('+'+str(i) if i>0 else str(i) for i in res)+')'\n",
    "        print(str_p)\n",
    "        if len(p) == 1:\n",
    "            p = p[0]\n",
    "\n",
    "\n",
    "        #print p\n",
    "    return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "P = '(+11 -9 -6 -13 -1 -10 -2 +14 +12 +8 +5 +3 +4 -7)'\n",
    "Q = '(+8 +13 -12 -3 +7 -14 +9 -1 +4 +5 +6 -11 -2 -10)'\n",
    "P = list(map(int,P.strip().lstrip('(').rstrip(')').split(' ')))\n",
    "Q = list(map(int,Q.strip().lstrip('(').rstrip(')').split(' ')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run this algorithm till last line = Q\n",
    "shortestrearrangement(P,Q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
