{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Implement the FarthestFirstTraversal clustering heuristic.\n",
    "     Input: Integers k and m followed by a set of points Data in m-dimensional space.\n",
    "     Output: A set Centers consisting of k points (centers) resulting from applying FarthestFirstTraversal(Data, k),\n",
    "     where the first point from Data is chosen as the first center to initialize the algorithm.\n",
    "Sample Input:\n",
    "3 2\n",
    "0.0 0.0\n",
    "5.0 5.0\n",
    "0.0 5.0\n",
    "1.0 1.0\n",
    "2.0 2.0\n",
    "3.0 3.0\n",
    "1.0 2.0\n",
    "Sample Output:\n",
    "0.0 0.0\n",
    "5.0 5.0\n",
    "0.0 5.0\n",
    "Although the k-Center Clustering Problem is easy to state, it is NP-Hard. The Farthest First Traversal heuristic, whose \n",
    "pseudocode is shown below, selects centers from the points in Data (instead of from all possible points in m-dimensional \n",
    "space). It begins by selecting an arbitrary point in Data as the first center and iteratively adds a new center as the point \n",
    "in Data that is farthest from the centers chosen so far, with ties broken arbitrarily.\n",
    "FarthestFirstTraversal(Data, k) \n",
    "    Centers ← the set consisting of a single randomly chosen point from Data\n",
    "    while |Centers| < k \n",
    "        DataPoint ← the point in Data maximizing d(DataPoint, Centers) \n",
    "        add DataPoint to Centers \n",
    "    return Centers \n",
    "'''\n",
    "\n",
    "import sys, math\n",
    "\n",
    "class FarthestFirstTravesal:\n",
    "\n",
    "    def __init__(self,data,k):\n",
    "        self.data = data\n",
    "        self.k = k\n",
    "\n",
    "    def euclideanDistance(self,v,m):\n",
    "        return math.sqrt(sum([(i-j)**2 for i,j in zip(v,m)]))\n",
    "\n",
    "    def distanceDataToCenters(self,point,centers):\n",
    "        distances = [self.euclideanDistance(point,center) for center in centers]\n",
    "        return min(distances)\n",
    "\n",
    "    def addNewCenter(self,centers):\n",
    "        #maxDistanceDatasToCenters\n",
    "        newCenter = None\n",
    "        maxLength = float('-inf')\n",
    "        for point in self.data:\n",
    "            segmentLen = self.distanceDataToCenters(point,centers)\n",
    "            if segmentLen > maxLength:\n",
    "                maxLength = segmentLen\n",
    "                newCenter = point\n",
    "        return newCenter\n",
    "\n",
    "    def fartherFirstTravesal(self):\n",
    "        centers = [self.data[0]]\n",
    "        while len(centers) < self.k:\n",
    "            centers.append(self.addNewCenter(centers))\n",
    "        return centers\n",
    "\n",
    "k = 4\n",
    "m = 5\n",
    "data = [(0.0,0.0),(5.0,5.0),(0.0,5.0),(1.0,1.0),(2.0,2.0),(3.0,3.0),(1.0,2.0)]\n",
    "fftObj = FarthestFirstTravesal(data,k)\n",
    "fftObj.fartherFirstTravesal()\n",
    "centers = fftObj.fartherFirstTravesal()\n",
    "for center in centers:\n",
    "    print(' '.join(map(str,center)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Squared Error Distortion Problem: Compute the squared error distortion of a set of data points with respect to a set of \n",
    "centers. \n",
    "  Input: A set of points Data and a set of centers Centers.\n",
    "  Output: The squared error distortion Distortion(Data, Centers). \n",
    "Code Challenge: Solve the Squared Error Distortion Problem.\n",
    "     Input: Integers k and m, followed by a set of centers Centers and a set of points Data.\n",
    "     Output: The squared error distortion Distortion(Data, Centers).\n",
    "Sample Input:\n",
    "2 2\n",
    "2.31 4.55\n",
    "5.96 9.08\n",
    "--------\n",
    "3.42 6.03\n",
    "6.23 8.25\n",
    "4.76 1.64\n",
    "4.47 4.33\n",
    "3.95 7.61\n",
    "8.93 2.97\n",
    "9.74 4.03\n",
    "1.73 1.28\n",
    "9.72 5.01\n",
    "7.27 3.77\n",
    "Sample Output:\n",
    "18.246\n",
    "To address limitations of MaxDistance, we will introduce a new scoring function. Given a set Data of n data points and a set \n",
    "Centers of k centers, the squared error distortion of Data and Centers, denoted Distortion(Data, Centers), is defined as the \n",
    "mean squared distance from each data point to its nearest center, \n",
    "Distortion(Data,Centers) = (1/n) ∑all points DataPoint in Datad(DataPoint, Centers)2 .\n",
    "Note that whereas MaxDistance(Data, Centers) only accounts for the length of the single red segment in the figure below, the \n",
    "squared error distortion accounts for the length of all segments.\n",
    "'''\n",
    "import sys,math\n",
    "\n",
    "class SquaredErrorDistortion():\n",
    "    def __init__(self,data,centers):\n",
    "        self.data = data\n",
    "        self.centers = centers\n",
    "\n",
    "    def euclideanDistance(self,v,m):\n",
    "        return math.sqrt(sum([(i-j)**2 for i,j in zip(v,m)]))\n",
    "\n",
    "    def distanceDataToCenters(self,point,centers):\n",
    "            distances = [self.euclideanDistance(point,center) for center in centers ]\n",
    "            return min(distances)\n",
    "\n",
    "    def SquaredErrorDistortion(self):\n",
    "        distortions = []\n",
    "        for data in self.data:\n",
    "            distortions.append(self.distanceDataToCenters(data,self.centers) ** 2)\n",
    "        print(\"MaxDistance:\",math.sqrt(max(distortions)))\n",
    "        return sum(distortions)/len(self.data)\n",
    "    \n",
    "k = 2\n",
    "m = 2\n",
    "centers = [(2.31,4.55),(5.96,9.08)]\n",
    "data = [(3.42,6.03),(6.23,8.25),(4.76,1.64),(4.47,4.33),(3.95,7.61),(8.93,2.97),(9.74,4.03),(1.73,1.28),(9.72,5.01),(7.27,3.77)]\n",
    "\n",
    "sedObj = SquaredErrorDistortion(data,centers)\n",
    "print(sedObj.SquaredErrorDistortion())\n",
    "\n",
    "'''\n",
    "trial.txt ---\n",
    "2.31 4.55\n",
    "5.96 9.08\n",
    "--------\n",
    "3.42 6.03\n",
    "6.23 8.25\n",
    "4.76 1.64\n",
    "4.47 4.33\n",
    "3.95 7.61\n",
    "8.93 2.97\n",
    "9.74 4.03\n",
    "1.73 1.28\n",
    "9.72 5.01\n",
    "7.27 3.77\n",
    "'''\n",
    "\n",
    "with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/trial.txt') as f:\n",
    "    lines = f.read().splitlines()\n",
    "centers = []\n",
    "for i in range(0,k):\n",
    "    centers.append(tuple(map(float,lines[i].split())))\n",
    "data = []\n",
    "for line in lines[k+1:]:\n",
    "    data.append(tuple(map(float,line.split(' '))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Implement the Lloyd algorithm for k-means clustering.\n",
    "     Input: Integers k and m followed by a set of points Data in m-dimensional space.\n",
    "     Output: A set Centers consisting of k points (centers) resulting from applying the Lloyd algorithm to Data and Centers,\n",
    "     where the first k points from Data are selected as the first k centers.\n",
    "Sample Input:\n",
    "2 2\n",
    "1.3 1.1\n",
    "1.3 0.2\n",
    "0.6 2.8\n",
    "3.0 3.2\n",
    "1.2 0.7\n",
    "1.4 1.6\n",
    "1.2 1.0\n",
    "1.2 1.1\n",
    "0.6 1.5\n",
    "1.8 2.6\n",
    "1.2 1.3\n",
    "1.2 1.0\n",
    "0.0 1.9\n",
    "Sample Output:\n",
    "1.800 2.867\n",
    "1.060 1.140\n",
    "The Lloyd algorithm is one of the most popular clustering heuristics for the k-Means Clustering Problem. It first chooses k \n",
    "arbitrary points Centers from Data as centers and then iteratively performs the following two steps (see figure below):\n",
    "Centers to Clusters: After centers have been selected, assign each data point to the cluster corresponding to its nearest \n",
    "center; ties are broken arbitrarily. \n",
    "Clusters to Centers: After data points have been assigned to clusters, assign each cluster’s center of gravity to be the \n",
    "cluster’s new center.\n",
    "'''\n",
    "import sys, math\n",
    "from decimal import *\n",
    "getcontext().prec = 3\n",
    "\n",
    "class Lloyd():\n",
    "    def __init__(self,data,k):\n",
    "        self.data = data\n",
    "        self.k = k\n",
    "\n",
    "    def euclideanDistance(self,v,m):\n",
    "        return math.sqrt(sum([(i-j)**2 for i,j in zip(v,m)]))\n",
    "\n",
    "    def centersToCluster(self,point,centers):\n",
    "        distances = [self.euclideanDistance(point,center) for center in centers ]\n",
    "        return distances.index(min(distances))\n",
    "\n",
    "    def centerOfGravity(self,data):\n",
    "        numOfPoints = len(data)\n",
    "        center = map(sum,zip(*data))\n",
    "        return [c/numOfPoints for c in center]\n",
    "\n",
    "    def checkConverge(self,cen1,cen2):\n",
    "        for i in range(len(cen1)):\n",
    "            for c in range(len(cen1[i])):\n",
    "                if Decimal(cen1[i][c]) != Decimal(cen2[i][c]):\n",
    "                    return False\n",
    "        return True\n",
    "\n",
    "    def kMeansCluster(self):\n",
    "        centers = self.data[:self.k]\n",
    "\n",
    "        while True:\n",
    "            toCenters = []\n",
    "            for data in self.data:\n",
    "                toCenters.append(self.centersToCluster(data,centers))\n",
    "            newCenters = []\n",
    "            for cluNum in range(self.k):\n",
    "                newData = []\n",
    "                for i,c in enumerate(toCenters):\n",
    "                    if c == cluNum:\n",
    "                        newData.append(self.data[i])\n",
    "                newCenters.append(self.centerOfGravity(newData))\n",
    "            if self.checkConverge(centers,newCenters):\n",
    "                break\n",
    "            else:\n",
    "                centers = newCenters\n",
    "        return centers\n",
    "\n",
    "k = 2\n",
    "m = 2\n",
    "data = [(1.3,1.1),(1.3,0.2),(0.6,2.8),(3.0,3.2),(1.2,0.7),(1.4,1.6),(1.2,1.0),(1.2,1.1),(0.6,1.5),(1.8,2.6),(1.2,1.3),(1.2,1.0),(0.0,1.9)]\n",
    "\n",
    "lloyedObj = Lloyd(data,k)\n",
    "centers = lloyedObj.kMeansCluster()\n",
    "for center in centers:\n",
    "    print(' '.join(map(str,center)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Implement HierarchicalClustering.\n",
    "     Input: An integer n, followed by an n x n distance matrix.\n",
    "     Output: The result of applying HierarchicalClustering to this distance matrix (using Davg), with each newly created cluster listed\n",
    "     on each line.\n",
    "Sample Input:\n",
    "7\n",
    "0.00 0.74 0.85 0.54 0.83 0.92 0.89\n",
    "0.74 0.00 1.59 1.35 1.20 1.48 1.55\n",
    "0.85 1.59 0.00 0.63 1.13 0.69 0.73\n",
    "0.54 1.35 0.63 0.00 0.66 0.43 0.88\n",
    "0.83 1.20 1.13 0.66 0.00 0.72 0.55\n",
    "0.92 1.48 0.69 0.43 0.72 0.00 0.80\n",
    "0.89 1.55 0.73 0.88 0.55 0.80 0.00\n",
    "Sample Output:\n",
    "4 6\n",
    "5 7\n",
    "3 4 6\n",
    "1 2\n",
    "5 7 3 4 6\n",
    "1 2 5 7 3 4 6\n",
    "HierarchicalClustering, whose pseudocode is shown below, progressively generates n different partitions of the underlying data into clusters, \n",
    "all represented by a tree in which each node is labeled by a cluster of genes. The first partition has n single-element clusters represented \n",
    "by the leaves of the tree, with each element forming its own cluster. The second partition merges the two “closest” clusters into a single \n",
    "cluster consisting of two elements. In general, the i-th partition merges the two closest clusters from the (i - 1)-th partition and has \n",
    "n - i + 1 clusters. We hope this algorithm looks familiar — it is UPGMA (from the chapter on evolutionary tree constuction) in disguise.\n",
    "HierarchicalClustering(D, n)\n",
    "    Clusters ← n single-element clusters labeled 1, ... , n \n",
    "       construct a graph T with n isolated nodes labeled by single elements 1, ... , n \n",
    "    while there is more than one cluster \n",
    "        find the two closest clusters Ci and Cj\n",
    "        merge Ci and Cj into a new cluster Cnew with |Ci| + |Cj| elements\n",
    "        add a new node labeled by cluster Cnew to T\n",
    "        connect node Cnew to Ci and Cj by directed edges\n",
    "        remove the rows and columns of D corresponding to Ci and Cj\n",
    "        remove Ci and Cj from Clusters\n",
    "        add a row/column to D for Cnew by computing D(Cnew, C) for each C in Clusters \n",
    "        add Cnew to Clusters \n",
    "    assign root in T as a node with no incoming edges\n",
    "    return T\n",
    "The distance function that we encountered with UPGMA uses the average distance between elements in two clusters,\n",
    "Davg(C1,C2)=∑all points i in cluster C1 ∑all points j in cluster C2Di,j / (|C1|⋅|C2|)\n",
    "'''\n",
    "\n",
    "\n",
    "import sys,math\n",
    "\n",
    "class DepthFirstPaths():\n",
    "    def __init__(self,tree):\n",
    "        self.tree = tree\n",
    "        self.edgeTo = dict((key,None) for key in self.tree.keys())\n",
    "        self.marked = dict((key,False) for key in self.tree.keys())\n",
    "        self.count = 0\n",
    "\n",
    "    def dfs(self,v):\n",
    "        self.count += 1\n",
    "        self.marked[v] = True\n",
    "        for w in self.tree[v]:\n",
    "            if not self.marked[w]:\n",
    "                self.edgeTo[w] = v\n",
    "                self.dfs(w)\n",
    "\n",
    "    def countLeaves(self,source):\n",
    "        if len(self.tree[source]) == 0:\n",
    "            return 1\n",
    "        self.dfs(source)\n",
    "        return (self.count+1)/2\n",
    "\n",
    "\n",
    "def findClosestClusters(D):\n",
    "    minValue = float('inf')\n",
    "    for i in range(len(D)):\n",
    "        for j in range(i+1,len(D)):\n",
    "            if D[i][j] < minValue:\n",
    "                minValue = D[i][j]\n",
    "                closestClusters = (i,j)\n",
    "    return closestClusters\n",
    "\n",
    "def HierarchicalClustering(D):\n",
    "    n = len(D)\n",
    "    clusters = { node:[node] for node in range(n) }\n",
    "    nodes = list(range(n))\n",
    "    T = { node:[] for node in range(n) }\n",
    "\n",
    "    while len(clusters) > 1:\n",
    "        (ci,cj) = findClosestClusters(D)\n",
    "        nodei = nodes[ci]\n",
    "        nodej = nodes[cj]\n",
    "        newCluster = clusters[nodei] + clusters[nodej]\n",
    "        newCluster = [nc+1 for nc in newCluster]\n",
    "        print(' '.join(map(str,newCluster)))\n",
    "\n",
    "        newNode = n\n",
    "        n += 1\n",
    "\n",
    "        newRow = []\n",
    "        for m in range(len(D)):\n",
    "            if m!= ci and m!=cj:\n",
    "                cmi = D[m][ci]\n",
    "                cmj = D[m][cj]\n",
    "                leaveNumi = len(clusters[nodei])\n",
    "                leaveNumj = len(clusters[nodej])\n",
    "                newRow.append((cmi*leaveNumi+cmj*leaveNumj) / (leaveNumi+leaveNumj))\n",
    "        newRow.append(0)\n",
    "\n",
    "        Dnew = []\n",
    "        for i in range(len(D)):\n",
    "            if i!=ci and i!=cj:\n",
    "                Dnew.append([D[i][j] for j in range(len(D[i])) if j != ci and j != cj])\n",
    "        Dnew.append(newRow)\n",
    "        for i in range(len(Dnew)-1):\n",
    "            Dnew[i].append(Dnew[-1][i])\n",
    "        D = Dnew\n",
    "\n",
    "        T[newNode] = [nodes[ci], nodes[cj]]\n",
    "\n",
    "        nodes.append(newNode)\n",
    "        nodes = [nodes[i] for i in range(len(nodes)) if i !=ci and i != cj]\n",
    "        clusters[newNode] = clusters[nodei] + clusters[nodej]\n",
    "        clusters.pop(nodei)\n",
    "        clusters.pop(nodej)\n",
    "\n",
    "    root = list(clusters.keys())[0]\n",
    "    T[root] = [clusters[root]]\n",
    "    return T\n",
    "\n",
    "n = 7\n",
    "D = [[0.00, 0.74, 0.85, 0.54, 0.83, 0.92, 0.89],[0.74, 0.00, 1.59, 1.35, 1.20, 1.48, 1.55],[0.85, 1.59, 0.00, 0.63, 1.13, 0.69, 0.73],[0.54, 1.35, 0.63, 0.00, 0.66, 0.43, 0.88],[0.83, 1.20, 1.13, 0.66, 0.00, 0.72, 0.55],[0.92, 1.48, 0.69, 0.43, 0.72, 0.00, 0.80],[0.89, 1.55, 0.73, 0.88, 0.55, 0.80, 0.00]]\n",
    "HierarchicalClustering(D)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Question 3: Imagine an isolated population in which each adult female has a distinct mitochondrial genome. In each generation, every adult female has two children (with a 50% chance of being male or female). After each generation, we will assume that all the adults die out, and all the children become adults.\n",
    "\n",
    "# After four generations, what is the expected percentage of mitochondria from the original population of females that will still exist in the population’s children? Express your answer as a decimal between 0 and 1, rounded to three decimal places.\n",
    "\n",
    "# Note: after one generation, there will be 100% of the mitochondria from the original population present in the population’s children since male children receive the mitochondria from their mothers (but do not pass them on to their own children.)\n",
    "\n",
    "# From the text: If we assume a simple population model in which every woman in this population has exactly two children, then about a quarter of the mitochondrial genomes will be lost every generation.\n",
    "\n",
    "# When every generation dies 1/4 of the ancestral genome is lost and the new population only has 3/4 of the original genome i.e.\n",
    "\n",
    "# When the first generation dies only 3/4 *1 = 0.75 of original genome retained\n",
    "\n",
    "# When the second generation dies 3/4*0.75 = 0.5625 of original genome retained\n",
    "\n",
    "# When the Third generation dies 3/4 *0.5625 = 0.421875 of original genome retained\n",
    "\n",
    "# This means that the fourth generation will receive approximately 0.422 of the original genome from their parents i.e (the third generation)\n",
    "\n",
    "# Question 5: How many possible columns of length 5 are compatible with the column (0, 1, 1, 0, 0)?\n",
    "\n",
    "# From the text: Using the notation Oi to denote the collection of rows possessing a 1 in their i-th element, we conclude that one of three possibilities must be true: Oi ⊆ Oj , Oj ⊆ Oi , or Oi and Oj are disjoint (the first two cases include the possibility that Oi = Oj). If columns i and j satisfy this condition, then we call\n",
    "\n",
    "# Taking an example of the first and second column i.e. (column i and j) from figure A in the text:\n",
    "\n",
    "# i = (1,0,0,0,0,0,0)\n",
    "\n",
    "# j = (1,1,0,0,0,0)\n",
    "\n",
    "# Oi and Oj will be sets representing where a \"1\" occurs in the binary vectors above i.e. column i contains 1 in the first position and column j contains a 1 in the first and second positions thus\n",
    "\n",
    "# Oi = {1}\n",
    "\n",
    "# Oj = {1,2}\n",
    "\n",
    "# Given the above conditions the proof of compatibility would proceed as follows:\n",
    "\n",
    "# Oi ⊆ Oj . This is true since Oi is clealy a subset of Oj\n",
    "\n",
    "# Oj ⊆ Oi. This is false since Oj contains a 2 which is not in Oi\n",
    "\n",
    "# Oi and Oj are disjoint. This is false since Oi and Oj both share the element 1\n",
    "\n",
    "# However since the first condition evaluated to be true, Then we conclude that columns i and j are compatible.\n",
    "\n",
    "# Now to solve the test question I painstakingly(well I wrote a simple script) to generate all possible binary vectors of length 5 (they are 32 in total) and i evaluated if their subsequent sets i.e. Oj satisfied any of the above conditions in my case the answer was 18.\n",
    "\n",
    "# Note: If the question asks for all compatible columns with (0,0,0,0,0) then the answer is 32 since it is disjoint to all other columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
