{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Given a string Text, its k-mer composition Compositionk(Text) is the collection of all k-mer substrings of Text (including repeated k-mers)\n",
    "# Generate the k-mer composition of a string.\n",
    "def composition(text, k):\n",
    "    patterns = []\n",
    "    for i in range(len(text)-k + 1):\n",
    "        pat = text[i:i+k] \n",
    "        patterns.append(pat)\n",
    "    return patterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# consecutive 3-mers in TAATGCCATGGGATGTT are linked together to form this string's genome path.\n",
    "# Reconstruct a string from its genome path\n",
    "def string_reconstruct(dna_list):\n",
    "    dna_first = dna_list[0]\n",
    "    for dna in dna_list[1:]:\n",
    "        dna_first = dna_first + dna[-1]\n",
    "    return dna_first\n",
    "\n",
    "with open(\"/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/dna_reconstruct.txt\") as f:\n",
    "    dna_list  = f.read().splitlines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To generalize the construction of the above graph to an arbitrary collection of k-mers Patterns, we form a node for each k-mer in Patterns and connect k-mers Pattern and Pattern' by a directed edge if Suffix(Pattern) is equal to Prefix(Pattern'). The resulting graph is called the overlap graph on these k-mers, denoted Overlap(Patterns)\n",
    "# The overlap graph Overlap(Patterns), in the form of an adjacency list\n",
    "\n",
    "def suffix(t):\n",
    "    return t[1:]\n",
    "\n",
    "def prefix(t):\n",
    "    return t[:-1]\n",
    "\n",
    "def overlap_graph(reads):\n",
    "    overlaps = []\n",
    "    for j in sorted(reads):\n",
    "        for i in reads:\n",
    "            if suffix(i) == prefix(j):\n",
    "                overlaps.append((i, j))\n",
    "    overlaps_df = pd.DataFrame(overlaps)\n",
    "    overlaps_df.columns = [\"i\",\"j\"]\n",
    "    overlaps_df = overlaps_df.groupby('i').agg({'j': ', '.join}).reset_index()\n",
    "    overlaps = dict(zip(overlaps_df.i, overlaps_df.j))\n",
    "    return overlaps\n",
    "\n",
    "for key, value in overlap_graph(dna_list).items():\n",
    "    print(key + \" -> \" + value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Solve the De Bruijn Graph from a String Problem.\n",
    "# The de Bruijn graph DeBruijnk(Text) is formed by gluing identically labeled nodes in PathGraphk(Text).\n",
    "\n",
    "def deBruijn(dna, k):\n",
    "    reads = []\n",
    "    for i in range(len(dna)-k+1):\n",
    "        pat = dna[i:i+k]\n",
    "        reads.append(pat)\n",
    "    de_bruijn_dict = dict()\n",
    "    for kmer in sorted(reads):\n",
    "        if kmer[:-1] in de_bruijn_dict:\n",
    "            de_bruijn_dict[kmer[:-1]].add(kmer[1:])\n",
    "        else:\n",
    "            de_bruijn_dict[kmer[:-1]] = {kmer[1:]}\n",
    "    de_buijn = [' -> '.join([item[0], ','.join(item[1])]) for item in sorted(de_bruijn_dict.items())]\n",
    "    return de_buijn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# For every k-mer in Patterns, we connect its prefix node to its suffix node by a directed edge in order to produce DeBruijn(Patterns).\n",
    "# Construct the de Bruijn graph from a set of k-mers.\n",
    "\n",
    "def build_deGruijnGraphFromKmers(kmers):\n",
    "    kmers = sorted(kmers)\n",
    "    ''' we need to merge the prefix and suffix set in case there are single isolated edges! '''\n",
    "    k_1mers = [k[:-1] for k in kmers] + [k[1:] for k in kmers]\n",
    "    k_1mers = sorted(set(k_1mers))\n",
    "\n",
    "    nodes = {}\n",
    "    for i in range(len(k_1mers)):\n",
    "        nodes[i] = k_1mers[i]\n",
    "    invnodes = {v:k for k,v in nodes.items()}\n",
    "\n",
    "    edges = {}\n",
    "    for pat in kmers:\n",
    "        prefixpat = pat[:-1]\n",
    "        suffixpat = pat[1:]\n",
    "        if invnodes[prefixpat] in edges:\n",
    "            edges[invnodes[prefixpat]].append(invnodes[suffixpat])\n",
    "        else:\n",
    "            edges[invnodes[prefixpat]] = [invnodes[suffixpat]]\n",
    "\n",
    "    temp = []\n",
    "    for key, vals in edges.items():\n",
    "        print(nodes[key],'->',','.join([nodes[val] for val in vals]))\n",
    "        temp.append(nodes[key] +' -> '+','.join([nodes[val] for val in vals]))\n",
    "    return temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# EulerianCycle(Graph)\n",
    "#     form a cycle Cycle by randomly walking in Graph (don't visit the same edge twice!)\n",
    "#     while there are unexplored edges in Graph\n",
    "#         select a node newStart in Cycle with still unexplored edges\n",
    "#         form Cycle’ by traversing Cycle (starting at newStart) and then randomly walking \n",
    "#         Cycle ← Cycle’\n",
    "#     return Cycle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Solve the Eulerian Cycle Problem.\n",
    "with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/dataset_203_2-2.txt') as f:\n",
    "    lines = f.read().splitlines()\n",
    "adjlist = {}\n",
    "for line in lines:\n",
    "    node = int(line.split(' -> ')[0])\n",
    "    edge = list(map(int,line.split(' -> ')[1].split(',')))\n",
    "    adjlist[node] = edge\n",
    "    \n",
    "def eulerian_cycle(adjlist):\n",
    "    nodes = list(adjlist.keys())\n",
    "    eulerianCycle = []\n",
    "\n",
    "    marked = {} # we need to mark the EDGE\n",
    "    edges = 0\n",
    "    for key,val in adjlist.items():\n",
    "        marked[key] = [False] * len(val)\n",
    "        edges += len(val)\n",
    "    # choose any vertex v and push it onto stack\n",
    "    v = random.choice(nodes)\n",
    "    # greedily add to cycle, depth-first search style\n",
    "    stack = []\n",
    "    stack.append(v)\n",
    "    # while the stack is not empty\n",
    "    while len(stack):\n",
    "        u = stack[-1]\n",
    "        unmarkedEdges = [i for i,mark in enumerate(marked[u]) if mark == False]\n",
    "        if len(unmarkedEdges):\n",
    "            w = adjlist[u][unmarkedEdges[0]]\n",
    "            marked[u][unmarkedEdges[0]] = True\n",
    "            stack.append(w)\n",
    "        else:\n",
    "            del stack[-1]\n",
    "            eulerianCycle.append(u)\n",
    "    return eulerianCycle[::-1]\n",
    "\n",
    "cycle = eulerian_cycle(adjlist)\n",
    "printcycle = ''.join([str(c) + '->' for c in cycle])\n",
    "print(printcycle[:-2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#find eulerian path\n",
    "from functools import reduce\n",
    "def eulerian_path(edge_dict):\n",
    "    '''Returns an Eulerian path from the given edges.'''\n",
    "    # Determine the unbalanced edges.\n",
    "    out_values = reduce(lambda a,b: a+b, edge_dict.values())\n",
    "    #print(out_values)\n",
    "    for node in set(out_values+list(edge_dict.keys())):\n",
    "        out_value = out_values.count(node)\n",
    "        if node in edge_dict:\n",
    "            in_value = len(edge_dict[node])\n",
    "        else:\n",
    "            in_value = 0\n",
    "\n",
    "        if in_value < out_value:\n",
    "            unbalanced_from = node\n",
    "        elif out_value < in_value:\n",
    "            unbalanced_to = node\n",
    "\n",
    "    # Add an edge connecting the unbalanced edges.\n",
    "    if unbalanced_from in edge_dict:\n",
    "        edge_dict[unbalanced_from].append(unbalanced_to)\n",
    "    else:\n",
    "        edge_dict[unbalanced_from] = [unbalanced_to]\n",
    "\n",
    "    # Get the Eulerian Cycle from the edges, including the unbalanced edge.\n",
    "    cycle = eulerian_cycle(edge_dict)\n",
    "\n",
    "    # Find the location of the unbalanced edge in the eulerian cycle.\n",
    "    divide_point = list(filter(lambda i: cycle[i:i+2] == [unbalanced_from, unbalanced_to], range(len(cycle)-1)))[0]\n",
    "\n",
    "    # Remove the unbalanced edge, and shift appropriately, overlapping the head and tail.\n",
    "    return cycle[divide_point+1:]+cycle[1:divide_point+1]\n",
    "\n",
    "with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/dataset_203_6.txt') as f:\n",
    "    lines = f.read().splitlines()\n",
    "adjlist = {}\n",
    "for line in lines:\n",
    "    node = int(line.split(' -> ')[0])\n",
    "    edge = list(map(int,line.split(' -> ')[1].split(',')))\n",
    "    adjlist[node] = edge\n",
    "path = eulerian_path(adjlist)\n",
    "print('->'.join(map(str,path)))\n",
    "with open('ans.txt', 'w') as output_data:\n",
    "    output_data.write('->'.join(map(str,path)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#a method to assemble a genome, since the String Reconstruction Problem reduces to finding an Eulerian path in the de Bruijn graph generated from reads\n",
    "#Solve the String Reconstruction Problem\n",
    "def make_edges(seqs):\n",
    "    edges = []\n",
    "    for read in seqs:\n",
    "        edges.append(read[:-1]+ ' -> ' +read[1:])\n",
    "    return edges\n",
    "\n",
    "def reconstruct_string(edges):\n",
    "    string_dict = {line.strip().split(' -> ')[0]:line.strip().split(' -> ')[1] for line in edges}\n",
    "    head = list(filter(lambda x: x not in string_dict.values(), string_dict.keys()))[0]\n",
    "    tail = list(filter(lambda x: x not in string_dict.keys(), string_dict.values()))[0]\n",
    "    reconstructed_str = head[0]\n",
    "    current_str = head\n",
    "    while current_str != tail:\n",
    "        current_str = string_dict[current_str]\n",
    "        reconstructed_str += current_str[0]\n",
    "    reconstructed_str += tail[1:]\n",
    "\n",
    "    return reconstructed_str\n",
    "\n",
    "data = []\n",
    "with open(\"/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/dataset_203_7.txt\") as input_data:\n",
    "    for line in input_data:\n",
    "        data.append(line[:-1])\n",
    "k = int(data[0])\n",
    "reads = data[1:]\n",
    "edges = make_edges(reads)\n",
    "ans = reconstruct_string(edges)\n",
    "print(ans)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Solve the k-Universal Circular String Problem.\n",
    "def build_deGruijnGraphFromKmers(kmers):\n",
    "    kmers = sorted(kmers)\n",
    "    ''' we need to merge the prefix and suffix set in case there are single isolated edges! '''\n",
    "    k_1mers = [k[:-1] for k in kmers] + [k[1:] for k in kmers]\n",
    "    k_1mers = sorted(set(k_1mers))\n",
    "\n",
    "    nodes = {}\n",
    "    for i in range(len(k_1mers)):\n",
    "        nodes[i] = k_1mers[i]\n",
    "    invnodes = {v:k for k,v in nodes.items()}\n",
    "\n",
    "    edges = {}\n",
    "    for pat in kmers:\n",
    "        prefixpat = pat[:-1]\n",
    "        suffixpat = pat[1:]\n",
    "        if invnodes[prefixpat] in edges:\n",
    "            edges[invnodes[prefixpat]].append(invnodes[suffixpat])\n",
    "        else:\n",
    "            edges[invnodes[prefixpat]] = [invnodes[suffixpat]]\n",
    "\n",
    "    temp = []\n",
    "    for key, vals in edges.items():\n",
    "        print(nodes[key],'->',','.join([nodes[val] for val in vals]))\n",
    "        temp.append(nodes[key] +' -> '+','.join([nodes[val] for val in vals]))\n",
    "    return edges, nodes\n",
    "\n",
    "def universal_string(k):\n",
    "    kmers = [ ''.join(x) for x in product('01', repeat=k) ]\n",
    "    adjlist,nodes = build_deGruijnGraphFromKmers(kmers)\n",
    "    cycle = eulerian_cycle(adjlist)\n",
    "\n",
    "    length = 2**k\n",
    "    epath = [nodes[p] for p in cycle]\n",
    "    string = [epath[0]]\n",
    "    for pa in epath[1:]:\n",
    "        string.append(pa[-1])\n",
    "    uni_string  = ''.join(string)\n",
    "    uni_string = uni_string[:length]\n",
    "    print(uni_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def build_deGruijnGraphFromPairedKmers(pairedKmers):\n",
    "    # remove the white space in the end of some lines\n",
    "    pairedKmers = [pk.strip() for pk in pairedKmers]\n",
    "    # first need to sork the pairedKmers\n",
    "    kmers = [pk.replace('|','') for pk in pairedKmers]\n",
    "    sortedIndex = sorted(range(len(kmers)), key=lambda k: kmers[k])\n",
    "    # extract the prefix and suffix of both the first and second part of the kmers separated by '|'\n",
    "    sortedKmers = [pairedKmers[i] for i in sortedIndex]\n",
    "\n",
    "    k_1mers = []\n",
    "    for kmer in sortedKmers:\n",
    "        kmer = kmer.split('|')\n",
    "        k_1mers.append(kmer[0][:-1]+'|'+kmer[1][:-1])\n",
    "        k_1mers.append(kmer[0][1:]+'|'+kmer[1][1:])\n",
    "\n",
    "    k_1mers = sorted(set(k_1mers))\n",
    "\n",
    "    nodes = {}\n",
    "    for i in range(len(k_1mers)):\n",
    "        nodes[i] = k_1mers[i]\n",
    "    invnodes = {v:k for k,v in nodes.items()}\n",
    "\n",
    "    edges = {}\n",
    "    for pat in pairedKmers:\n",
    "        pat = pat.split('|')\n",
    "        node1 = pat[0][:-1]+'|'+pat[1][:-1]\n",
    "        node2 = pat[0][1:]+'|'+pat[1][1:]\n",
    "        if invnodes[node1] in edges:\n",
    "            edges[invnodes[node1]].append(invnodes[node2])\n",
    "        else:\n",
    "            edges[invnodes[node1]] = [invnodes[node2]]\n",
    "\n",
    "    adjlist = {}\n",
    "    temp = []\n",
    "    for key, vals in edges.items():\n",
    "        adjlist[key] = vals\n",
    "    return adjlist,nodes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def stringSpelledByGappedPatterns(gappedPatterns, k, d):\n",
    "    firstPatterns = [ pat.split('|')[0] for pat in gappedPatterns ]\n",
    "    secondPatterns = [ pat.split('|')[1].rstrip() for pat in gappedPatterns ]\n",
    "    preffixString = stringSpelledByPatterns(firstPatterns,k)\n",
    "    suffixString = stringSpelledByPatterns(secondPatterns,k)\n",
    "    if preffixString[k+d:] == preffixString[k+d:]:\n",
    "        return preffixString + suffixString[-k-d:]\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "def stringSpelledByPatterns(patterns,k):\n",
    "    string = [patterns[0]]\n",
    "    for pat in patterns[1:]:\n",
    "        string.append(pat[-1])\n",
    "    return ''.join(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Solve the String Reconstruction from Read-Pairs Problem.\n",
    "#reconstruct a string from read-pairs \n",
    "pairedKmers = ['GAGA|TTGA', 'TCGT|GATG', 'CGTG|ATGT', 'TGGT|TGAG', 'GTGA|TGTT', 'GTGG|GTGA', 'TGAG|GTTG', 'GGTC|GAGA', 'GTCG|AGAT']\n",
    "\n",
    "with open(filename) as f:\n",
    "    lines = f.read().splitlines()\n",
    "k = int(lines[0].split()[0])\n",
    "d = int(lines[0].split()[1])\n",
    "pairedKmers = lines[1:]\n",
    "\n",
    "\n",
    "adjlist,nodes = build_deGruijnGraphFromPairedKmers(pairedKmers)\n",
    "print(\"number of nodes:\",len(nodes))\n",
    "numedge = 0\n",
    "for v in adjlist.keys():\n",
    "    numedge += len(adjlist[v])\n",
    "print(\"number of outgoing edges:\",numedge)\n",
    "\n",
    "path = eulerian_path(adjlist)\n",
    "patterns = [nodes[p] for p in path]\n",
    "string = stringSpelledByGappedPatterns(patterns,k,d)\n",
    "print(string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate the contigs from a collection of reads (with imperfect coverage).\n",
    "# Solve the Contig Generation Problem\n",
    "with open(filename) as file:\n",
    "    kmers = [line.strip() for line in file.readlines()]\n",
    "    \n",
    "import collections\n",
    "def flatten(x):\n",
    "    result = []\n",
    "    for el in x:\n",
    "        if isinstance(x, collections.Iterable) and not isinstance(el, str):\n",
    "            result.extend(flatten(el))\n",
    "        else:\n",
    "            result.append(el)\n",
    "    return result\n",
    "\n",
    "edges = {}\n",
    "for kmer in kmers:\n",
    "    if kmer[:-1] in edges:\n",
    "        edges[kmer[:-1]].append(kmer[1:])\n",
    "    else:\n",
    "        edges[kmer[:-1]] = [kmer[1:]]\n",
    "\n",
    "balanced, unbalanced = [], []\n",
    "out_values = reduce(lambda a,b: a+b, edges.values())\n",
    "for node in set(out_values+list(edges.keys())):\n",
    "    out_value = out_values.count(node)\n",
    "    if node in edges:\n",
    "        in_value = len(edges[node])\n",
    "    else:\n",
    "        in_value = 0\n",
    "    if in_value == out_value == 1:\n",
    "        balanced.append(node)\n",
    "    else:\n",
    "        unbalanced.append(node)\n",
    "\n",
    "get_contigs = lambda s, c: flatten([c+e[-1] if e not in balanced else get_contigs(e,c+e[-1]) for e in edges[s]])\n",
    "contigs = sorted(flatten([get_contigs(start,start) for start in set(unbalanced) & set(edges.keys())]))\n",
    "\n",
    "print('\\n'.join(contigs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CODE CHALLENGE: Solve the String Reconstruction Problem.\n",
    "#      Input: An integer k followed by a list of k-mers Patterns.\n",
    "#      Output: A string Text with k-mer composition equal to Patterns. \n",
    "#      (If multiple answers exist, you may return any one.)\n",
    "\n",
    "def build_deGruijnGraphFromKmers(kmers):\n",
    "    kmers = sorted(kmers)\n",
    "    ''' we need to merge the prefix and suffix set in case there are single isolated edges! '''\n",
    "    k_1mers = [k[:-1] for k in kmers] + [k[1:] for k in kmers]\n",
    "    k_1mers = sorted(set(k_1mers))\n",
    "\n",
    "    nodes = {}\n",
    "    for i in range(len(k_1mers)):\n",
    "        nodes[i] = k_1mers[i]\n",
    "    invnodes = {v:k for k,v in nodes.items()}\n",
    "\n",
    "    edges = {}\n",
    "    for pat in kmers:\n",
    "        prefixpat = pat[:-1]\n",
    "        suffixpat = pat[1:]\n",
    "        if invnodes[prefixpat] in edges:\n",
    "            edges[invnodes[prefixpat]].append(invnodes[suffixpat])\n",
    "        else:\n",
    "            edges[invnodes[prefixpat]] = [invnodes[suffixpat]]\n",
    "\n",
    "    temp = []\n",
    "    for key, vals in edges.items():\n",
    "        print(nodes[key],'->',','.join([nodes[val] for val in vals]))\n",
    "        temp.append(nodes[key] +' -> '+','.join([nodes[val] for val in vals]))\n",
    "    return edges, nodes\n",
    "\n",
    "from functools import reduce\n",
    "def eulerian_path(edge_dict):\n",
    "    '''Returns an Eulerian path from the given edges.'''\n",
    "    # Determine the unbalanced edges.\n",
    "    out_values = reduce(lambda a,b: a+b, edge_dict.values())\n",
    "    #print(out_values)\n",
    "    for node in set(out_values+list(edge_dict.keys())):\n",
    "        out_value = out_values.count(node)\n",
    "        if node in edge_dict:\n",
    "            in_value = len(edge_dict[node])\n",
    "        else:\n",
    "            in_value = 0\n",
    "\n",
    "        if in_value < out_value:\n",
    "            unbalanced_from = node\n",
    "        elif out_value < in_value:\n",
    "            unbalanced_to = node\n",
    "\n",
    "    # Add an edge connecting the unbalanced edges.\n",
    "    if unbalanced_from in edge_dict:\n",
    "        edge_dict[unbalanced_from].append(unbalanced_to)\n",
    "    else:\n",
    "        edge_dict[unbalanced_from] = [unbalanced_to]\n",
    "\n",
    "    # Get the Eulerian Cycle from the edges, including the unbalanced edge.\n",
    "    cycle = eulerian_cycle(edge_dict)\n",
    "\n",
    "    # Find the location of the unbalanced edge in the eulerian cycle.\n",
    "    divide_point = list(filter(lambda i: cycle[i:i+2] == [unbalanced_from, unbalanced_to], range(len(cycle)-1)))[0]\n",
    "\n",
    "    # Remove the unbalanced edge, and shift appropriately, overlapping the head and tail.\n",
    "    return cycle[divide_point+1:]+cycle[1:divide_point+1]\n",
    "\n",
    "#example\n",
    "kmers = ['CTTA','ACCA','TACC','GGCT','GCTT','TTAC']\n",
    "\n",
    "adjlist,nodes = build_deGruijnGraphFromKmers(kmers)\n",
    "path = eulerian_path(adjlist)\n",
    "epath = [nodes[p] for p in path]\n",
    "string = [epath[0]]\n",
    "for pa in epath[1:]:\n",
    "    string.append(pa[-1])\n",
    "print(''.join(string))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Translate an RNA string into an amino acid string.\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "text = 'AUGGCAUUGGAUCUCGUUCAAUACCCUAUCAUCGGUGUAAGCUUGUGUCAUUGCGCCGGCUUGCGCGGACGUCCGAAAAUCGGUAGUGCGCUACCAUUGCCGUGUGCAUCGCUUCGCAUGGUUUACGAAAGUGGAUGCAUACCCUACCACAAGGUUACCCAUUCCAACUGGAGCUACGAAGCACUUGGGGCAAGUAUUUCCUUGACAGUAUGGAUUAACGGAUACAUUUAUGGAUGUUCAAAAGGUCGAACAGUGACCGACCGGUUUCUACUUGUUGUGCCCAAAGGAUACGGUCUGCUUUAUAAUGAAGCCAAGGAACUCUUACCAAUGGGAAUAUAUACUGUCAGCAUGCUGCCACCCAAGUACCUAGUAUUUAACGGACCCAAGGCGUUCCGGGCGAAGAGGGGUAACCUGCGGCCCAACAGUAAUAAUGAGCCACCCAGUGGGCCGCAACGGGAACUUAUUAGGAGAGUGCCAUCGUAUUAUUACGACUUGUCACGGGCCUACUCCGUAGAGAACACGCCACCGCCUGACUUUGCCAACACCGCUAUCUCCCCGGGAGUUGCAUCCCCAAUGUAUCAUGCAGGGUUACAAAGCCCCCUCUCUUACCUUCAUCAACGUGAUAUGCCUGCCGGAUUGAAACCUUUUCUGAGAAGGGCGAGUUCCCUCCUGACAGCACAUUAUCUUAUUUUAACUCAUCAGGCUGGGGGGGACCUCUGGCUGAGCAAUCUUAUUUGCCACCCAACACCCACACGGUUACACGUGAUUAGCGCACAUGCUUUUGCACUAGGAAAUAACCCUGGAGGCAGGCUAGGCCUUUGCCAGUGGAAUGGAUCAUCGCAUUUUUACCCACAUCCCGCGAGCGCGAUCCACGACUACAUCGGGGUCGGUAUAGCUACAUUACGGGUAUCGUUGAAAAGCGUAGUUACAGAGACCAUAGCAGGUACCGAGCCUAGCGCAUCCGAAUCAUGUUUACGCCUGCCGGGUGCUCGCUCAUACUUAUUCCAGGUUCGCACCGGGCGUAUAGACAGAAAGAUCCUAUUCGAUCCGAUAUGCCGGAGGGCAGAUCUAUGUUUUGAGCGUGUAAGAAGCCGAAGAGCCCGUACUCCUGACGGGGUCUUAAGAUUACCCCUUCGCCCUGAAGACUUGUACGCGCAAUGCCGUCUUCUGUCCCUAACGCUGAGGACAUAUACUCAGUUCCGACCUAAAUGUGCGGGAGCUCCGAGCGUGGGCACGCGUGUUCCUGAUAUGGUAUUAGAUCUGUCUUGUCACGAUUUAAGACUUCAAAGCGACGGCAUGAACAGUAAGUCACAUCGCCUGGCACUUUGCGUGCGUCAUACGGGCUGCCAUUCAAAAAGAUCAGCCCAGAAAGAACCCCCGGAAUGGUUCGGCGCAAUGGAUCCCAAAUUUAUGGGUCCGCUUUGGCUCUACAAUGACGUGGGAGAUCCUGUUAGGUUGCGGCCUUUAGUACACGAGUAUAGGACCCAAGUGGGACGGUACAGAGGGUCACUGACGGAUCCAAAACCCCACCUCUCAGCGGUGACUCCCGCCAGCGGCUCACUCUGGCUACUUAAAGACGUCGGGAUCGUUAGCUGUCCGCCCCAACCCAAGACUCUAUACGAAAUUAAAAAUCGGCACCGUCCCAGUGUUGAAGGACCUGUGCAGAAAGGCUGGUACGAUGCCUCUGAAAACAGCUGUCAAACCCGGAGACGUCGGUGUUCUCCGCAGAUGCCACGAUCCCAACAGCGCAACGGACAUGCGCUGCCUGCUGUCUGGCCGAGAAUACAGGAAAAGUUAUUACAUUUAAGAGUUGGUAUCGUCCUGCCUGAAAUCCCGUUGAAUUAUGAGGGUCAUUUGUGCACUGCAGACGACCGAGUCGGUAUCCUCCCUACCAGGAUAGCAUGGAAGUGGGAAGACACCAAUGAUACGGCAUCGGCACCCGGACCAGACGGCGCUAUUAGUCGCGAGGCAUUUUACACAUCCGUCCCAACAGCGGAUUUAAUCCUGAGGGGGUGUCAGGAAUCCCUAUCUACCGCCAGAGAGUGGAAUAAGUUGCACGCACGAGUGGACCAGGAUGGGGUCCCUCCCGCAUGUUCUGCUUGCAGAGCAACAUUUCGGCUAAGGUCAAGCGAUAUUCAUACUGCUUUCGACCCCGAAUUCGCGAACUGCCCCCCGUCCACUAUAUCGCAUCGACCGAUCGGGCGUGCUCGGCACCUAUCUUCACCGAAAACUGUGUUCAGUCUUUUCCAGAAUGCACUAUCAACGGGACCGCGGCGAGCCGGCGGUACUUACUCUGGCCCAACGGACAUGAUACGAAGAGUGAAUUCAAGAGCGUCUUGUAGCGAGGAACCGACGAACUGUAUCCAGGCAGCACAAGCCGCUCAUGCUCCUGGGCUAAUGGUACAUUUCAGGAGUCCCCGGAUGUUACAACGCCGGGAAAUUCAAGGCAAAAGACGGCGAACCUGGAGUACUCAGAGGGUGGCCCAGUGCCGAGCAUAUACCAUAGACCGUCUACACAGACUCCUAACCCGUUGUAUUACGGUCAGAGCAGUCCCACAUAAAUAUUCAGAAUACUUCGGGACAACCCCGCAUAGUUGCGAGGCGUCUCGUCAGGCCCAUUUGCGGGCUCCAAGUAGCCUACUUGGCGACUUAUACCUACGUCGCCUCUCACUACACACGAAUAUGGCAUGCUGGCCGUGCGGGAUUGGACUUUUCAGAAAUUAUACAGCUGCAACCGGCCUUAAUUGCGGUCUUACUUUCCUGAGAGGAACCUUUAUGGGUCGUGCCAGUGGACGUCCCGUGGAAAAUGGCAUCCUGGAUUCAACAUCGCAGCACAUCCCCGUUUGCACUUACCCGCCAACCCGCCUUGAAUCGACGACAUGCUCGGCCACUAGGCGGGCAAGCAGAGAGAAAUCAGGGACCCCCCAACACGUAUUUAACGUGACAUGUAAUCGGGUAAUUAGGGGCGGUAGACGGAACUCAUUCCACCACCGCACUUCGGGUUCAGUGUCGGCAGUAACUGAUUGCUCUGCCUUGCGGCACGUGCAUUGGAUGCUACUCACGUUCAGUGAAAGUACCGCUAGACUCCUGAGGGUAUUACGAACAAACCCGAAGCCACUAACGCGCUUGCAGGCGCGCCUCCGUUCUCGGGGUAUCGUGCAGUGUAAACAACGUCACAAACUGUCGUACGAGUUGCUGUACGUAUGUAAGAGUAGCCACUUGUGCUUCUCGGGAUAUGAGACCGCCCCAUUGGGGUCAUCCGACUGGAUAAAGUUUAUGGUCUGCGGCAAGGUAAGGUUGAUGGCCAUCGCUGGGCCAAACGAGCCUCCAAGGGUGGUUAGCUUUCACGCCGUGCGAACGGGAAAGGAGAUUGCUAUUCUGUUCGGUGCGACUAAGGUUACACCCCAUUACCAACCCGCGGCCCAAACCACCGCGUUCACAGCUCCGCUAUCGCAAAUCCCUCUUUCUAAAUGCCGACGAAGUCCCACCCCUCCAUCAUGGACGUGCUACCUAAGGCCCGUUUAUCACACUACUCAUCAGAGCCGACCCCCGAAAGGUUUGCAAAAGACCAGGACUACGAAGCUGCAAACACGACGAGAGCUGGACAAAGAUCGUACGUGGAGUGUGACGUCCAUUAUCACCACGCAGCCUUCAGUGUUUAAGCACCAAGGACUAAAUGGACUAUCAAUCGAUGUAAUGGACGAUCGGCUGACCAACGUAUACCACUCAAGUACGCCGGACAGUGGGAGUGUAGAUGCGAAUAUCAGACCGGAGCACAGCCCAAGAAACUACAACCUACUGUGUUACUUAGAGUCAGGCCUAAUGCUUAGGCGGUCGGGAUUCUCCGCAGUAACCCCGCCAGCGGAGCCGAGGAAGAUGGAGAUCUACUUACUCCAACAGUGCAUAUCGCUACCCAACGUGAGCGUACAUCCUCAGGGCGGCAUUUCAGGGCUUCGCAUGGACUGGGAAGAUGUGCGGAACAUACCUGAAGAGUCAGUCAAUUGGGGGGACGGACGUCCGGACGGGACUGAAGGUUUAUUAAGAAGAAGCGUAUCUCCACUUGGGGUAUUCCAAAUAGAAACUCCUAUACAAGACAGGGAAUGGGAGGAUUCACCCAACGCACGUUACAAUGUGGCCGCUAUGCCUUGCACGAAGGGUCAAAUAUGCCCGGAGUGCCAAUCUAACAAGAACCGAACCUGGGCGUUUGCAUUUAACUGGGGACCGCUACUAAUGUGCACAUCUAACCCUGACCGUCUGGACACCGUACGAGUAACGAAUCUCCCGCUGCAGAUCAUAGAAUUCAGCUCCAUCUCACCAUGUGCAGUAUCACAGCCUGAACUCCGAUCCGGGUCUAAUCUGAGGGACAGCCGGGGAAACGAGUCAUGGUUAGCGAUUAUGACCCGGUAUGAGACGGAUUCCCUACAAUCUGUUAUAUUCAUUCACCGUCGGACGAGUCUAACGACAAACCCGGACGUCGAGGGUAUCGCAUACAGACCGGUGUACGGGCGUUAUGUAGACGCUCCAGUAGAUCAACAAAAGUGGAAUGUCACGACCCGAAGCAUCCAUCGAGAUUUUGGCGGUGGAUUCGAGCACGACAACGCUGAGAGUCCAAGAGGGACACAUUGCUCCCGCUGCAGUACAGCGAACACUUUUCUCGGUGUAUCAGCGUUUGGCACAGCUGAGACCCUGAAGAAGAGUGCCUCGAGGGUACAGGAACACGGAUGUAUUCGUCAUCAUGAUAUCAGCGCUUUUCGCGGUAGUAGGGGAACGUCUUCUCUUGAGGGCAUAAACGUUGUGGCGUGUCCUGGUACAAUAUCCCUCGUUUUACACAGAAUAACCGACGAAGCGGUAGUGGUGAAAACCGUCUGGCUGCGUCCGGAAAAGUUCGUCAGCGCCAAGAUCGCAUCAACGGGGGGAAGCCAGACUAGAGGUAUGAGUCAUCUCUACCAGAUACUUUUGAGAAGCCCCAUGAUGGACGCAAGGAGUGAGUCCGAUCAAGCCGGUAUGUCGGAGCGCAUUAAAAUACGUCAACGUCCCUGCGUUAGACCCUACGUAUCGAGCCCCCAGAGCCUCACCUGCCAGAUAGAUUCCUUUCUUCCAUCCGGCUUGGCUUACUCUAGACUACGAGAAAGAGUUCUUAGUAUGAGGACUAUAUGUUAUAUAGCAACCAAAUCUGAGGCCUGGAAGGGAGCGCGUAUGCCUACGCAGGGCCUAAACUCUCCCACGACAACCGAUUUCUUCAGCCGAGGAGGACAGCUCAAACGCGAUGAACUCGUCCACUUUUGCCAGCGUGUUAUAUACCCCUCGCCAUGGGGACUGGGGGCGCAACGAGGUUACCUGGCGUGGCAUGAGGGAAAUGCUUCCCAUACGUACGUUUCUGUGAUUGAAAACAGAACAAGCUCGCUCGUGCGCUUGGGGUGGGAAGAUUUCAUCUAUAACGGCCCUUCUGCUUUCUGGACGUCAGCUUAUCCACUUUCCCACUCUCCUUCCGAGAUAAUCAGUAACAGUGUCCUCAUUGCUUCUGUUGGUACCAAGUACAAUCUACUCGCCGUAACUGUUUCCAAUAGCCAUAAGAGUUUCGAUAACGUCACAGCUAAUCGACUAAAGUACGACAAAGAAUCCACUGCAACCGUUUUUCAGAAACUCGUGUCGACCCAUAAAUUACGCUACAGCAAUUGUAUAACGAGAACAGACUCCAAUCUGCGAGUUAAGAUCACGCAGCCGCCAAGCUCUCAUUGGGCGAGGCGGACUCGCUGCGAGUUAAUAUUAUCUCCGAUUCGAGCUUUUCAAAUGUGUAUCGACAGGUGUAGUUUCCCGAGUAGAAGUGGACAUCCAAGAUCCACUCGAACAAAGCCCCCAUCAACCGACGUUGUCUCCAACGCGGGCGAAUGUUCGCCCAUCCUUAGACAUUCGAGAGGGUUGGCAUGUACGGGAAUUGUUCUUGUCGGCUGUCUAGCGAUCGCAGCCCUGAGCGUUGUGCAGUGUGAGUCCUUUGUUCCGCGGCUCAAAGAAGCUAUCGUGUUCACGAUUGCUACUCUAGUCAAUCCAUUUUUCAUUCGUAUUCGCACCGAACUGUACGGGUUAGCUUGGGAUCGUCGUUUAUACGCACCCGUACAACGUAGGGCUGGACCCCUAGAGUAUCUCUGGGGAGGCAAAAUAGUGAGCAAUAUAACACGGUUAGACCACUGGAUAUACAGGCAGGCGGGUCAGUUUUUUACGUCCUCUCAUCGAGGCGGCAGUCGAUCUGGAUGGACCUCGCCUAAAAACGAGCGCGGAUUGCACGCGUACGAUAAGUUUACCCGGGGUUGUUGUCAUGGCCUUCUCUCCCCCACGGUUGGCAGUUGCGAAAAAACUGCUGUCGACCAGCCACGCUGGUACAAUAACGAUGAUCUAAUGAAUCUACAUGCUCCCCAUGCAUCCGUAUUAGUGCACACUCCCCUGCGCAAGGCAUUUCUAGUGUGGUAUAAGACUCUUGCAUCCACAUCAAAUCCCGCUCAGAUGGUCUUCGAAGAACUUGGUGCGAAAGUCCUACCUUGUAUAAUUGUACUGGGUUUGCUCGACGAACGUGUUGCGAGCUGGGCACAACUGCCGCCGUCCAUCGGCAAUAUAGCCAAACGUUCGCAACGCUCCGACGAUGAGACUCGCUCGUCUGCCUUGGAUCACCGGAAAGUAAGACGGCCUUAUAUUAUUUCAGAGAGUGAAGUCAGACUACGCUGCGUUUCGGAAUAUAAUCACUUUUUAACCUUUCUGCUCCCAAUAGAUCGGUCCAGGGGUAUCCAGGCUUGGUCAGAUAUGUGGUUUACAGAGUCGGACACGAACACCGCGAAGGAACCUUACGGCAUCUAUAAACCGCACGCUGGAAAUCAAGGGCACUGUAUUGUGUUGAACCUUAGUAUGCGGGCUUAUGUCUUGCUGUCCAUUAAGACAAUAUACGAUAUCGUCAGACAAUCGGGGGGGUGUGAUUUCGUAAGCGACCCCCGAAGUCGUCUACGCCCGCCUGAAACUCCCCCCCGCGAGCAUCUGUUCCCUUAUACGCGGGAAUGCGUGCGCCUACAUGUCACUGCACCCGAAGCCAGGGUAAUGAUGCUAAGGAUUCCGUGGCGCUUAAUGGCUUCCGCAAAACAUAAUUUCAAUCCAUGUGGCCUGUCGCAUGACAAUCGCAUGGAGGCGGGGCCCGCUUACCAGCCGCUACUACGUAUUGCAGAGAUCUUUAAGUGCGUUCAUCUAUGCUGUCACUUACUCCAUUGCCUAGGGGUGUCCUUUGGUACUAAGUUUUCUACAUACGAGGGGAGACCGUUAUCGGCUUCUACUAAGAUACAGUCUCACACUCUAUCGCUCCCUACGACCCUCUUGACGACUAGACUCACGUCUGGACACUUGUGUAUUGUGGGACUAGUCGUCCGGUUUGAUGCAACGCGUCCAAUGCCUAGCGGGUGUUUAGGAACUAACAUAGGAGGCCUUCAGCCUGGAGGGGUAAGAGAUAAAAAGGUGCGUUUAAGGUCGUCCGAGGUUGAGCCAAGAAGCAAGCAACGCAUUGUAAGUGGCCAAAGUGCCAAUCUUCUCGGCGCAGGCGUUUCCAGCGAGCUGGAGGCGGCUGCAACUGAACGAAUGCCCGUCAUCCGCUUAUCUCCCUGGGUAGUACCGACGGAUCAGAGUGUAUAUUUGUCCCAGGAUUUUCAACUGCGCAAACUCAGCCGCGAUAACCUCGCUAAUUCCUCAACGCUUCUGGAAGGGGAGGUGACACCGUCCCAUGUUGGGAAUACCCCGGCUGCAAGACACACUUCUCCGAUCAUACAGGGAUAUCGCACAGGUCGGAAUAUGAGCGCUCUCUUCGUUCACGUUCGCAGAGAUACCAGAGGUCCGGAGGUACGAAGCUUAGCAUCAACAUUCAGUACUCCAGAGGGCACCAUGCAGAACCCGCAGUUGUGCUUCCGCGCUAAUCGGGCCUAUCCUAUACUAUCAUCGGAAAUGCCUGCAAACCAGUGCACUUGCGGAUGUUGCAUUAAAAUGCCCCACGCUGCACGGGAUGGUCUCUCUCUUUUGAGUCCUCCGAGCUUAACGUAUCAUGCCCACAGACCAUCAGCGAGCAAUGAAAAUUCGAGGUCGACGGAGUUAACUCGGAUUACCGCUCCCGACUGCACCGAGACAUCUCCCGAAAUCGAACGAUCUACGGGCUUAACCCCUACGUCUAGGAGCGCUUUUGCAAGCGUAGAGAUUGCAAAAUUCCAAACAGCCUCACCCGGCACUCCGCGUACCGGGCAAAAGUCUCCAAUUGGUAGGUGCCAAAGCGGAGCGGAGGCAGCUAUAGCAGGGUUUGCGGGAGGUUGCAUUUCUUGUAUCGUAAGUAGCAAACUUCAAAUCGUCACUUUACUCAGCACGAUUUACAAACUUGCCGUAAUGCAAUUGCUUACUGCGAGUGUACGUCUAAGAGGAUCCCGUCUUAUUAUCUGUAGUGUCCAUGGCCAGUCCGGCUUCCACCAGUUAUACCAAGGCCAGAUCUUCUUAACAAAACCUGUUAUAGACCUCUGCCUGAAGUACCAUGCAUUCCCGGUCUACUUUGAGAUGGUCCUAGGCGCGUGGUCGCCCGUGGCGAGGGAGUUUGUAUGGUCCUCCGGGCCAAGUAAGACCGGAGUCCAGAAAUUCGUACCCAGCCAGCGGAGAACGGGGUUUCAUAAGCACAAGGCCUCGACGUCAGGCUUGGACCCGCUAGUGUGGGCAGCUCUAAAUUGUAAAGCUCAGCUCUGCCGUUUUGGCGUUGGACUCAUCGAUCAGUUUCUCCGGGGUCACUCUGCGAUCGUAGCUGAGUGCGAUCCUCAACGUAUAUCUCGACAUGGUUACCCCUCAUCCUCCUACAAGGGCAGCCUGGAUGACCGCCCGGAACCGCCUAGCCACGCAUGGUGUUGUCACAGCUUUUACAAGGAAAAGAGUUAUUUGCAUACUAAUUUAGGAAGCAACUUGGCGAAAGUGGCAGCAGCAUAUGAUUUUCCCACUGUCCCCCCGGGAGAUCAUCUAAAGAUACGCUGUGGUUCUAUUGAUAACGAGUUACAAGUGCAGGUCACAGACGACACGGCCCAGGCUGAUCAUGUAAACCAUUUGCUUAGGCAAAGUGACGGCCGGAGCUCUGACACGCACUCCAUAUUUCUGACCCCGACACCCGCUAUUGGCGCCUGUUUACGAAGGAAUAUCAUUUUGAACGUCAGGGGAAGUGCAUUAAUCGAAAACGUUGAAUCGGUGGCCCCAUCCCAGAUAUGCCAUCUACGGGAUAUGAGUACGACGGUUUCGCUUCCCAUUGUAUAUGCUCCUAGAGGCCAGAUCAACUUCUGGAGCGUUACGCGAUAA'\n",
    "\n",
    "coding_rna = Seq(text, IUPAC.unambiguous_rna)\n",
    "coding_rna\n",
    "\n",
    "protein_seq = coding_rna.translate()\n",
    "print(protein_seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find substrings of a genome encoding a given amino acid sequence.\n",
    "# A DNA string Text, an amino acid string Peptide, and the array GeneticCode\n",
    "# find All substrings of Text encoding Peptide (if any such substrings exist).\n",
    "GeneticCode = {'ACC': 'T', 'GCA': 'A', 'AAG': 'K', 'AAA': 'K', 'GUU': 'V', 'AAC': 'N', 'AGG': 'R', 'UGG': 'W', 'GUC': 'V', 'AGC': 'S', 'ACA': 'T', 'AGA': 'R', 'AAU': 'N', 'ACU': 'T', 'GUG': 'V', 'CAC': 'H', 'ACG': 'T', 'AGU': 'S', 'CCA': 'P', 'CAA': 'Q', 'CCC': 'P', 'UGU': 'C', 'GGU': 'G', 'UCU': 'S', 'GCG': 'A', 'CGA': 'R', 'CAG': 'Q', 'CGC': 'R', 'UAU': 'Y', 'CGG': 'R', 'UCG': 'S', 'CCU': 'P', 'GGG': 'G', 'GGA': 'G', 'GGC': 'G', 'CCG': 'P', 'UCC': 'S', 'UAC': 'Y', 'CGU': 'R', 'GAA': 'E', 'AUA': 'I', 'AUC': 'I', 'CUU': 'L', 'UCA': 'S', 'AUG': 'M', 'UGA': '', 'CUG': 'L', 'GAG': 'E', 'AUU': 'I', 'CAU': 'H', 'CUA': 'L', 'UAA': '', 'GCC': 'A', 'UUU': 'F', 'GAC': 'D', 'GUA': 'V', 'UGC': 'C', 'GCU': 'A', 'UAG': '', 'CUC': 'L', 'UUG': 'L', 'UUA': 'L', 'GAU': 'D', 'UUC': 'F'}\n",
    "\n",
    "def rna_translate_protein(rna):\n",
    "    i = 0\n",
    "    protein = []\n",
    "    while i < len(rna)-2:\n",
    "        if len(GeneticCode[rna[i:i+3]]) ==0:\n",
    "            return protein\n",
    "        else:\n",
    "            protein.append(GeneticCode[rna[i:i+3]])\n",
    "        i = i + 3\n",
    "    return ''.join(protein)\n",
    "\n",
    "def peptide_encoding(text,peptide):\n",
    "    k = 3 * len(peptide)\n",
    "    ret = []\n",
    "    for i in range(len(text)-k+1):\n",
    "        rna = text[i:i+k].replace('T','U')\n",
    "        protein = rna_translate_protein(rna)\n",
    "        if protein == peptide:\n",
    "            ret.append(text[i:i+k])\n",
    "    reversetext = reverse_complement(text)\n",
    "    for i in range(len(reversetext)-k+1):\n",
    "        rna = reversetext[i:i+k].replace('T','U')\n",
    "        protein = rna_translate_protein(rna)\n",
    "        if protein == peptide:\n",
    "            ret.append(reverse_complement(reversetext[i:i+k]))\n",
    "    return ret\n",
    "\n",
    "def count_seq(peptide):\n",
    "    ret = 1\n",
    "    for pep in peptide:\n",
    "        ret = ret * len([i for i, v in GeneticCode.items() if v == pep])\n",
    "    print(ret)\n",
    "\n",
    "def reverse_complement(dna):\n",
    "    dnadict = {'A':'T','C':'G','G':'C','T':'A'}\n",
    "    reverseDna = [ dnadict[c] for c in dna ]\n",
    "    reverseDna = reverseDna[::-1]\n",
    "    return ''.join(reverseDna)\n",
    "\n",
    "# text = \"ATGGCCATGGCCCCCAGAACTGAGATCAATAGTACCCGTATTAACGGGTGA\"\n",
    "# peptide = \"MA\"\n",
    "protein = peptide_encoding(text,peptide)\n",
    "print('\\n'.join(protein))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#How many subpeptides does a cyclic peptide of length n have\n",
    "def subpeptide(n):\n",
    "    subpeptides = n*(n-1)\n",
    "    return subpeptides\n",
    "\n",
    "subpeptide(35046)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generate the theoretical spectrum of a cyclic/linear/other peptide.\n",
    "#The theoretical spectrum of a cyclic peptide Peptide, denoted Cyclospectrum(Peptide), is the collection of all of the masses of its subpeptides, in addition to the mass 0 and the mass of the entire peptide, with masses ordered from smallest to largest\n",
    "MassTable = {'A': 71, 'C': 103, 'E': 129, 'D': 115, 'G': 57, 'F': 147, 'I': 113, 'H': 137, 'K': 128, 'M': 131, 'L': 113, 'N': 114, 'Q': 128, 'P': 97, 'S': 87, 'R': 156, 'T': 101, 'W': 186, 'V': 99, 'Y': 163}\n",
    "\n",
    "def theoretical_spectrum(peptide):\n",
    "    k = len(peptide)\n",
    "    spec = []\n",
    "    while k > 0:\n",
    "        for i in range(len(peptide)-k+1):\n",
    "            subpep = peptide[i:i+k]\n",
    "            spec.append(sum([MassTable[s] for s in subpep]))\n",
    "        k -= 1\n",
    "    spec.append(0)\n",
    "    return sorted(spec)\n",
    "\n",
    "def linearSpectrum(peptide):\n",
    "    prefixMass = [0]*(len(peptide)+1)\n",
    "    for i in range(len(peptide)):\n",
    "        prefixMass[i+1] = prefixMass[i] + MassTable[peptide[i]]\n",
    "    linearSpec = [0]\n",
    "    for i in range(len(peptide)):\n",
    "        for j in range(i+1,len(peptide)+1):\n",
    "            linearSpec.append(prefixMass[j]-prefixMass[i])\n",
    "    return sorted(linearSpec)\n",
    "\n",
    "def cyclicSpectrum(peptide):\n",
    "    prefixMass = [0]*(len(peptide)+1)\n",
    "    for i in range(len(peptide)):\n",
    "        prefixMass[i+1] = prefixMass[i] + MassTable[peptide[i]]\n",
    "    peptideMass = prefixMass[-1]\n",
    "    cyclicSpec = [0]\n",
    "    for i in range(len(peptide)):\n",
    "        for j in range(i+1,len(peptide)+1):\n",
    "            cyclicSpec.append(prefixMass[j]-prefixMass[i])\n",
    "            if i > 0 and j < len(peptide):\n",
    "                cyclicSpec.append(peptideMass - prefixMass[j] + prefixMass[i]) #<--Nice!\n",
    "    return sorted(cyclicSpec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "peptide = 'MKIMSAFWIDSSED'\n",
    "spec = cyclicSpectrum(peptide)\n",
    "print(' '.join(map(str,spec)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "50747776"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#How many subpeptides does a linear peptide of given length n have\n",
    "def linear_subpeptide(n):\n",
    "    subpeptides = int(n * ((n+1)/2)) + 1\n",
    "    return subpeptides\n",
    "\n",
    "linear_subpeptide(10074)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#match peptides with the spectrum. Returns index of target peptides that matches the spectrum             \n",
    "def linearspectrum(peptide):\n",
    "    amino_acid_mass_table = {'G':57,'A':71,'S':87,'P':97,'V':99,'T':101,'C':103,'I':113,'L':113,'N':114,'D':115,'K':128,'Q':128,'E':129,'M':131,'H':137,'F':147,'R':156,'Y':163,'W':186} \n",
    "    sub_peptides = [['',0]]\n",
    "    for l in range(len(peptide))[1:]:\n",
    "        for pos in range(len(peptide)):\n",
    "            if pos + l <= len(peptide):\n",
    "                sub_peptide = peptide[pos:pos+l]\n",
    "                sub_peptides.append([sub_peptide,0])\n",
    "    if (peptide):\n",
    "        sub_peptides.append([peptide,0])\n",
    "\n",
    "    for entry in sub_peptides:\n",
    "        entry[1] = _get_peptide_mass(amino_acid_mass_table, entry[0])\n",
    "    return sorted(sub_peptides, key=lambda entry: entry[1])\n",
    "\n",
    "def _get_peptide_mass(amino_acid_mass_table, peptide):\n",
    "    mass = 0\n",
    "    for pos in range(len(peptide)):\n",
    "        mass += amino_acid_mass_table[peptide[pos]]\n",
    "    return mass\n",
    "\n",
    "def quiz6():\n",
    "    print(\"----- Quiz 6 -----\")\n",
    "    #peptides = [\"TCE\", \"AQV\", \"VAQ\", \"ETC\", \"CTV\", \"CET\"]\n",
    "    peptides = [\"QCV\", \"ETC\", \"TVQ\", \"AVQ\", \"TCE\", \"TCQ\"]\n",
    "    spectrum_set = set(\"0 71 99 101 103 128 129 199 200 204 227 230 231 298 303 328 330 332 333\".split(' '))\n",
    "    for peptide in peptides:\n",
    "        sub_spectrum = [str(item[1]) for item in linearspectrum(peptide)]\n",
    "        if set(sub_spectrum).issubset(spectrum_set):\n",
    "            print(str(peptides.index(peptide)+1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "    CYCLOPEPTIDESEQUENCING(Spectrum)\n",
    "        Peptides ? a set containing only the empty peptide\n",
    "        while Peptides is nonempty\n",
    "            Peptides ? Expand(Peptides)\n",
    "            for each peptide Peptide in Peptides\n",
    "                if Mass(Peptide) = ParentMass(Spectrum)\n",
    "                    if Cyclospectrum(Peptide) = Spectrum\n",
    "                        output Peptide\n",
    "                    remove Peptide from Peptides\n",
    "                else if Peptide is not consistent with Spectrum\n",
    "                    remove Peptide from Peptides\n",
    "Sample Input:\n",
    "     0 113 128 186 241 299 314 427\n",
    "Sample Output:\n",
    "     186-128-113 186-113-128 128-186-113 128-113-186 113-186-128 113-128-186\n",
    "'''\n",
    "\n",
    "input = '0 101 103 113 128 128 128 128 129 147 163 214 231 248 256 256 257 257 266 275 276 359 361 376 377 379 384 385 385 394 404 480 487 489 505 507 513 513 522 524 532 608 616 618 627 633 635 641 650 652 660 736 744 746 755 755 761 763 779 781 788 864 874 883 883 884 889 891 892 907 909 992 993 1002 1011 1011 1012 1012 1020 1037 1054 1105 1121 1139 1140 1140 1140 1140 1155 1165 1167 1268'\n",
    "with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/integer_mass_table.txt') as f:\n",
    "    masses = list(set([int(line.strip().split()[1]) for line in f]))\n",
    "\n",
    "spectrum = [int(i) for i in input.strip().split()]\n",
    "#print spectrum\n",
    "\n",
    "\n",
    "def expand_list(peptides, masses):\n",
    "    if len(peptides) == 0:\n",
    "        return [([m], [0, m]) for m in masses]\n",
    "\n",
    "    def combine_spectrum(peptide, mass):\n",
    "        def extend_spectrum(masses, m):\n",
    "            return masses + [m] + [(sum(peptide[0][i:]) + m)\n",
    "                                   for i in range(len(peptide[0]))]\n",
    "        return (peptide[0]+[mass], extend_spectrum(peptide[1], mass))\n",
    "    return [combine_spectrum(p, m) for p in peptides for m in masses]\n",
    "\n",
    "\n",
    "def is_consistent(linear, spectrum):\n",
    "    for e in linear:\n",
    "        if linear.count(e) > spectrum.count(e):\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "consistent = []\n",
    "\n",
    "for i in range(20):\n",
    "    consistent = [cand for cand in expand_list(consistent, masses)\n",
    "                  if is_consistent(cand[1], spectrum)]\n",
    "    print (\"i: %s, len(consistent): %s\" % (i, len(consistent)))\n",
    "    if len(consistent) == 0:\n",
    "        break\n",
    "    print (*(sorted(['-'.join([str(i) for i in c[0]])\n",
    "                   for c in consistent])), sep = ' ')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Cyclopeptide Scoring Problem: Compute the score of a cyclic peptide against a spectrum\n",
    "# To generalize the Cyclopeptide Sequencing Problem to handle noisy spectra, we need to relax the requirement that a candidate peptide’s theoretical spectrum must match the experimental spectrum exactly, and instead incorporate a scoring function that will select the peptide whose theoretical spectrum matches the given experimental spectrum the most closely.\n",
    "\n",
    "MassTable = {'A': 71, 'C': 103, 'E': 129, 'D': 115, 'G': 57, 'F': 147, 'I': 113, 'H': 137, 'K': 128, 'M': 131, 'L': 113, 'N': 114, 'Q': 128, 'P': 97, 'S': 87, 'R': 156, 'T': 101, 'W': 186, 'V': 99, 'Y': 163}\n",
    "\n",
    "def cyclicSpectrum(peptide):\n",
    "    prefixMass = [0]*(len(peptide)+1)\n",
    "    for i in range(len(peptide)):\n",
    "        prefixMass[i+1] = prefixMass[i] + MassTable[peptide[i]]\n",
    "    peptideMass = prefixMass[-1]\n",
    "    cyclicSpec = [0]\n",
    "    for i in range(len(peptide)):\n",
    "        for j in range(i+1,len(peptide)+1):\n",
    "            cyclicSpec.append(prefixMass[j]-prefixMass[i])\n",
    "            if i > 0 and j < len(peptide):\n",
    "                cyclicSpec.append(peptideMass - prefixMass[j] + prefixMass[i]) #<--Nice!\n",
    "    return sorted(cyclicSpec)\n",
    "\n",
    "import collections\n",
    "def cyclic_scoring(peptide,expSpectrum):\n",
    "    theoSpectrum = cyclicSpectrum(peptide)\n",
    "    theoSpec_multiset = collections.Counter(theoSpectrum)\n",
    "    expSpec_multiset = collections.Counter(expSpectrum)\n",
    "    overlap = list((theoSpec_multiset & expSpec_multiset).elements())\n",
    "    theoSpec_remainder = list((theoSpec_multiset - expSpec_multiset).elements())\n",
    "    expSpect_remainder = list((expSpec_multiset - theoSpec_multiset).elements())\n",
    "    return len(overlap)\n",
    "\n",
    "#peptide = 'NQEL' \n",
    "#expSpectrum = [0,99,113,114,128,227,257,299,355,356,370,371,484]\n",
    "score = cyclic_scoring(peptide,expSpectrum)\n",
    "print(score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Implement LeaderboardCyclopeptideSequencing\n",
    "#output - LeaderPeptide after running LeaderboardCyclopeptideSequencing(Spectrum, N)\n",
    "aminoacid = ['G', 'A', 'S', 'P', 'V', 'T', 'C', 'L', 'N', 'D', 'K', 'E', 'M', 'H', 'F', 'R', 'Y', 'W']\n",
    "aminoacidMass = {'G':57, 'A':71, 'S':87, 'P':97, 'V':99, 'T':101, 'C':103, 'L':113, 'N':114, 'D':115, 'K':128, 'E':129, 'M':131, 'H':137, 'F':147, 'R':156, 'Y':163, 'W':186}\n",
    "def expand(leaderboard):\n",
    "    \"\"\"Expands each peptide/aminoacid in leaderboard by all 18 aminoacids with distinct masses.\"\"\"\n",
    "    expanded = []\n",
    "    for i in leaderboard:\n",
    "        expanded += [i+j for j in aminoacidMass.keys()]\n",
    "    return expanded   \n",
    "\n",
    "def mass(peptide):\n",
    "    \"\"\"Calculates the mass of peptide using the aminoacidMass dictionary\"\"\"\n",
    "    massOfPeptide = 0\n",
    "    for i in peptide:\n",
    "        massOfPeptide += aminoacidMass[i]\n",
    "    return massOfPeptide\n",
    "\n",
    "def cyclicSpectrum(peptide):\n",
    "    \"\"\"Input: An amino acid string Peptide.\n",
    "     Output: The cyclic spectrum of Peptide.\"\"\"\n",
    "    prefixMass = [0]*((len(peptide)+1))\n",
    "    for i in range(len(peptide)):\n",
    "        prefixMass[i+1] = prefixMass[i] + aminoacidMass[peptide[i]]\n",
    "    peptideMass = prefixMass[len(peptide)]\n",
    "    cyclic_spectrum = [0]\n",
    "    for i in range(len(prefixMass)-1):\n",
    "        for j in range(i+1, len(prefixMass)):\n",
    "            cyclic_spectrum.append(prefixMass[j] - prefixMass[i])\n",
    "            if i > 0 and j < (len(prefixMass)-1):\n",
    "                cyclic_spectrum.append(peptideMass - (prefixMass[j] - prefixMass[i]))\n",
    "    return sorted(cyclic_spectrum) \n",
    "\n",
    "from collections import Counter\n",
    "def score_peptide(peptide, spectrum):\n",
    "    \"\"\"Cyclopeptide Scoring Problem: Compute the score of a cyclic peptide against a spectrum.\n",
    "     Input: An amino acid string Peptide and a collection of integers Spectrum. \n",
    "     Output: The score of Peptide against Spectrum, Score(Peptide, Spectrum).\"\"\"\n",
    "    spectrum_peptide = cyclicSpectrum(peptide)\n",
    "    c1, c2 = Counter(spectrum_peptide), Counter(spectrum)\n",
    "    return sum([min(n, c2[k]) for k,n in c1.items()])\n",
    "\n",
    "def linearSpectrum(peptide):\n",
    "    \"\"\"Input: An amino acid string Peptide.\n",
    "     Output: The linear spectrum of Peptide.\"\"\"\n",
    "    prefixMass = [0]*((len(peptide)+1))\n",
    "    for i in range(len(peptide)):\n",
    "        prefixMass[i+1] = prefixMass[i] + aminoacidMass[peptide[i]]\n",
    "    #print 'prefixMass', prefixMass\n",
    "    linear_spectrum = [0]\n",
    "    for i in range(len(prefixMass)-1):\n",
    "        for j in range(i+1, len(prefixMass)):\n",
    "            linear_spectrum.append(prefixMass[j] - prefixMass[i])\n",
    "    return sorted(linear_spectrum) \n",
    "\n",
    "def score_linear_peptide(peptide, spectrum):\n",
    "    \"\"\"Compute the score of a linear peptide with respect to a spectrum.\n",
    "     Input: An amino acid string Peptide and a collection of integers Spectrum.\n",
    "     Output: The linear score of Peptide with respect to Spectrum, LinearScore(Peptide, Spectrum).\"\"\"\n",
    "    spectrum_linear_peptide = linearSpectrum(peptide)\n",
    "    c3, c4 = Counter(spectrum_linear_peptide), Counter(spectrum)\n",
    "    return sum([min(n, c4[k]) for k,n in c3.items()])\n",
    "\n",
    "def trim_leaderboard(leaderboard, spectrum, N):\n",
    "    \"\"\"Input: A collection of peptides Leaderboard, a collection of integers Spectrum, and an integer N.\n",
    "     Output: The N highest-scoring linear peptides on Leaderboard with respect to Spectrum.\"\"\"\n",
    "    scores =  [[score_linear_peptide(peptide, spectrum), peptide] for peptide in leaderboard]\n",
    "    sorted_scores = sorted(scores, reverse = True)\n",
    "    if len(leaderboard) <= N:\n",
    "        return [i[1] for i in sorted_scores]\n",
    "    else:\n",
    "        return [i[1] for i in sorted_scores if i[0] >= sorted_scores[int(N)-1][0]]\n",
    "\n",
    "def leaderboard_cyclopeptide_sequencing(spectrum, N):\n",
    "    \"\"\" Input: An integer N and a collection of integers Spectrum.\n",
    "     Output: LeaderPeptide after running LEADERBOARDCYCLOPEPTIDESEQUENCING(Spectrum, N)\"\"\"\n",
    "    leaderboard = aminoacid\n",
    "    leaderpeptide = ''\n",
    "    parentmass = max(spectrum)\n",
    "    while len(leaderboard) > 0:\n",
    "        leaderboard = expand(leaderboard)\n",
    "        for peptide in leaderboard[:]:\n",
    "            if mass(peptide) == parentmass:\n",
    "                if score_peptide(peptide, spectrum) > score_peptide(leaderpeptide, spectrum):\n",
    "                    leaderpeptide = peptide\n",
    "            elif mass(peptide) > parentmass:\n",
    "                leaderboard.remove(peptide)\n",
    "        leaderboard = trim_leaderboard(leaderboard, spectrum, N)   \n",
    "    return leaderpeptide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute the convolution of a spectrum\n",
    "# The list of elements in the convolution of Spectrum. If an element has multiplicity k, it should appear exactly k times; you may return the elements in any order.\n",
    "\n",
    "def spectral_convolution(spectrum):\n",
    "    convolution = []\n",
    "    #spectrum = sorted(spectrum)\n",
    "    for i in range(len(spectrum)):\n",
    "        for j in range(i+1,len(spectrum)):\n",
    "            if spectrum[j]-spectrum[i] > 0 :\n",
    "                convolution.append(spectrum[j]-spectrum[i])\n",
    "    return convolution\n",
    "\n",
    "#spectrum = [0,137,186,323]\n",
    "convolution = spectral_convolution(spectrum)\n",
    "print(' '.join(map(str,convolution)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We now have the outline for a new cyclopeptide sequencing algorithm. Given an experimental spectrum, we first compute the convolution of an experimental spectrum. We then select the M most frequent elements between 57 and 200 in the convolution to form an extended alphabet of candidate amino acid masses. In order to be fair, we should include the top M elements of the convolution \"with ties\". Finally, we run the algorithm LeaderboardCyclopeptideSequencing, where the amino acid masses are restricted to this alphabet. We call this algorithm ConvolutionCyclopeptideSequencing.\n",
    "#output - A cyclic peptide LeaderPeptide with amino acids taken only from the top M elements (and ties) of the convolution of Spectrum that fall between 57 and 200, and where the size of Leaderboard is restricted to the top N (and ties).\n",
    "\n",
    "from collections import Counter\n",
    "\n",
    "def spectral_convolution(spectrum):\n",
    "    convolution = []\n",
    "    spectrum = sorted(spectrum)\n",
    "    for i in range(len(spectrum)):\n",
    "        for j in range(i+1,len(spectrum)):\n",
    "            if spectrum[j]-spectrum[i] >=57 and  spectrum[j]-spectrum[i] < 200:\n",
    "                convolution.append(spectrum[j]-spectrum[i])\n",
    "    return convolution \n",
    "\n",
    "def convolution_cyclopeptide_sequencing(spectrum,M,N):\n",
    "    convolution = spectral_convolution(spectrum)\n",
    "    theoSpecScore = [ c[1] for c in Counter(convolution).most_common(M)] #<-- DIDN'T CONSIDER TIES\n",
    "    threshold = min(theoSpecScore)\n",
    "    convolution = dict(Counter(convolution))\n",
    "    theoSpec = [k for k,v in convolution.items() if v>=threshold]\n",
    "    leadPeptide = leaderboard_cyclopeptide_sequencing(spectrum, N)\n",
    "    return leadPeptide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def linear_spectrum(peptide):\n",
    "    prefix_mass = [0]\n",
    "    for i in range(len(peptide)):\n",
    "        prefix_mass.append(prefix_mass[i]+amino_acid_mass_table[peptide[i]])\n",
    "\n",
    "    theoretical_spectrum = [0]\n",
    "    for i in range(len(prefix_mass)-1):\n",
    "        for j in range(i+1, len(prefix_mass)):\n",
    "            theoretical_spectrum.append(prefix_mass[j]-prefix_mass[i])\n",
    "    return sorted(theoretical_spectrum)\n",
    "\n",
    "def _linear_score(peptide, spectrum):\n",
    "    ls = linear_spectrum(peptide)\n",
    "    cs = spectrum.copy()\n",
    "    score = 0\n",
    "    for c in ls:\n",
    "        if c in cs:\n",
    "            score += 1\n",
    "            cs.remove(c)\n",
    "    return score\n",
    "\n",
    "def quiz4():\n",
    "    print(\"------ Quiz 4 ------\")\n",
    "    print(_linear_score(\"PEEP\", [int(mass) for mass in \"0 97 129 129 129 194 226 323 323 355 452\".split(' ')]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
