{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#https://github.com/atorang/replication-origin/blob/master/bioinformatics1.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count frequency of a pattern in DNA string\n",
    "def pattern(text, pattern):\n",
    "    results = 0\n",
    "    sub_len = len(pattern)\n",
    "    for i in range(len(text)):\n",
    "        if text[i:i+sub_len] == pattern:\n",
    "            results += 1\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#most frequency k-mers in DNA string\n",
    "def frequency(text, k):\n",
    "    frequent_patterns = []\n",
    "    for i in range(len(text)-k + 1):\n",
    "        pat = text[i:i+k]\n",
    "        count = pattern(text,pat)\n",
    "        frequent_patterns.append((pat,count))\n",
    "    frequent_patterns = list(set(frequent_patterns))\n",
    "    frequent_patterns.sort(key=lambda tup: tup[1], reverse=True)\n",
    "    max_count = max(frequent_patterns,key=itemgetter(1))[1] \n",
    "    pattern_return = [s[0] for s in frequent_patterns if s[1] == max_count] \n",
    "    return pattern_return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find reverse complement of DNA string\n",
    "def reverse(seq):\n",
    "    \"\"\"Returns a reversed string\"\"\"\n",
    "    return seq[::-1]\n",
    "\n",
    "\n",
    "def complement(seq):\n",
    "    \"\"\"Returns a complement DNA sequence\"\"\"\n",
    "    complement_dict = {'A': 'T', 'C': 'G', 'T': 'A', 'G': 'C'}\n",
    "    seq_list = list(seq)\n",
    "    # I can think of 3 ways to do this but first is faster I think ???\n",
    "    # first list comprehension\n",
    "    seq_list = [complement_dict[base] for base in seq_list]\n",
    "    # second complicated lambda\n",
    "    # seq_list = list(map(lambda base: complement_dict[base], seq_list))\n",
    "    # third easy for loop\n",
    "    # for i in range(len(seq_list)):\n",
    "    #    seq_list[i] = complement_dict[seq_list[i]]\n",
    "    return ''.join(seq_list)\n",
    "\n",
    "\n",
    "def reverse_complement(seq):\n",
    "    \"\"\"\"Returns a reverse complement DNA sequence\"\"\"\n",
    "    seq = reverse(seq)\n",
    "    seq = complement(seq)\n",
    "    return seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find start positions of pattern in a string\n",
    "def pattern_start(text, pattern):\n",
    "    results = []\n",
    "    sub_len = len(pattern)\n",
    "    for i in range(len(text)):\n",
    "        if text[i:i+sub_len] == pattern:\n",
    "            results.append(i)\n",
    "    return results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# find most frequent patterns within a clump in DNA string with frequency above certain threshold\n",
    "def frequency(text, k):\n",
    "    frequent_patterns = []\n",
    "    for i in range(len(text)-k + 1):\n",
    "        pat = text[i:i+k]\n",
    "        count = pattern(text,pat)\n",
    "        frequent_patterns.append((pat,count))\n",
    "    frequent_patterns = list(set(frequent_patterns))\n",
    "    frequent_patterns.sort(key=lambda tup: tup[1], reverse=True)\n",
    "    max_count = max(frequent_patterns,key=itemgetter(1))[1] \n",
    "    pattern_return = [s for s in frequent_patterns if s[1] == max_count] \n",
    "    return pattern_return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ClumpFinding(text, k, t, L):\n",
    "    new_patterns = []\n",
    "    for i in range(len(text) - L + 1):\n",
    "        frequent_patterns = []\n",
    "        new_text = text[i:i + L]\n",
    "        for j in range(len(new_text) - k + 1):\n",
    "            pat = new_text[j:j+k]\n",
    "            \n",
    "            count = pattern(new_text,pat)         \n",
    "            frequent_patterns.append((pat,count))\n",
    "        frequent_patterns = list(set(frequent_patterns))\n",
    "        pattern_return = [s for s in frequent_patterns if s[1] >= t]\n",
    "        for l in pattern_return:\n",
    "            new_patterns.append(l)\n",
    "    \n",
    "    new_patterns = list(set(new_patterns))\n",
    "    new_patterns = [to for to in new_patterns if to[1] >= t]\n",
    "    return new_patterns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Return skews over entire length of DNA from O - len(genome)\n",
    "def gcdiff(text):\n",
    "    skew = 0\n",
    "    skew_list = [0]\n",
    "    minimum = skew\n",
    "    for i in range(len(text)):\n",
    "        if text[i] == 'G':\n",
    "            skew += 1\n",
    "        elif text[i] == 'C':\n",
    "            skew -= 1 \n",
    "        skew_list.append(skew)\n",
    "    return skew_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Return index of DNA string with min skew\n",
    "def skew_min(text):\n",
    "    skew = 0\n",
    "    min_skew = 0\n",
    "    skew_list = [0]\n",
    "    for i in range(len(text)):\n",
    "        \n",
    "        if text[i] == 'G':\n",
    "            skew += 1\n",
    "        elif text[i] == 'C':\n",
    "            skew -= 1 \n",
    "        \n",
    "        skew_list.append(skew)\n",
    "        \n",
    "    min_skew = min(skew_list)\n",
    "    index_list = [i for i, e in enumerate(skew_list) if e == min_skew]\n",
    "    print(index_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get hamming distance between 2 DNA strings of equal length\n",
    "def hammingd(text1, text2):\n",
    "    ham_dis = 0\n",
    "    for i in range(len(text1)):\n",
    "        if text1[i] == text2[i]:\n",
    "            ham_dis = ham_dis\n",
    "        elif text1[i] != text2[i]:\n",
    "            ham_dis += 1\n",
    "    return ham_dis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#find pattern that matches DNA string with mismatch\n",
    "def mismatch_pattern(text, pattern, mismatch):\n",
    "    results = []\n",
    "    sub_len = len(pattern)\n",
    "    for i in range(len(text) - sub_len + 1):\n",
    "        hamd = hammingd(text[i:i+sub_len],pattern)\n",
    "        if hamd <= mismatch:\n",
    "            results.append(i) \n",
    "    string = ' '.join([str(elem) for elem in results])\n",
    "    return string\n",
    "\n",
    "# returns index of matched pattern with a mismatch in a DNA string\n",
    "def mismatch_pattern_return(text, pattern, mismatch):\n",
    "    results = []\n",
    "    sub_len = len(pattern)\n",
    "    for i in range(len(text) - sub_len + 1):\n",
    "        hamd = hammingd(text[i:i+sub_len],pattern)\n",
    "        if hamd <= mismatch:\n",
    "            results.append(i) \n",
    "    string = ' '.join([str(elem) for elem in results])\n",
    "    return string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#count number of times pattern occurs in text with slight mismatch\n",
    "def count_hamd(text, pattern, mismatch):\n",
    "    results = 0\n",
    "    sub_len = len(pattern)\n",
    "    for i in range(len(text) - sub_len + 1):\n",
    "        hamd = hammingd(text[i:i+sub_len],pattern)\n",
    "        if hamd <= mismatch:\n",
    "            results += 1\n",
    "    return results\n",
    "\n",
    "#count highest frequency pattern that occurs in text with slight mismatch\n",
    "def frequency_hamd(text, k, d):\n",
    "    frequent_patterns = []\n",
    "    for i in range(len(text)-k + 1):\n",
    "        pat = text[i:i+k]\n",
    "        count = count_hamd(text, pat, d)\n",
    "        frequent_patterns.append((pat,count))\n",
    "    frequent_patterns = list(set(frequent_patterns))\n",
    "    frequent_patterns.sort(key=lambda tup: tup[1], reverse=True)\n",
    "    max_count = max(frequent_patterns,key=itemgetter(1))[1] \n",
    "    pattern_return = [s for s in frequent_patterns if s[1] == max_count] \n",
    "    \n",
    "    return pattern_return\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get most frequent k-mers with upto d-mismatches\n",
    "from itertools import combinations, product\n",
    "\n",
    "def generate(s, d):\n",
    "    N = len(s)\n",
    "    letters = 'ACGT'\n",
    "    pool = list(s)\n",
    "\n",
    "    for indices in combinations(range(N), d):\n",
    "        for replacements in product(letters, repeat=d):\n",
    "            skip = False\n",
    "            for i, a in zip(indices, replacements):\n",
    "                if pool[i] == a: skip = True\n",
    "            if skip: continue\n",
    "\n",
    "            keys = dict(zip(indices, replacements))\n",
    "            yield ''.join([pool[i] if i not in indices else keys[i] \n",
    "                           for i in range(N)])\n",
    "            \n",
    "def frequency_hamd(text, k, d):\n",
    "    frequent_patterns = []\n",
    "    pattern_combinations = []\n",
    "    final_patterns = []\n",
    "    for i in range(len(text)-k + 1):\n",
    "        pat = text[i:i+k]\n",
    "        pattern_combinations.append(pat)\n",
    "        temp = list(generate(pat, d))\n",
    "        for p in temp:\n",
    "            pattern_combinations.append(p)\n",
    "    pattern_combinations = list(set(pattern_combinations))\n",
    "    for t in pattern_combinations:\n",
    "        count = count_hamd(text, t, d)\n",
    "        frequent_patterns.append((t,count))     \n",
    "    frequent_patterns = list(set(frequent_patterns))\n",
    "    frequent_patterns.sort(key=lambda tup: tup[1], reverse=True)\n",
    "    max_count = max(frequent_patterns,key=itemgetter(1))[1] \n",
    "    pattern_return = [s[0] for s in frequent_patterns if s[1] == max_count] \n",
    "    return pattern_return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#match k-mer patterns in DNA with upto d-mismatches with reverse complements\n",
    "\n",
    "def count_hamd_reverse(text, pattern, mismatch):\n",
    "    results = 0\n",
    "    sub_len = len(pattern)\n",
    "    for i in range(len(text) - sub_len + 1):\n",
    "        hamd = hammingd(text[i:i+sub_len],pattern)\n",
    "        hamd_r = hammingd(text[i:i+sub_len],reverse_complement(pattern))\n",
    "        if hamd <= mismatch:\n",
    "            results += 1\n",
    "        if hamd_r <= mismatch:\n",
    "            results += 1\n",
    "    return results\n",
    "\n",
    "def frequency_hamd_reverse(text, k, d):\n",
    "    frequent_patterns = []\n",
    "    pattern_combinations = []\n",
    "    final_patterns = []\n",
    "    for i in range(len(text)-k + 1):\n",
    "        pat = text[i:i+k]\n",
    "        pattern_combinations.append(pat)\n",
    "        pattern_combinations.append(reverse_complement(pat))\n",
    "        temp = list(generate(pat, d))\n",
    "        for p in temp:\n",
    "            pattern_combinations.append(p)\n",
    "            pattern_combinations.append(reverse_complement(p))\n",
    "    pattern_combinations = list(set(pattern_combinations))\n",
    "    for t in pattern_combinations:\n",
    "        count = count_hamd_reverse(text, t, d)\n",
    "        frequent_patterns.append((t,count))     \n",
    "    frequent_patterns = list(set(frequent_patterns))\n",
    "    frequent_patterns.sort(key=lambda tup: tup[1], reverse=True)\n",
    "    max_count = max(frequent_patterns,key=itemgetter(1))[1] \n",
    "    pattern_return = [s[0] for s in frequent_patterns if s[1] == max_count] \n",
    "    return pattern_return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# given collection of strings Dna and an integer d, a k-mer is a (k,d)-motif if it appears in every string from Dna with at most d mismatches\n",
    "\n",
    "def motifenumeration(DNA_list, k, d):\n",
    "    n = len(DNA_list)\n",
    "    new_list = []\n",
    "    final_pattern = []\n",
    "    text = DNA_list[0]\n",
    "    for i in range(len(text) - k + 1):\n",
    "        pattern = text[i:i+k]\n",
    "        for i in range(d+1):\n",
    "            pattern_list = list(generate(pattern, i))\n",
    "            for p in pattern_list:\n",
    "                new_list.append(p)\n",
    "        for pat in new_list:\n",
    "            final_c = 0\n",
    "            for dna in DNA_list[1:]:\n",
    "                count = count_hamd(dna, pat, d)\n",
    "                if count > 0:\n",
    "                    final_c += 1\n",
    "            if final_c == n - 1:\n",
    "                final_pattern.append(pat)\n",
    "    final_pattern = list(set(final_pattern))   \n",
    "    return final_pattern "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Median string finding function\n",
    "import sys\n",
    "\n",
    "switcher = {\n",
    "    \"A\": 0,\n",
    "    \"C\": 1,\n",
    "    \"G\": 2,\n",
    "    \"T\": 3\n",
    "}\n",
    "\n",
    "def __number_to_symbol(val):\n",
    "    for symbol, value in switcher.items():\n",
    "        if(value == val):\n",
    "            return symbol\n",
    "\n",
    "def number_to_pattern(index, k):\n",
    "    if(k == 1):\n",
    "        return __number_to_symbol(index)\n",
    "    \n",
    "    reminder = index % 4\n",
    "    index = index // 4\n",
    "\n",
    "    return number_to_pattern(index, k-1) + __number_to_symbol(reminder)\n",
    "\n",
    "def median_string(dna, k):\n",
    "    distance = sys.maxsize\n",
    "    median = []\n",
    "    for i in range(4**k -1):\n",
    "        pattern = number_to_pattern(i, k)\n",
    "        interm_distance = __distance_between_pattern_and_strings(pattern, dna)\n",
    "        if distance >= interm_distance:\n",
    "            distance = interm_distance\n",
    "            median.append((pattern, distance))\n",
    "\n",
    "    return median\n",
    "        \n",
    "def hammingd(text1, text2):\n",
    "    ham_dis = 0\n",
    "    for i in range(len(text1)):\n",
    "        if text1[i] != text2[i]:\n",
    "            ham_dis += 1\n",
    "    return ham_dis\n",
    "    \n",
    "def __distance_between_pattern_and_strings(pattern, dna):\n",
    "    \n",
    "    distance = 0\n",
    "    for str in dna:\n",
    "        hamming_distance = sys.maxsize\n",
    "        for i in range(len(str) - len(pattern) + 1):\n",
    "            interm_distance = hammingd(pattern, str[i:i+len(pattern)])\n",
    "            if hamming_distance > interm_distance:\n",
    "                hamming_distance = interm_distance\n",
    "\n",
    "        distance += hamming_distance\n",
    "    \n",
    "    return distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Profile-most probable k-mer in Text\n",
    "def most_probable_motif(text, k, profile_matrix):\n",
    "    max_prob = 0\n",
    "    most_prop_kmer = \"\"\n",
    "    for i in range(len(text) - k + 1):\n",
    "        cur_prop = 1\n",
    "        kmer = text[i:i+k]\n",
    "        for i, c in enumerate(kmer):\n",
    "            cur_prop *= profile_matrix[c][i]\n",
    "        if max_prob < cur_prop:\n",
    "            max_prob = cur_prop\n",
    "            most_prop_kmer = kmer\n",
    "    \n",
    "    if most_prop_kmer == \"\":\n",
    "        most_prop_kmer = text[:k]\n",
    "\n",
    "    return most_prop_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# greedy-search motif\n",
    "from collections import Counter\n",
    "def generate_profile_matrix(kmers, count):\n",
    "    profile_matrix = {}\n",
    "    matrix = {\n",
    "      \"A\": [],\n",
    "      \"C\": [],\n",
    "      \"G\": [],\n",
    "      \"T\": []\n",
    "    } \n",
    "    for i, c in enumerate(kmers[0]):\n",
    "        for letter in [\"A\", \"C\", \"G\", \"T\"]:\n",
    "            matrix[letter].append((letter == c) if float(1) else float(0))\n",
    "    \n",
    "    for i in range(1, len(kmers)):\n",
    "        for i, c in enumerate(kmers[i]):\n",
    "            matrix[c][i] += 1\n",
    "\n",
    "    #isZeroPresent = False\n",
    "    #for key, value in matrix.items():\n",
    "    #    for i, v in enumerate(value):\n",
    "    #        if matrix[key][i] == 0:\n",
    "    #            isZeroPresent = True\n",
    "\n",
    "    #if isZeroPresent:\n",
    "    #    count += 4\n",
    "    #    for key, value in matrix.items():\n",
    "    #        for i, v in enumerate(value):\n",
    "    #            matrix[key][i] += 1\n",
    "\n",
    "    for key, value in matrix.items():\n",
    "        for i, v in enumerate(value):\n",
    "            matrix[key][i] = round(v/count, 4)\n",
    "\n",
    "    for i in range(len(kmers[0])):\n",
    "        c = 0\n",
    "        for key, value in matrix.items():\n",
    "            c += matrix[key][i]\n",
    "        assert abs(c - 1) < 1e-3\n",
    "\n",
    "    return matrix\n",
    "\n",
    "def __score_motifs(motifs, k):\n",
    "    score = 0    \n",
    "    for i in range(k):\n",
    "        col = []\n",
    "        for motif in motifs: \n",
    "            col.append(motif[i]);\n",
    "        most_common_letter = Counter(col).most_common(1)       \n",
    "        score += sum(1 for v in col if v != most_common_letter[0][0])  \n",
    "    \n",
    "    return score\n",
    "\n",
    "def greedy_motif_search(sequences, k):\n",
    "    best_motifs = [seq[:k] for seq in sequences]\n",
    "    motifs = []\n",
    "\n",
    "    for i in range(len(sequences[0]) - k + 1):\n",
    "        kmer = sequences[0][i:i+k]\n",
    "        motifs.append(kmer)\n",
    "        profile_matrix = generate_profile_matrix(motifs, len(motifs))\n",
    "        \n",
    "        for j in range(1, len(sequences)):\n",
    "            most_prop_kmer_in_seq = most_probable_motif(sequences[j], k, profile_matrix)\n",
    "            motifs.append(most_prop_kmer_in_seq)\n",
    "            profile_matrix = generate_profile_matrix(motifs, len(motifs))\n",
    "        \n",
    "        if __score_motifs(motifs, k) < __score_motifs(best_motifs, k):\n",
    "            best_motifs = motifs  \n",
    "        motifs = []\n",
    "    return best_motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Profile_Pseudocounts(Motifs,pseudocount):\n",
    "    count=Count(Motifs)\n",
    "    profile=count\n",
    "    for i in range(4):\n",
    "        for j in range(len(Motifs[0])):\n",
    "            profile[i][j]=(count[i][j]+pseudocount)/(len(Motifs)+pseudocount*4)\n",
    "    return profile\n",
    "\n",
    "def Probability(Pattern, Profile):\n",
    "    set = {'A': 0, 'C': 1, 'G': 2, 'T': 3}\n",
    "    probability=1\n",
    "    k=len(Pattern)\n",
    "    for j in range(k):\n",
    "        probability = probability * Profile[set[Pattern[j]]][j]\n",
    "    return probability\n",
    "\n",
    "def ProfileMostProbableKmer(Text, k, profile):\n",
    "    prob=[]\n",
    "    Kmer=[]\n",
    "    for i in range(len(Text)-k+1):\n",
    "        pattern=Text[i:i+k]\n",
    "        probability=Probability(pattern,profile)\n",
    "        prob.append(probability)\n",
    "    maxProb=max(prob)\n",
    "    for i in range(len(Text)-k+1):\n",
    "        if prob[i]==maxProb:\n",
    "            Kmer=Text[i:i+k]\n",
    "            break\n",
    "    return Kmer\n",
    "\n",
    "def Count(Motifs):\n",
    "    count=[]\n",
    "    if len(Motifs) != 1:\n",
    "        for i in range(len(Motifs[0])):\n",
    "            A = 0;C = 0;G = 0; T = 0\n",
    "            for j in range(len(Motifs)):\n",
    "                if Motifs[j][i] == 'A': A += 1\n",
    "                if Motifs[j][i] == 'C': C += 1\n",
    "                if Motifs[j][i] == 'G': G += 1\n",
    "                if Motifs[j][i] == 'T': T += 1\n",
    "            count.append([A, C, G, T])\n",
    "        count=[list(x) for x in zip(*count)]\n",
    "    else:\n",
    "        for j in range(len(Motifs[0])):\n",
    "            A = 0;C = 0;G = 0; T = 0\n",
    "            if Motifs[0][j] == 'A': A += 1\n",
    "            if Motifs[0][j] == 'C': C += 1\n",
    "            if Motifs[0][j] == 'G': G += 1\n",
    "            if Motifs[0][j] == 'T': T += 1\n",
    "            count.append([A, C, G, T])\n",
    "        count = [list(x) for x in zip(*count)]\n",
    "    return count\n",
    "\n",
    "def Score(Motifs):\n",
    "    consensus=Consensus(Motifs)\n",
    "    score=0\n",
    "    for i in range(len(Motifs)):\n",
    "        score+=hammingd(consensus,Motifs[i])\n",
    "    return score\n",
    "\n",
    "def Consensus(Motifs):\n",
    "    set = {0:'A', 1:'C', 2:'G', 3:'T'}\n",
    "    consensus=''\n",
    "    if len(Motifs[0])!=1:\n",
    "        for i in range(len(Motifs[0])):\n",
    "            A = 0; C = 0; G = 0; T = 0\n",
    "            for j in range(len(Motifs)):\n",
    "                if Motifs[j][i]=='A': A+=1\n",
    "                if Motifs[j][i]=='C': C+=1\n",
    "                if Motifs[j][i]=='G': G+=1\n",
    "                if Motifs[j][i]=='T': T+=1\n",
    "            S=[A,C,G,T]\n",
    "            Max=max(S)\n",
    "            for n in range(4):\n",
    "                if S[n]==Max:\n",
    "                    consensus=consensus+set[n]\n",
    "                    break\n",
    "    else: print('Motifs has one string')\n",
    "    return consensus\n",
    "\n",
    "def GreedyMotifSearchWithPseudocounts(Dna, k, t, pseudocount):\n",
    "    BestMotifs = []\n",
    "    for i in range(len(Dna)):\n",
    "        BestMotifs.append(Dna[i][:k])\n",
    "    for i in range(len(Dna[0]) - k + 1):\n",
    "        Motif1 = Dna[0][i:i + k]\n",
    "        Motifs = []\n",
    "        Motifs.append(Motif1)\n",
    "        for j in range(1, t):\n",
    "            Motifi = ProfileMostProbableKmer(Dna[j], k, Profile_Pseudocounts(Motifs,pseudocount))\n",
    "            Motifs.append(Motifi)\n",
    "        if Score(Motifs) < Score(BestMotifs):\n",
    "            BestMotifs = Motifs\n",
    "    return BestMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Random Motif search \n",
    "import random\n",
    "\n",
    "N = 1000\n",
    "def randomized_motif_search(Dna, k, t):\n",
    "    M = random_motifs(Dna, k, t)\n",
    "    bestMotifs = M\n",
    "    while True:\n",
    "        profile = profile_with_pseudocounts(M)\n",
    "        M = _motifs(profile, Dna)\n",
    "        if _score(M) < _score(bestMotifs):\n",
    "            bestMotifs = M\n",
    "        else:\n",
    "            return bestMotifs\n",
    "\n",
    "###############################################MAINCODE#########################################################################\n",
    "\n",
    "###################################################SubCodes#####################################################################\n",
    "def random_motifs(Dna, k, t):\n",
    "    randMotifs = []\n",
    "\n",
    "    for i in range(t):\n",
    "        x = random.randint(0, t)\n",
    "        randMotifs.append(Dna[i][x:x+k])\n",
    "\n",
    "    return randMotifs\n",
    "\n",
    "def _motifs(profile, Dna):\n",
    "    motifs = []\n",
    "    t = len(Dna)\n",
    "    k = len(profile['A'])\n",
    "\n",
    "    for i in range(t):\n",
    "        motifs.append(profile_most_probable_kmer(Dna[i], k, profile))\n",
    "\n",
    "    return motifs\n",
    "\n",
    "def profile_most_probable_kmer(text, k, profile):\n",
    "    mostProbVal = -1\n",
    "    mostProbKmer = ''\n",
    "\n",
    "    for i in range(0, 1 + len(text) - k):\n",
    "        kmer = text[i:i+k]\n",
    "        probKmerVal = _pr(kmer, profile)\n",
    "        if probKmerVal > mostProbVal:\n",
    "            mostProbVal = probKmerVal\n",
    "            mostProbKmer = kmer\n",
    "\n",
    "    return mostProbKmer\n",
    "\n",
    "def _pr(text, profile):\n",
    "    P = 1\n",
    "\n",
    "    for i in range(len(text)):\n",
    "        P = P * profile[text[i]][i]\n",
    "\n",
    "    return P\n",
    "\n",
    "def profile_with_pseudocounts(motifs):\n",
    "    profile = {}\n",
    "    t = len(motifs)\n",
    "    k = len(motifs[0])\n",
    "    countMotifs = count_with_pseudocounts(motifs)\n",
    "\n",
    "    for symbol in \"ACGT\":\n",
    "        profile[symbol] = []\n",
    "\n",
    "    for x in countMotifs:\n",
    "        for y in countMotifs[x]:\n",
    "            z = y/float(t+4)\n",
    "            profile[x].append(z)\n",
    "\n",
    "    return profile\n",
    "\n",
    "def count_with_pseudocounts(motifs):\n",
    "    count = {}\n",
    "    pseudocounts = {}\n",
    "    t = len(motifs)\n",
    "    k = len(motifs[0])\n",
    "\n",
    "    for symbol in \"GACT\":\n",
    "        count[symbol] = []\n",
    "        for j in range(k):\n",
    "            count[symbol].append(0)\n",
    "\n",
    "    for i in range(t):\n",
    "        for j in range(k):\n",
    "            symbol = motifs[i][j]\n",
    "            count[symbol][j] += 1\n",
    "\n",
    "    for symbol in \"GACT\":\n",
    "        pseudocounts[symbol] = []\n",
    "\n",
    "    for x in count:\n",
    "        for y in count[x]:\n",
    "            z = y + 1\n",
    "            pseudocounts[x].append(z)\n",
    "\n",
    "    return pseudocounts\n",
    "\n",
    "def _score(motifs):\n",
    "    count = 0\n",
    "    k = len(motifs[0])\n",
    "    t = len(motifs)\n",
    "    consensusMotif = _consensus(motifs)\n",
    "\n",
    "    for i in range(t):\n",
    "        for j in range(k):\n",
    "            if motifs[i][j] != consensusMotif[j]:\n",
    "                count += 1\n",
    "\n",
    "    return count\n",
    "\n",
    "def _consensus(motifs):\n",
    "    k = len(motifs[0])\n",
    "    count = count_with_pseudocounts(motifs)\n",
    "    consensus = \"\"\n",
    "\n",
    "    for j in range(k):\n",
    "        M = 0\n",
    "        frequentSymbol = \"\"\n",
    "        for symbol in \"ACGT\":\n",
    "            if count[symbol][j] > M:\n",
    "                M = count[symbol][j]\n",
    "                frequentSymbol = symbol\n",
    "        consensus += frequentSymbol\n",
    "\n",
    "    return consensus\n",
    "###################################################SubCodes#####################################################################\n",
    "randomized_motif_search(dna, 15, 20)\n",
    "M = randomized_motif_search(dna, 15, 20)\n",
    "bMotifs = M\n",
    "\n",
    "for i in range(N+1):\n",
    "    M = randomized_motif_search(dna, 15, 20)\n",
    "    if _score(M) < _score(bMotifs):\n",
    "         bMotifs = M\n",
    "    else:\n",
    "        bestMotifs = bMotifs\n",
    "\n",
    "print ('\\n'.join(bestMotifs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Gibb's sampler motif search\n",
    "from collections import Counter\n",
    "def ProfileRandomlyGeneratedKmer(Text,profile,k):\n",
    "    prob=[]\n",
    "    for i in range(len(Text)-k+1):\n",
    "        pattern=Text[i:i+k]\n",
    "        prob.append(Probability(pattern,profile))\n",
    "    Sum=sum(prob)\n",
    "    for i in range(len(Text)-k+1):\n",
    "        prob[i]=prob[i]/Sum\n",
    "    Range = [0]\n",
    "    for i in range(len(Text) - k+1):\n",
    "        Range.append(prob[i]+Range[i])\n",
    "    Range.append(1)\n",
    "    Random=random.uniform(0,1)\n",
    "    RandomPattern = ''\n",
    "    for i in range(len(Text) - k+1):\n",
    "        if Random>=Range[i] and Random<=Range[i+1]:\n",
    "            RandomPattern=Text[i:i+k]\n",
    "    return RandomPattern\n",
    "\n",
    "def Profile_Pseudocounts(Motifs,pseudocount):\n",
    "    count=Count(Motifs)\n",
    "    profile=count\n",
    "    for i in range(4):\n",
    "        for j in range(len(Motifs[0])):\n",
    "            profile[i][j]=(count[i][j]+pseudocount)/(len(Motifs)+pseudocount*4)\n",
    "    return profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def GibbsSampler(Dna, k, t, N):\n",
    "    motifs = []\n",
    "    for i in range(t):\n",
    "        x = random.randint(0, len(Dna[0]) - k)\n",
    "        motifs.append(Dna[i][x:x + k])\n",
    "    BestMotifs = motifs.copy()\n",
    "    for j in range(N):\n",
    "        i=random.randint(0,t-1)\n",
    "        motifs_i=motifs[0:i]+motifs[i+1:]\n",
    "        Profile = Profile_Pseudocounts(motifs_i, 1)\n",
    "        motifs[i]=ProfileRandomlyGeneratedKmer(Dna[i],Profile,k)\n",
    "        if Score(motifs) < Score(BestMotifs):\n",
    "            BestMotifs=motifs.copy()\n",
    "    return BestMotifs\n",
    "###################################################################\n",
    "def GibbsSampler_nTime(Dna, k, t, N, repeat):\n",
    "    score = []\n",
    "    motifs = []\n",
    "    for i in range(repeat):\n",
    "        m = GibbsSampler(Dna,k,t,N)\n",
    "        score.append(Score(m))\n",
    "        motifs.append(m)\n",
    "    Min = min(score)\n",
    "    Finalmotifs = []\n",
    "    for i in range(repeat):\n",
    "        if score[i] == Min:\n",
    "            Finalmotifs = motifs[i]\n",
    "    return Finalmotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#GibbsSampler_nTime(dna_list, 15, 20, 2000, 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
