{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "In this chapter, we define the length of a path in a tree as the sum of the lengths of its edges (rather than the number of edges \n",
    "on the path). As a result, the evolutionary distance between two present-day species corresponding to leaves i and j in a tree T is \n",
    "equal to the length of the unique path connecting i and j, denoted di,j(T).\n",
    "Distances Between Leaves Problem: Compute the distances between leaves in a weighted tree.\n",
    "     Input:  An integer n followed by the adjacency list of a weighted tree with n leaves.\n",
    "     Output: An n x n matrix (di,j), where di,j is the length of the path between leaves i and j.\n",
    "Code Challenge: Solve the Distances Between Leaves Problem. The tree is given as an adjacency list of a graph whose leaves are \n",
    "integers between 0 and n - 1; the notation a->b:c means that node a is connected to node b by an edge of weight c. The matrix \n",
    "you return should be space-separated.\n",
    "Sample Input:\n",
    "4\n",
    "0->4:11\n",
    "1->4:2\n",
    "2->5:6\n",
    "3->5:7\n",
    "4->0:11\n",
    "4->1:2\n",
    "4->5:4\n",
    "5->4:4\n",
    "5->3:7\n",
    "5->2:6\n",
    "Sample Output:\n",
    "0 13 21 22\n",
    "13 0 12 13\n",
    "21 12 0 13\n",
    "22 13 13 0\n",
    "\n",
    "filename = trial.txt-\n",
    "4\n",
    "0->4:11\n",
    "1->4:2\n",
    "2->5:6\n",
    "3->5:7\n",
    "4->0:11\n",
    "4->1:2\n",
    "4->5:4\n",
    "5->4:4\n",
    "5->3:7\n",
    "5->2:6\n",
    "'''\n",
    "\n",
    "from collections import defaultdict, deque\n",
    "with open(filename) as file:\n",
    "    data = []\n",
    "    for line in file:\n",
    "        data.append(line[:-1])\n",
    "        \n",
    "n = int(data[0])\n",
    "\n",
    "from_node = []\n",
    "to_node = []\n",
    "weight = []\n",
    "\n",
    "for line in data[1:]:\n",
    "    x = line.split('->')\n",
    "    from_node.append(int(x[0]))\n",
    "    x = x[1].split(':')\n",
    "    to_node.append(int(x[0]))\n",
    "    weight.append(int(x[1]))\n",
    "    \n",
    "#print from_node,to_node,weight\n",
    "\n",
    "nodes = list(set(from_node))\n",
    "\n",
    "class Graph(object):\n",
    "    def __init__(self):\n",
    "        self.nodes = set()\n",
    "        self.edges = defaultdict(list)\n",
    "        self.distances = {}\n",
    "    def add_node(self, value):\n",
    "        self.nodes.add(value)\n",
    "\n",
    "    def add_edge(self, from_node, to_node, distance):\n",
    "        self.edges[from_node].append(to_node)\n",
    "        self.edges[to_node].append(from_node)\n",
    "        self.distances[(from_node, to_node)] = distance\n",
    "\n",
    "\n",
    "def dijkstra(graph, initial):\n",
    "    visited = {initial: 0}\n",
    "    path = {}\n",
    "\n",
    "    nodes = set(graph.nodes)\n",
    "\n",
    "    while nodes:\n",
    "        min_node = None\n",
    "        for node in nodes:\n",
    "            if node in visited:\n",
    "                if min_node is None:\n",
    "                    min_node = node\n",
    "                elif visited[node] < visited[min_node]:\n",
    "                    min_node = node\n",
    "        if min_node is None:\n",
    "            break\n",
    "\n",
    "        nodes.remove(min_node)\n",
    "        current_weight = visited[min_node]\n",
    "\n",
    "        for edge in graph.edges[min_node]:\n",
    "            try:\n",
    "                weight = current_weight + graph.distances[(min_node, edge)]\n",
    "            except:\n",
    "                continue\n",
    "            if edge not in visited or weight < visited[edge]:\n",
    "                visited[edge] = weight\n",
    "                path[edge] = min_node\n",
    "\n",
    "    return visited, path\n",
    "\n",
    "\n",
    "def shortest_path(graph, origin, destination):\n",
    "\n",
    "    if origin == destination:\n",
    "        return 0\n",
    "    else:\n",
    "        visited, paths = dijkstra(graph, origin)\n",
    "        full_path = deque()\n",
    "        _destination = paths[destination]\n",
    "\n",
    "        while _destination != origin:\n",
    "            full_path.appendleft(_destination)\n",
    "            _destination = paths[_destination]\n",
    "\n",
    "        full_path.appendleft(origin)\n",
    "        full_path.append(destination)\n",
    "\n",
    "        return visited[destination]#, list(full_path)\n",
    "\n",
    "\n",
    "graph = Graph()\n",
    "\n",
    "for node in nodes:\n",
    "    graph.add_node(node)\n",
    "\n",
    "for i in range(len(from_node)):\n",
    "    graph.add_edge(from_node[i],to_node[i],weight[i])\n",
    "\n",
    "leaf_nodes = []\n",
    "for i in graph.edges:\n",
    "    if len(graph.edges[i])==2:\n",
    "        leaf_nodes.append(i)\n",
    "\n",
    "\n",
    "for i in leaf_nodes:\n",
    "    final_path = []\n",
    "    for j in leaf_nodes:\n",
    "        final_path.append(str(shortest_path(graph,i,j)))\n",
    "    print(' '.join(final_path)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "We now have an algorithm for solving the Limb Length Problem. For each j, we can compute LimbLength(j) by finding the minimum value \n",
    "of (Di,j+Dj,k−Di,k)/2(Di,j+Dj,k−Di,k)/2 over all pairs of leaves i and k.\n",
    "Code Challenge: Solve the Limb Length Problem.\n",
    "     Input: An integer n, followed by an integer j between 0 and n - 1, followed by a space-separated\n",
    "     additive distance matrix D (whose elements are integers).\n",
    "     Output: The limb length of the leaf in Tree(D) corresponding to row j of this distance\n",
    "     matrix (use 0-based indexing).\n",
    "Sample Input:\n",
    "4\n",
    "1\n",
    "0 13 21 22\n",
    "13 0 12 13\n",
    "21 12 0 13\n",
    "22 13 13 0\n",
    "Sample Output:\n",
    "2\n",
    "'''\n",
    "\n",
    "from itertools import permutations\n",
    "\n",
    "def limbLength(mat,n,j):\n",
    "    leaves = range(n)\n",
    "    leavesLeft = [l for l in leaves if l != j]\n",
    "    minlen = float(\"inf\")\n",
    "    for i,k in permutations(leavesLeft,2):\n",
    "        if i < k:\n",
    "            lenik = (mat[j][i]+mat[j][k]-mat[i][k]) / 2\n",
    "            if lenik < minlen:\n",
    "                minlen = lenik\n",
    "    return minlen\n",
    "\n",
    "mat = [[0,13,21,22],[13,0,12,13],[21,12,0,13],[22,13,13,0]]\n",
    "n = 4\n",
    "j = 1\n",
    "\n",
    "print(limbLength(mat,n,j))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Implement AdditivePhylogeny to solve the Distance-Based Phylogeny Problem.\n",
    "     Input: An integer n followed by a space-separated n x n distance matrix.\n",
    "     Output: A weighted adjacency list for the simple tree fitting this matrix.\n",
    "Note on formatting: The adjacency list must have consecutive integer node labels starting from 0. The n leaves must be labeled 0, 1, \n",
    "..., n - 1 in order of their appearance in the distance matrix. Labels for internal nodes may be labeled in any order but must start \n",
    "from n and increase consecutively.\n",
    "Sample Input:\n",
    "4\n",
    "0 13 21 22\n",
    "13 0 12 13\n",
    "21 12 0 13\n",
    "22 13 13 0\n",
    "Sample Output:\n",
    "0->4:11\n",
    "1->4:2\n",
    "2->5:6\n",
    "3->5:7\n",
    "4->0:11\n",
    "4->1:2\n",
    "4->5:4\n",
    "5->4:4\n",
    "5->3:7\n",
    "5->2:6\n",
    "AdditivePhylogeny(D, n)\n",
    "    if n = 2\n",
    "        return the tree consisting of a single edge of length D1,2\n",
    "    limbLength ← Limb(D, n)\n",
    "    for j ← 1 to n - 1\n",
    "        Dj,n ← Dj,n - limbLength\n",
    "        Dn,j ← Dj,n\n",
    "    (i, k) ← two leaves such that Di,k = Di,n + Dn,k\n",
    "    x ← Di,n\n",
    "    remove row n and column n from D\n",
    "    T ← AdditivePhylogeny(D, n - 1)\n",
    "    v ← the (potentially new) node in T at distance x from i on the path between i and k\n",
    "    add leaf n back to T by creating a limb (v, n) of length limbLength\n",
    "    return T\n",
    "'''\n",
    "from itertools import permutations\n",
    "from collections import defaultdict\n",
    "'''refer to the additive_phylogency's idea about add branch and I need to master this concept.'''\n",
    "\n",
    "def additive_phylogence(D,n):\n",
    "    global Tree\n",
    "    global nodesnum\n",
    "    if n == 2:\n",
    "        Tree = defaultdict(list)\n",
    "        Tree[0].append((1,D[0][1]))\n",
    "        Tree[1].append((0,D[0][1]))\n",
    "        return\n",
    "    limblen = limbLength(D,n-1)\n",
    "    '''calculate D_bald to find out (i,L,k)'''\n",
    "    for j in range(n-1):\n",
    "        D[j][n-1] = D[j][n-1]-limblen\n",
    "        D[n-1][j] = D[j][n-1]\n",
    "    ik = [l for l in range(len(D[n-1][:-1])) if D[n-1][l]!=0]\n",
    "    for i,k in permutations(ik,2):\n",
    "        if i < k and D[i][n-1] + D[k][n-1] == D[i][k]:\n",
    "            x = D[i][n-1]\n",
    "            ret = ((i,x),(n-1,limblen),(k,D[i][k]-x))\n",
    "            break\n",
    "    '''calculate D_trimmed'''\n",
    "    D = D[0:n-1][0:n-1]\n",
    "    '''recursively'''\n",
    "    additive_phylogence(D,n-1)\n",
    "    '''FORMAT: ret = ((i,x),(l,limblen),(k,D[i][k]-x))'''\n",
    "    pathiTok = pathBetweenLeaves(Tree,ret[0][0],ret[2][0])\n",
    "    x = ret[0][1]\n",
    "    l = ret[1][0]\n",
    "    limblen = ret[1][1]\n",
    "    start_node = ret[0][0]\n",
    "    for node, dist in pathiTok:\n",
    "        if dist ==0:\n",
    "            continue\n",
    "        if x > dist:\n",
    "            x -= dist\n",
    "            start_node = node\n",
    "        elif x == dist:\n",
    "            Tree[node].append((l,limblen))\n",
    "            Tree[l].append((node,limblen))\n",
    "            break\n",
    "        elif x < dist:\n",
    "            '''this part is reallly good'''\n",
    "            new_node = nodesnum\n",
    "            nodesnum += 1\n",
    "            Tree[new_node].append((l,limblen))\n",
    "            Tree[l].append((new_node,limblen))\n",
    "            Tree[start_node].remove((node,dist))\n",
    "            Tree[node].remove((start_node,dist))\n",
    "            Tree[start_node].append((new_node,x))\n",
    "            Tree[new_node].append((start_node,x))\n",
    "            Tree[node].append((new_node,dist-x))\n",
    "            Tree[new_node].append((node,dist-x))\n",
    "            break\n",
    "\n",
    "def pathBetweenLeaves(graph,i,k):\n",
    "    tree = defaultdict(list)\n",
    "    weights = defaultdict(list)\n",
    "    for nodei,childList in graph.items():\n",
    "        for j in range(len(childList)):\n",
    "            tree[nodei].append(childList[j][0])\n",
    "            weights[nodei].append(childList[j][1])\n",
    "    edgeTo, marked = depthFirstPaths(tree,i)\n",
    "    path = pathTo(tree,weights,k,i,edgeTo)\n",
    "    return path\n",
    "\n",
    "def depthFirstPaths(tree,s):\n",
    "    '''edgeTo and marked all save the idnex of nodes'''\n",
    "    edgeTo = dict((key,None) for key in tree.keys())\n",
    "    marked = dict((key,False) for key in tree.keys())\n",
    "    edgeTo, marked = dfs(tree,s,edgeTo,marked)\n",
    "    return edgeTo, marked\n",
    "\n",
    "def dfs(tree,v,edgeTo,marked):\n",
    "    marked[v] = True\n",
    "    for w in tree[v]:\n",
    "        if not marked[w]:\n",
    "            edgeTo[w] = v\n",
    "            edgeTo,marked = dfs(tree,w,edgeTo,marked)\n",
    "    return edgeTo, marked\n",
    "\n",
    "def pathTo(tree,weights,v,s,edgeTo):\n",
    "    stack = []\n",
    "    w = edgeTo[v]\n",
    "    stack.append((v,weights[w][tree[w].index(v)]))\n",
    "    while w != s:\n",
    "        v = w\n",
    "        w = edgeTo[v]\n",
    "        stack.append((v,weights[w][tree[w].index(v)]))\n",
    "    stack.append((s,0))\n",
    "    return stack[::-1]\n",
    "\n",
    "def limbLength(mat,j):\n",
    "    n = len(mat)\n",
    "    leaves = range(n)\n",
    "    leavesLeft = [l for l in leaves if l != j]\n",
    "    minlen = float(\"inf\")\n",
    "    for i,k in permutations(leavesLeft,2):\n",
    "        if i < k:\n",
    "            lenik = (mat[j][i]+mat[j][k]-mat[i][k]) / 2\n",
    "            if lenik < minlen:\n",
    "                minlen = lenik\n",
    "    return minlen\n",
    "\n",
    "D = [[0,13,21,22],[13,0,12,13],[21,12,0,13],[22,13,13,0]]\n",
    "n = 4\n",
    "global Tree\n",
    "global nodesnum\n",
    "nodesnum = n\n",
    "additive_phylogence(D,n)\n",
    "for key,vals in Tree.items():\n",
    "    for val in vals:\n",
    "        print(str(key)+'->'+str(val[0])+':'+str(int(val[1])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "If we had a molecular clock measuring evolutionary time, then we could assign an age to every node v in a rooted binary tree \n",
    "(denoted Age(v)), where all of the leaves of the tree have age 0 because they correspond to present-day species. We could then \n",
    "define the weight of an edge (v, w) in the tree as the difference Age(v) - Age(w). Consequently, the length of a path between \n",
    "the root and any node would be equal to the difference between their ages. Such a tree, in which the distance from the root to \n",
    "any leaf is the same, is called ultrametric.\n",
    "Our aim is to derive an ultrametric tree that explains a given distance matrix (even if it does so only approximately). UPGMA \n",
    "(which stands for Unweighted Pair Group Method with Arithmetic Mean) is a simple clustering heuristic that introduces a hypothetical \n",
    "molecular clock for constructing an ultrametric evolutionary tree.\n",
    "Given an n × n matrix D, UPGMA (which is illustrated in the figure on the next step) first forms n trivial clusters, each containing \n",
    "a single leaf. The algorithm then finds a pair of “closest” clusters. To clarify the notion of closest clusters, UPGMA defines the \n",
    "distance between clusters C1 and C2 as the average pairwise distance between elements of C1 and C2.\n",
    "Once UPGMA has identified a pair of closest clusters C1 and C2, it merges them into a cluster C with |C1| + |C2| elements and then \n",
    "creates a node for C, which it connects to each of C1 and C2 by a directed edge. The age of C is set to be DC1, C2 /2. UPGMA then \n",
    "iterates this process of merging the two closest clusters until only a single cluster remains, which corresponds to the root.\n",
    "Implement UPGMA.\n",
    "     Input: An integer n followed by a space separated n x n distance matrix.\n",
    "     Output: An adjacency list for the ultrametric tree returned by UPGMA. Edge weights should be accurate to two decimal places\n",
    "     (answers in the sample dataset below are provided to three decimal places).\n",
    "Note on formatting: The adjacency list must have consecutive integer node labels starting from 0. The n leaves must be labeled \n",
    "0, 1, ..., n - 1 in order of their appearance in the distance matrix. Labels for internal nodes may be labeled in any order but \n",
    "must start from n and increase consecutively.\n",
    "Sample Input:\n",
    "4\n",
    "0 20 17 11\n",
    "20 0 20 13\n",
    "17 20 0 10\n",
    "11 13 10 0\n",
    "Sample Output:\n",
    "0->5:7.000\n",
    "1->6:8.833\n",
    "2->4:5.000\n",
    "3->4:5.000\n",
    "4->2:5.000\n",
    "4->3:5.000\n",
    "4->5:2.000\n",
    "5->0:7.000\n",
    "5->4:2.000\n",
    "5->6:1.833\n",
    "6->5:1.833\n",
    "6->1:8.833\n",
    "UPGMA(D, n) \n",
    "        Clusters ← n single-element clusters labeled 1, ... , n\n",
    "        construct a graph T with n isolated nodes labeled by single elements 1, ... , n\n",
    "    for every node v in T \n",
    "        Age(v) ← 0\n",
    "    while there is more than one cluster\n",
    "        find the two closest clusters Ci and Cj\n",
    "        merge Ci and Cj into a new cluster Cnew with |Ci| + |Cj| elements\n",
    "        add a new node labeled by cluster Cnew to T\n",
    "        connect node Cnew to Ci and Cj by directed edges \n",
    "        Age(Cnew) ← DCi, Cj / 2\n",
    "        remove the rows and columns of D corresponding to Ci and Cj\n",
    "        remove Ci and Cj from Clusters\n",
    "        add a row/column to D for Cnew by computing D(Cnew, C) for each C in Clusters\n",
    "        add Cnew to Clusters\n",
    "    root ← the node in T corresponding to the remaining cluster\n",
    "    for each edge (v, w) in T\n",
    "        length of (v, w) ← Age(v) - Age(w)\n",
    "    return T\n",
    "'''\n",
    "\n",
    "from __future__ import division\n",
    "import sys\n",
    "from collections import defaultdict\n",
    "\n",
    "''' as an UPGMA tree, the degree of internal nodes must be three SOGA'''\n",
    "'''20151030: it took me a while to realize that it is the number of leaves in a cluster, not the number of nodes'''\n",
    "def dfs(tree,v,edgeTo,marked,count):\n",
    "    count += 1\n",
    "    marked[v] = True\n",
    "    for w in tree[v]:\n",
    "        if not marked[w]:\n",
    "            edgeTo[w] = v\n",
    "            edgeTo,marked,count = dfs(tree,w,edgeTo,marked,count)\n",
    "    return edgeTo, marked,count\n",
    "\n",
    "def count(tree,s):\n",
    "    if len(tree[s]) == 0:\n",
    "        return 1\n",
    "    '''edgeTo and marked all save the index of nodes'''\n",
    "    edgeTo = dict((key,None) for key in tree.keys())\n",
    "    marked = dict((key,False) for key in tree.keys())\n",
    "    count = 0\n",
    "    edgeTo, marked,count = dfs(tree,s,edgeTo,marked,count)\n",
    "    return (count+1)/2#THE NUMBER OF LEAVES:) n + n - 1 = count\n",
    "\n",
    "def upgma1(D,nodes):\n",
    "    #create the tree first\n",
    "    T = {}\n",
    "    age = {}\n",
    "    for node in nodes:\n",
    "        T[node] = []\n",
    "        age[node] = 0\n",
    "\n",
    "    nodeNum = len(D)\n",
    "    while len(nodes)>1:\n",
    "        cv,ci,cj = findMin(D) #ci, cj are the index of D\n",
    "        newnode = nodeNum\n",
    "        nodeNum += 1\n",
    "\n",
    "        newRow = []\n",
    "        for m in range(len(D)):\n",
    "            if m!=ci and m!=cj:\n",
    "                cmi = D[m][ci]\n",
    "                cmj = D[m][cj]\n",
    "                numi = count(T,nodes[ci])\n",
    "                numj = count(T,nodes[cj])\n",
    "                newRow.append((cmi*numi+cmj*numj) / (numi+numj))\n",
    "        newRow.append(0)\n",
    "\n",
    "        T[nodes[ci]].append(newnode)\n",
    "        T[nodes[cj]].append(newnode)\n",
    "        T[newnode] = [nodes[ci], nodes[cj]]\n",
    "        age[newnode] = cv/2.0\n",
    "\n",
    "        Dnew = []\n",
    "        for i in range(len(D)):\n",
    "            if i != ci and i != cj:\n",
    "                Dnew.append([D[i][j] for j in range(len(D[i])) if j != ci and j != cj])\n",
    "        nodes = [nodes[i] for i in range(len(nodes)) if i !=ci and i != cj]\t\t\n",
    "        nodes.append(newnode)\n",
    "        Dnew.append(newRow)\n",
    "        for i in range(len(Dnew)-1):\n",
    "            Dnew[i].append(Dnew[-1][i])\n",
    "        D = Dnew\n",
    "\n",
    "    for v in T.keys():\n",
    "        T[v] = [ (w,abs(age[v]-age[w])) for w in T[v] ] \n",
    "    return T\n",
    "\n",
    "def findMin(D):\n",
    "    '''return the row and column index of the min value, not necessarily the node number'''\n",
    "    n = len(D)\n",
    "    imin = None\n",
    "    jmin = None\n",
    "    vmin = float(\"inf\")\n",
    "    for i in range(n):\n",
    "        rowmin = min([x for x in D[i] if x>0])\n",
    "        if rowmin < vmin:\n",
    "            imin = i\n",
    "            jmin = D[i].index(rowmin)\n",
    "            vmin = rowmin\n",
    "    return (vmin,imin,jmin)\n",
    "\n",
    "def printTree(T):\n",
    "    for v in T.keys():\n",
    "        for (w,edge) in T[v]:\n",
    "            print(str(v)+\"->\"+str(w)+\":\"+\"%.3f\"%(edge))\n",
    "            \n",
    "n = 4\n",
    "D =[[0,20,17,11],[20,0,20,13],[17,20,0,10],[11,13,10,0]]\n",
    "nodes = range(n)\n",
    "\n",
    "tree = upgma1(D,nodes)\n",
    "printTree(tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "In 1987, Naruya Saitou and Masatoshi Nei developed the neighbor-joining algorithm for evolutionary tree reconstruction. Given \n",
    "an additive distance matrix, this algorithm, which we call NeighborJoining, finds a pair of neighboring leaves and substitutes \n",
    "them by a single leaf, thus reducing the size of the tree. NeighborJoining can thus recursively construct a tree fitting the \n",
    "additive matrix. This algorithm also provides a heuristic for non-additive distance matrices that performs well in practice.\n",
    "The central idea of NeighborJoining is that although finding a minimum element in a distance matrix D is not guaranteed to \n",
    "yield a pair of neighbors in Tree(D), we can convert D into a different matrix whose minimum element does yield a pair of \n",
    "neighbors. First, given an n × n distance matrix D, we define TotalDistanceD(i) as the sum ∑1≤k≤n Di,k of distances from \n",
    "leaf i to all other leaves. The neighbor-joining matrix D* (see below) is defined such that for any i and j, D*i,i = 0 \n",
    "and D*i,j = (n - 2) · Di,j - TotalDistanceD(i) - TotalDistanceD(j).\n",
    "Implement NeighborJoining.\n",
    "     Input: An integer n, followed by an n x n distance matrix.\n",
    "     Output: An adjacency list for the tree resulting from applying the neighbor-joining algorithm. Edge-weights should be \n",
    "     accurate to two decimal places (they are provided to three decimal places in the sample output below).\n",
    "Note on formatting: The adjacency list must have consecutive integer node labels starting from 0. The n leaves must be \n",
    "labeled 0, 1, ..., n - 1 in order of their appearance in the distance matrix. Labels for internal nodes may be labeled \n",
    "in any order but must start from n and increase consecutively.\n",
    "Sample Input:\n",
    "4\n",
    "0 23 27 20\n",
    "23 0 30 28\n",
    "27 30 0 30\n",
    "20 28 30 0\n",
    "Sample Output:\n",
    "0->4:8.000\n",
    "1->5:13.500\n",
    "2->5:16.500\n",
    "3->4:12.000\n",
    "4->5:2.000\n",
    "4->0:8.000\n",
    "4->3:12.000\n",
    "5->1:13.500\n",
    "5->2:16.500\n",
    "5->4:2.000\n",
    "'''\n",
    "\n",
    "from __future__ import division\n",
    "import sys\n",
    "from collections import defaultdict\n",
    "#Chunyu Zhao 20151031\n",
    "\n",
    "def neighbor_joining(D,nodes):\n",
    "    global nodeNum\n",
    "    n = len(D)\n",
    "    if n == 2:\n",
    "        T = defaultdict(list)\n",
    "        T[nodes[0]].append((nodes[1],D[0][1]))\n",
    "        T[nodes[1]].append((nodes[0],D[1][0]))\n",
    "        return T\n",
    "    Dstar = getDstar(D)\n",
    "    ci,cj = findMin(Dstar)\n",
    "\n",
    "    newnode = nodeNum\n",
    "    nodeNum += 1\n",
    "    totalDist = [ sum(row) for row in D ]\n",
    "    delta = (totalDist[ci] - totalDist[cj])/ (n-2)\n",
    "    limbi = (D[ci][cj] + delta) / 2.0\n",
    "    limbj = (D[ci][cj] - delta) / 2.0\n",
    "    nodei = nodes[ci]\n",
    "    nodej = nodes[cj]\n",
    "\n",
    "    newRow = [(D[k][ci]+D[k][cj]-D[ci][cj])/2.0 for k in range(n) if k!=ci and k!=cj]\n",
    "    newRow.append(0)\n",
    "    Dnew = []\n",
    "    for i in range(n):\n",
    "        if i != ci and i != cj:\n",
    "            Dnew.append([D[i][j] for j in range(len(D[i])) if j != ci and j != cj])\n",
    "    nodes = [nodes[i] for i in range(len(nodes)) if i !=ci and i != cj]\t\t\n",
    "    nodes.append(newnode)\n",
    "    Dnew.append(newRow)\n",
    "    for i in range(n-2):\n",
    "        Dnew[i].append(Dnew[-1][i])\n",
    "    D = Dnew\n",
    "\n",
    "    ''' recursively '''\n",
    "    T = neighbor_joining(D,nodes)\n",
    "\n",
    "    T[nodei].append((newnode,limbi))\n",
    "    T[nodej].append((newnode,limbj))\n",
    "    T[newnode].append((nodei,limbi))\n",
    "    T[newnode].append((nodej,limbj))\n",
    "\n",
    "    return T\n",
    "\n",
    "def getDstar(D):\n",
    "    n = len(D)\n",
    "    totalDist = [ sum(row) for row in D ]\n",
    "    Dstar = []\n",
    "    for i in range(n):\n",
    "        Dstar.append([(n-2)*D[i][j]-totalDist[i]-totalDist[j] for j in range(n)])\n",
    "    for i in range(n):\n",
    "        for j in range(n):\n",
    "            if i==j:\n",
    "                Dstar[i][j] = 0\n",
    "    return Dstar\n",
    "\n",
    "def findMin(D):\n",
    "    '''return the row and column index of the min value, not necessarily the node number'''\n",
    "    n = len(D)\n",
    "    imin = None\n",
    "    jmin = None\n",
    "    vmin = float(\"inf\")\n",
    "    for i in range(n):\n",
    "        rowmin = min([x for x in D[i] if x != 0])\n",
    "        if rowmin < vmin:\n",
    "            imin = i\n",
    "            jmin = D[i].index(rowmin)\n",
    "            vmin = rowmin\n",
    "    return (imin,jmin)\n",
    "\n",
    "def printTree(T):\n",
    "    for v in T.keys():\n",
    "        for (w,edge) in T[v]:\n",
    "            print(str(v)+\"->\"+str(w)+\":\"+\"%.3f\"%(edge))\n",
    "            \n",
    "n = 4\n",
    "D =[[0,23,27,20],[23,0,30,28],[27,30,0,30],[20,28,30,0]]\n",
    "nodes = range(n)\n",
    "global nodeNum \n",
    "nodeNum = n\n",
    "T = neighbor_joining(D,nodes)\n",
    "printTree(T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Implement SmallParsimony to solve the Small Parsimony Problem.\n",
    "     Input: An integer n followed by an adjacency list for a rooted binary tree with n leaves labeled by DNA strings.\n",
    "     Output: The minimum parsimony score of this tree, followed by the adjacency list of the tree corresponding to labeling\n",
    "     internal nodes by DNA strings in order to minimize the parsimony score of the tree.\n",
    "Note: Remember to run SmallParsimony on each individual index of the strings at the leaves of the tree.\n",
    "Sample Input:\n",
    "4\n",
    "4->CAAATCCC\n",
    "4->ATTGCGAC\n",
    "5->CTGCGCTG\n",
    "5->ATGGACGA\n",
    "6->4\n",
    "6->5\n",
    "Sample Output:\n",
    "16\n",
    "ATTGCGAC->ATAGCCAC:2\n",
    "ATAGACAA->ATAGCCAC:2\n",
    "ATAGACAA->ATGGACTA:2\n",
    "ATGGACGA->ATGGACTA:1\n",
    "CTGCGCTG->ATGGACTA:4\n",
    "ATGGACTA->CTGCGCTG:4\n",
    "ATGGACTA->ATGGACGA:1\n",
    "ATGGACTA->ATAGACAA:2\n",
    "ATAGCCAC->CAAATCCC:5\n",
    "ATAGCCAC->ATTGCGAC:2\n",
    "ATAGCCAC->ATAGACAA:2\n",
    "CAAATCCC->ATAGCCAC:5\n",
    "SmallParsimony(T, Character)\n",
    "    for each node v in tree T\n",
    "        Tag(v) ← 0\n",
    "        if v is a leaf\n",
    "            Tag(v) ← 1\n",
    "            for each symbol k in the alphabet\n",
    "                if Character(v) = k\n",
    "                    sk(v) ← 0\n",
    "                else\n",
    "                    sk(v) ← ∞\n",
    "    while there exist ripe nodes in T\n",
    "        v ← a ripe node in T\n",
    "        Tag(v) ← 1\n",
    "        for each symbol k in the alphabet\n",
    "            sk(v) ← minimumall symbols i {si(Daughter(v))+δi,k} + minimumall symbols j {sj(Son(v))+δj,k}\n",
    "    return minimum over all symbols k {sk(v)}\n",
    "    \n",
    "trial.txt --\n",
    "\n",
    "4\n",
    "4->CAAATCCC\n",
    "4->ATTGCGAC\n",
    "5->CTGCGCTG\n",
    "5->ATGGACGA\n",
    "6->4\n",
    "6->5\n",
    "\n",
    "'''\n",
    "\n",
    "import sys\n",
    "from collections import defaultdict\n",
    "\n",
    "alphabet = ['A', 'C', 'G', 'T', ]\n",
    "\n",
    "\n",
    "def dist(a, b):\n",
    "    if a == b:\n",
    "        return 0\n",
    "    else:\n",
    "        return 1\n",
    "\n",
    "\n",
    "def seq_dist(a, b):\n",
    "    summ = 0\n",
    "    for i in range(len(a)):\n",
    "        if a[i] != b[i]:\n",
    "            summ += 1\n",
    "    return summ\n",
    "\n",
    "\n",
    "class Node(object):\n",
    "    def __init__(self, index, leaf):\n",
    "        self.index = index\n",
    "        self.leaf = leaf\n",
    "        self.children = []\n",
    "        self.edges = []\n",
    "        self.text = ''\n",
    "        self.score = defaultdict(int)\n",
    "        self.scored = False\n",
    "\n",
    "    def reinit(self):\n",
    "        self.score = defaultdict(int)\n",
    "        self.scored = False\n",
    "        return None\n",
    "\n",
    "    def is_ripe(self):\n",
    "        if self.leaf:\n",
    "            return True\n",
    "        else:\n",
    "            for item in self.children:\n",
    "                if not nodelist[item].scored:\n",
    "                    return False\n",
    "        return True\n",
    "\n",
    "    def add_node(self, c_index):\n",
    "        self.children.append(c_index)\n",
    "        self.edges.append(0)\n",
    "        return None\n",
    "\n",
    "    def scoring(self, curr_letter):\n",
    "        if self.leaf:\n",
    "            for letter in alphabet:\n",
    "                if letter == self.text[curr_letter]:\n",
    "                    self.score[letter] = 0\n",
    "                else:\n",
    "                    self.score[letter] = float('inf')\n",
    "        else:\n",
    "            for letter in alphabet:\n",
    "                for item in self.children:\n",
    "                    self.score[letter] += min([nodelist[item].score[i]+dist(letter, i) for i in alphabet])\n",
    "        self.scored = True\n",
    "        return None\n",
    "\n",
    "\n",
    "def interpreter(conn):\n",
    "    nn = int(conn.readline().strip())\n",
    "    nl = [Node(i, True) for i in range(nn)]\n",
    "    for j, raw_line in enumerate(conn):\n",
    "        line = raw_line.strip().split('->')\n",
    "        node_val = int(line[0])\n",
    "        if node_val == len(nl):\n",
    "            nl.append(Node(node_val, False))\n",
    "        if node_val > len(nl):\n",
    "            print(\"Unexpected node\")\n",
    "        if j < nn:\n",
    "            nl[node_val].add_node(j)\n",
    "            nl[j].text = line[1]\n",
    "        if j >= nn:\n",
    "            nl[node_val].add_node(int(line[1]))\n",
    "    return nl\n",
    "\n",
    "\n",
    "def tree_scoring(curr_letter):\n",
    "    for node in nodelist:\n",
    "        if node.leaf:\n",
    "            node.scoring(curr_letter)\n",
    "    while not nodelist[-1].scored:\n",
    "        for node in nodelist:\n",
    "            if not node.scored:\n",
    "                if node.is_ripe():\n",
    "                    node.scoring(curr_letter)\n",
    "    return None\n",
    "\n",
    "\n",
    "def tree_pruning(node, parent_letter=''):\n",
    "    if node.leaf:\n",
    "        return None\n",
    "    mini = float('inf')\n",
    "    choice = ''\n",
    "    for letter in alphabet:\n",
    "        s = node.score[letter]\n",
    "        if parent_letter != letter:\n",
    "            s += 1\n",
    "        if s < mini:\n",
    "            mini = s\n",
    "            choice = letter\n",
    "    node.text += choice\n",
    "    for i in node.children:\n",
    "        tree_pruning(nodelist[i], choice)\n",
    "    return None\n",
    "\n",
    "\n",
    "def tree_edge_weighter(node):\n",
    "    global running_sum\n",
    "    if node.leaf:\n",
    "        return None\n",
    "    for index, item in enumerate(node.children):\n",
    "        dd = seq_dist(node.text, nodelist[item].text)\n",
    "        running_sum += dd\n",
    "        node.edges[index] = dd\n",
    "        tree_edge_weighter(nodelist[item])\n",
    "    return None\n",
    "\n",
    "\n",
    "def tree_printer(node, conn):\n",
    "    if node.leaf:\n",
    "        return None\n",
    "    for index, item in enumerate(node.children):\n",
    "        child = nodelist[item]\n",
    "        conn.write(node.text+'->'+child.text+':'+str(node.edges[index])+'\\n')\n",
    "        conn.write(child.text+'->'+node.text+':'+str(node.edges[index])+'\\n')\n",
    "        tree_printer(child, conn)\n",
    "    return None\n",
    "\n",
    "filename = '/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/trial.txt'\n",
    "with open(filename, 'r') as f:\n",
    "    nodelist = interpreter(f)\n",
    "seq_len = len(nodelist[0].text)\n",
    "for i in range(seq_len):\n",
    "    tree_scoring(i)\n",
    "    tree_pruning(nodelist[-1])\n",
    "    for node in nodelist:\n",
    "        node.reinit()\n",
    "        \n",
    "running_sum = 0\n",
    "tree_edge_weighter(nodelist[-1])\n",
    "with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/out_final.txt', 'w') as g:\n",
    "    g.write(str(running_sum)+'\\n')\n",
    "    tree_printer(nodelist[-1], g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "When the position of the root in the tree is unknown, we can simply assign the root to any edge that we like, apply \n",
    "SmallParsimony to the resulting rooted tree, and then remove the root. It can be shown that this method provides a solution \n",
    "to the following problem.\n",
    "Solve the Small Parsimony in an Unrooted Tree Problem.\n",
    "     Input: An integer n followed by an adjacency list for an unrooted binary tree with n leaves labeled by DNA strings.\n",
    "     Output: The minimum parsimony score of this tree, followed by the adjacency list of the tree corresponding to labeling\n",
    "     internal nodes by DNA strings in order to minimize the parsimony score of the tree.\n",
    "Sample Input:\n",
    "4\n",
    "TCGGCCAA->4\n",
    "4->TCGGCCAA\n",
    "CCTGGCTG->4\n",
    "4->CCTGGCTG\n",
    "CACAGGAT->5\n",
    "5->CACAGGAT\n",
    "TGAGTACC->5\n",
    "5->TGAGTACC\n",
    "4->5\n",
    "5->4\n",
    "Sample Output:\n",
    "17\n",
    "TCGGCCAA->CCAGGCAC:4\n",
    "CCTGGCTG->CCAGGCAC:3\n",
    "TGAGTACC->CAAGGAAC:4\n",
    "CCAGGCAC->CCTGGCTG:3\n",
    "CCAGGCAC->CAAGGAAC:2\n",
    "CCAGGCAC->TCGGCCAA:4\n",
    "CACAGGAT->CAAGGAAC:4\n",
    "CAAGGAAC->CACAGGAT:4\n",
    "CAAGGAAC->TGAGTACC:4\n",
    "CAAGGAAC->CCAGGCAC:2\n",
    "\n",
    "\n",
    "trial.txt --\n",
    "\n",
    "4\n",
    "TCGGCCAA->4\n",
    "4->TCGGCCAA\n",
    "CCTGGCTG->4\n",
    "4->CCTGGCTG\n",
    "CACAGGAT->5\n",
    "5->CACAGGAT\n",
    "TGAGTACC->5\n",
    "5->TGAGTACC\n",
    "4->5\n",
    "5->4\n",
    "'''\n",
    "from collections import defaultdict\n",
    "\n",
    "alphabet = ['A', 'C', 'G', 'T', ]\n",
    "\n",
    "\n",
    "def dist(a, b):\n",
    "    if a == b:\n",
    "        return 0\n",
    "    else:\n",
    "        return 1\n",
    "\n",
    "\n",
    "def seq_dist(a, b):\n",
    "    summ = 0\n",
    "    for i in range(len(a)):\n",
    "        if a[i] != b[i]:\n",
    "            summ += 1\n",
    "    return summ\n",
    "\n",
    "\n",
    "def is_int(s):\n",
    "    try:\n",
    "        int(s)\n",
    "        return True\n",
    "    except ValueError:\n",
    "        return False\n",
    "\n",
    "\n",
    "class Node(object):\n",
    "    def __init__(self, index, leaf):\n",
    "        self.index = index\n",
    "        self.leaf = leaf\n",
    "        self.children = []\n",
    "        self.edges = []\n",
    "        self.text = ''\n",
    "        self.score = defaultdict(int)\n",
    "        self.scored = False\n",
    "\n",
    "    def reinit(self):\n",
    "        self.score = defaultdict(int)\n",
    "        self.scored = False\n",
    "        return None\n",
    "\n",
    "    def is_ripe(self):\n",
    "        if self.leaf:\n",
    "            return True\n",
    "        else:\n",
    "            for item in self.children:\n",
    "                if not nodelist[item].scored:\n",
    "                    return False\n",
    "        return True\n",
    "\n",
    "    def add_node(self, c_index):\n",
    "        self.children.append(c_index)\n",
    "        self.edges.append(0)\n",
    "        return None\n",
    "\n",
    "    def remove_node(self, c_index):\n",
    "        self.children.remove(c_index)\n",
    "        self.edges.pop()\n",
    "        return None\n",
    "\n",
    "    def scoring(self, curr_letter):\n",
    "        if self.leaf:\n",
    "            for letter in alphabet:\n",
    "                if letter == self.text[curr_letter]:\n",
    "                    self.score[letter] = 0\n",
    "                else:\n",
    "                    self.score[letter] = float('inf')\n",
    "        else:\n",
    "            for letter in alphabet:\n",
    "                for item in self.children:\n",
    "                    self.score[letter] += min([nodelist[item].score[i]+dist(letter, i) for i in alphabet])\n",
    "        self.scored = True\n",
    "        return None\n",
    "\n",
    "\n",
    "def interpreter(conn):\n",
    "    nn = int(conn.readline().strip())\n",
    "    nl = dict()\n",
    "    running = 0\n",
    "    for raw_line in conn:\n",
    "        line = raw_line.strip().split('->')\n",
    "        if is_int(line[0]):\n",
    "            a = int(line[0])\n",
    "            if a not in nl:\n",
    "                nl[a] = Node(a, False)\n",
    "            if is_int(line[1]):\n",
    "                b = int(line[1])\n",
    "                if b not in nl:\n",
    "                    nl[b] = Node(b, False)\n",
    "                nl[a].add_node(b)\n",
    "            else:\n",
    "                seq = line[1]\n",
    "                nl[running] = Node(running, True)\n",
    "                nl[running].text = seq\n",
    "                nl[a].add_node(running)\n",
    "                running += 1\n",
    "    return nl\n",
    "\n",
    "\n",
    "def tree_scoring(curr_letter):\n",
    "    for _, node in nodelist.items():\n",
    "        if node.leaf:\n",
    "            node.scoring(curr_letter)\n",
    "    while not nodelist[new_root].scored:\n",
    "        for _, node in nodelist.items():\n",
    "            if not node.scored:\n",
    "                if node.is_ripe():\n",
    "                    node.scoring(curr_letter)\n",
    "    return None\n",
    "\n",
    "\n",
    "def tree_pruning(node, parent_letter=''):\n",
    "    if node.leaf:\n",
    "        return None\n",
    "    mini = float('inf')\n",
    "    choice = ''\n",
    "    for letter in alphabet:\n",
    "        s = node.score[letter]\n",
    "        if parent_letter != letter:\n",
    "            s += 1\n",
    "        if s < mini:\n",
    "            mini = s\n",
    "            choice = letter\n",
    "    node.text += choice\n",
    "    for i in node.children:\n",
    "        tree_pruning(nodelist[i], choice)\n",
    "    return None\n",
    "\n",
    "\n",
    "def tree_edge_weighter(node):\n",
    "    global running_sum\n",
    "    if node.leaf:\n",
    "        return None\n",
    "    for index, item in enumerate(node.children):\n",
    "        dd = seq_dist(node.text, nodelist[item].text)\n",
    "        running_sum += dd\n",
    "        node.edges[index] = dd\n",
    "        tree_edge_weighter(nodelist[item])\n",
    "    return None\n",
    "\n",
    "\n",
    "def tree_printer(node, conn):\n",
    "    if node.leaf:\n",
    "        return None\n",
    "    for index, item in enumerate(node.children):\n",
    "        child = nodelist[item]\n",
    "        conn.write(node.text+'->'+child.text+':'+str(node.edges[index])+'\\n')\n",
    "        conn.write(child.text+'->'+node.text+':'+str(node.edges[index])+'\\n')\n",
    "        tree_printer(child, conn)\n",
    "    return None\n",
    "\n",
    "\n",
    "def tree_rooting(node):\n",
    "    if node.leaf:\n",
    "        return None\n",
    "    ii = node.index\n",
    "    for child in node.children:\n",
    "        if ii in nodelist[child].children:\n",
    "            nodelist[child].remove_node(ii)\n",
    "        tree_rooting(nodelist[child])\n",
    "        \n",
    "with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/trial.txt', 'r') as f:\n",
    "    nodelist = interpreter(f)\n",
    "seq_len = len(nodelist[0].text)\n",
    "# Tree rooting process\n",
    "new_root = len(nodelist)\n",
    "last_node = new_root - 1\n",
    "connected_node = max(nodelist[last_node].children)\n",
    "nodelist[new_root] = Node(new_root, False)\n",
    "nodelist[new_root].add_node(last_node)\n",
    "nodelist[new_root].add_node(connected_node)\n",
    "nodelist[last_node].remove_node(connected_node)\n",
    "nodelist[connected_node].remove_node(last_node)\n",
    "tree_rooting(nodelist[new_root])\n",
    "\n",
    "# Tree scoring\n",
    "for i in range(seq_len):\n",
    "    tree_scoring(i)\n",
    "    tree_pruning(nodelist[new_root])\n",
    "    for _, node in nodelist.items():\n",
    "        node.reinit()\n",
    "running_sum = 0\n",
    "tree_edge_weighter(nodelist[new_root])\n",
    "\n",
    "# Tree uprooting\n",
    "top_edge_weight = sum(nodelist[new_root].edges)\n",
    "del nodelist[new_root]\n",
    "nodelist[last_node].add_node(connected_node)\n",
    "nodelist[last_node].edges[-1] = top_edge_weight\n",
    "with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/out.txt', 'w') as g:\n",
    "    g.write(str(running_sum)+'\\n')\n",
    "    tree_printer(nodelist[last_node], g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Solve the Nearest Neighbors of a Tree Problem.\n",
    "     Input: Two internal nodes a and b specifying an edge e, followed by an adjacency list of an unrooted binary tree.\n",
    "     Output: Two adjacency lists representing the nearest neighbors of the tree with respect to e. Separate the\n",
    "     adjacency lists with a blank line.\n",
    "Sample Input:\n",
    "5 4\n",
    "0->4\n",
    "4->0\n",
    "1->4\n",
    "4->1\n",
    "2->5\n",
    "5->2\n",
    "3->5\n",
    "5->3\n",
    "4->5\n",
    "5->4\n",
    "Sample Output:\n",
    "1->4\n",
    "0->5\n",
    "3->4\n",
    "2->5\n",
    "5->2\n",
    "5->4\n",
    "5->0\n",
    "4->1\n",
    "4->5\n",
    "4->3\n",
    "1->5\n",
    "0->4\n",
    "3->4\n",
    "2->5\n",
    "5->2\n",
    "5->4\n",
    "5->1\n",
    "4->0\n",
    "4->5\n",
    "4->3\n",
    "\n",
    "trial.txt --\n",
    "\n",
    "5 4\n",
    "0->4\n",
    "4->0\n",
    "1->4\n",
    "4->1\n",
    "2->5\n",
    "5->2\n",
    "3->5\n",
    "5->3\n",
    "4->5\n",
    "5->4\n",
    "'''\n",
    "\n",
    "from collections import defaultdict\n",
    "from copy import copy, deepcopy\n",
    "\n",
    "\n",
    "class Node(object):\n",
    "    def __init__(self, index, leaf):\n",
    "        self.index = index\n",
    "        self.leaf = leaf\n",
    "        self.children = []\n",
    "        self.edges = []\n",
    "        self.text = ''\n",
    "        self.score = defaultdict(int)\n",
    "        self.scored = False\n",
    "\n",
    "    def reinit(self):\n",
    "        self.score = defaultdict(int)\n",
    "        self.scored = False\n",
    "        return None\n",
    "\n",
    "    def is_ripe(self):\n",
    "        if self.leaf:\n",
    "            return True\n",
    "        else:\n",
    "            for item in self.children:\n",
    "                if not nodelist[item].scored:\n",
    "                    return False\n",
    "        return True\n",
    "\n",
    "    def add_node(self, c_index):\n",
    "        self.children.append(c_index)\n",
    "        self.edges.append(0)\n",
    "        return None\n",
    "\n",
    "    def remove_node(self, c_index):\n",
    "        self.children.remove(c_index)\n",
    "        self.edges.pop()\n",
    "        return None\n",
    "\n",
    "\n",
    "def interpreter(conn):\n",
    "    edge = conn.readline().strip().split(' ')\n",
    "    nl = dict()\n",
    "    for raw_line in conn:\n",
    "        line = raw_line.strip().split('->')\n",
    "        a = int(line[0])\n",
    "        if a not in nl:\n",
    "            nl[a] = Node(a, False)\n",
    "        b = int(line[1])\n",
    "        if b not in nl:\n",
    "            nl[b] = Node(b, False)\n",
    "        nl[a].add_node(b)\n",
    "    return edge, nl\n",
    "\n",
    "\n",
    "def tree_printer(nodelist, conn):\n",
    "    for _, node in nodelist.items():\n",
    "        for child in node.children:\n",
    "            conn.write(str(node.index)+'->'+str(child)+'\\n')\n",
    "    conn.write('\\n')\n",
    "    return None\n",
    "\n",
    "\n",
    "def exchanger(aa, bb, tree):\n",
    "    first_node = tree[aa]\n",
    "    second_node = tree[bb]\n",
    "    first_children = copy(first_node.children)\n",
    "    second_children = copy(second_node.children)\n",
    "    first_children.remove(bb)\n",
    "    second_children.remove(aa)\n",
    "    fixed_node = first_children[0]\n",
    "    treelist = list()\n",
    "    for ii in range(2):\n",
    "        new_node1 = Node(aa, False)\n",
    "        new_node1.add_node(bb)\n",
    "        new_node1.add_node(fixed_node)\n",
    "        new_node1.add_node(second_children[ii])\n",
    "        new_node2 = Node(bb, False)\n",
    "        new_node2.add_node(aa)\n",
    "        new_node2.add_node(first_children[1])\n",
    "        new_node2.add_node(second_children[(ii+1)%2])\n",
    "        new_tree = deepcopy(tree)\n",
    "        new_tree[aa] = new_node1\n",
    "        new_tree[bb] = new_node2\n",
    "        new_tree[first_children[1]].remove_node(aa)\n",
    "        new_tree[first_children[1]].add_node(bb)\n",
    "        new_tree[second_children[ii]].remove_node(bb)\n",
    "        new_tree[second_children[ii]].add_node(aa)\n",
    "        treelist.append(new_tree)\n",
    "    return treelist\n",
    "\n",
    "filename = '/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/trial.txt'\n",
    "\n",
    "with open(filename, 'r') as f:\n",
    "    edge, nodelist = interpreter(f)\n",
    "a, b = int(edge[0]), int(edge[1])\n",
    "treelist = exchanger(a, b, nodelist)\n",
    "with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/out.txt', 'w') as g:\n",
    "    for nodes in treelist:\n",
    "        tree_printer(nodes, g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Implement the nearest neighbor interchange heuristic for the Large Parsimony Problem.\n",
    "     Input: An integer n, followed by an adjacency list for an unrooted binary tree whose n leaves are labeled by DNA strings and\n",
    "     whose internal nodes are labeled by integers.\n",
    "     Output: The parsimony score and unrooted labeled tree obtained after every step of the nearest neighbor interchange heuristic.\n",
    "     Each step should be separated by a blank line.\n",
    "Note: Depending on how your code breaks ties, you may obtain a different solution than the one we provide.  As a result, the parsimony \n",
    "score at each step may vary.\n",
    "Sample Input:\n",
    "5\n",
    "GCAGGGTA->5\n",
    "TTTACGCG->5\n",
    "CGACCTGA->6\n",
    "GATTCCAC->6\n",
    "5->TTTACGCG\n",
    "5->GCAGGGTA\n",
    "5->7\n",
    "TCCGTAGT->7\n",
    "7->5\n",
    "7->6\n",
    "7->TCCGTAGT\n",
    "6->GATTCCAC\n",
    "6->CGACCTGA\n",
    "6->7\n",
    "Sample Output:\n",
    "22\n",
    "TCCGTAGT->TCAGCGGA:4\n",
    "GATTCCAC->GAACCCGA:4\n",
    "CGACCTGA->GAACCCGA:3\n",
    "TTTACGCG->TCAGCGGA:5\n",
    "TCAGCGGA->TTTACGCG:5\n",
    "TCAGCGGA->GCAGCGGA:1\n",
    "TCAGCGGA->TCCGTAGT:4\n",
    "GCAGGGTA->GCAGCGGA:2\n",
    "GCAGCGGA->GAACCCGA:3\n",
    "GCAGCGGA->GCAGGGTA:2\n",
    "GCAGCGGA->TCAGCGGA:1\n",
    "GAACCCGA->GATTCCAC:4\n",
    "GAACCCGA->CGACCTGA:3\n",
    "GAACCCGA->GCAGCGGA:3\n",
    "21\n",
    "TCCGTAGT->TCTGCGGA:4\n",
    "GATTCCAC->GCTGCGGA:5\n",
    "CGACCTGA->GCAGCGGA:4\n",
    "TTTACGCG->TCTGCGGA:4\n",
    "TCTGCGGA->TTTACGCG:4\n",
    "TCTGCGGA->GCTGCGGA:1\n",
    "TCTGCGGA->TCCGTAGT:4\n",
    "GCAGGGTA->GCAGCGGA:2\n",
    "GCTGCGGA->GCAGCGGA:1\n",
    "GCTGCGGA->GATTCCAC:5\n",
    "GCTGCGGA->TCTGCGGA:1\n",
    "GCAGCGGA->CGACCTGA:4\n",
    "GCAGCGGA->GCAGGGTA:2\n",
    "GCAGCGGA->GCTGCGGA:1\n",
    "\n",
    "\n",
    "trial.txt --\n",
    "\n",
    "5\n",
    "GCAGGGTA->5\n",
    "TTTACGCG->5\n",
    "CGACCTGA->6\n",
    "GATTCCAC->6\n",
    "5->TTTACGCG\n",
    "5->GCAGGGTA\n",
    "5->7\n",
    "TCCGTAGT->7\n",
    "7->5\n",
    "7->6\n",
    "7->TCCGTAGT\n",
    "6->GATTCCAC\n",
    "6->CGACCTGA\n",
    "6->7\n",
    "\n",
    "'''\n",
    "\n",
    "import sys\n",
    "import queue\n",
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "class LargeParsimony:\n",
    "    def __init__(self):\n",
    "        n, adj, nodes, lastEdge = self.readFromFile()\n",
    "        trees = self.runNearestNeighborInterchange(n, adj, nodes, lastEdge)\n",
    "        self.saveTrees(trees)\n",
    "        self.printTrees(trees)\n",
    "    \n",
    "    def _input(self):\n",
    "        data = sys.stdin.read().strip().split('\\n')\n",
    "        n = int(data[0])\n",
    "        adj = [dict() for _ in range(n)]\n",
    "        nodes = ['' for _ in range(n)]\n",
    "        currNode = 0\n",
    "        for d in data[1:]:\n",
    "            d = d.split('->')\n",
    "            try:\n",
    "                p = int(d[0])\n",
    "            except:\n",
    "                p = currNode\n",
    "                nodes[p] = d[0]\n",
    "                currNode += 1\n",
    "            try:\n",
    "                c = int(d[1])\n",
    "            except:\n",
    "                continue\n",
    "            if p > len(adj)-1 or c > len(adj)-1:\n",
    "                adj.extend([dict() for _ in range(max([p,c])-len(adj)+1)])\n",
    "            adj[p][c] = 0\n",
    "            adj[c][p] = 0\n",
    "        nodes.extend(['' for _ in range(len(adj)-n+1)])\n",
    "        lastEdge = [int(i) for i in data[-1].split('->')]\n",
    "        return n, adj, nodes, lastEdge\n",
    "\n",
    "    def readFromFile(self):\n",
    "        f = open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/trial.txt', 'r')\n",
    "        data = []\n",
    "        for line in f:\n",
    "            data.append(line.strip())\n",
    "        n = int(data[0])\n",
    "        adj = [dict() for _ in range(n)]\n",
    "        nodes = ['' for _ in range(n)]\n",
    "        currNode = 0\n",
    "        for d in data[1:]:\n",
    "            d = d.split('->')\n",
    "            try:\n",
    "                p = int(d[0])\n",
    "            except:\n",
    "                p = currNode\n",
    "                nodes[p] = d[0]\n",
    "                currNode += 1\n",
    "            try:\n",
    "                c = int(d[1])\n",
    "            except:\n",
    "                continue\n",
    "            if p > len(adj)-1 or c > len(adj)-1:\n",
    "                adj.extend([dict() for _ in range(max([p,c])-len(adj)+1)])\n",
    "            adj[p][c] = 0\n",
    "            adj[c][p] = 0\n",
    "        nodes.extend(['' for _ in range(len(adj)-n+1)])\n",
    "        lastEdge = [int(i) for i in data[-1].split('->')]\n",
    "        f.close()\n",
    "        return n, adj, nodes, lastEdge\n",
    "\n",
    "    def printResults(self, s, adj, nodes):\n",
    "        print(s)\n",
    "        for i, d in enumerate(adj):\n",
    "            for j, w in d.items():\n",
    "                print(nodes[i]+'->'+nodes[j]+':'+str(w))\n",
    "    \n",
    "    def printTrees(self, trees):\n",
    "        for s, adj, nodes in trees:\n",
    "            print(s)\n",
    "            for i, d in enumerate(adj):\n",
    "                for j, w in d.items():\n",
    "                    print(nodes[i]+'->'+nodes[j]+':'+str(w))\n",
    "            print('')\n",
    "    \n",
    "    def saveTrees(self, trees):\n",
    "        f = open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/result.txt', 'w')\n",
    "        for s, adj, nodes in trees:\n",
    "            f.write(str(s)+'\\n')\n",
    "            for i, d in enumerate(adj):\n",
    "                for j, w in d.items():\n",
    "                    f.write(nodes[i]+'->'+nodes[j]+':'+str(w)+'\\n')\n",
    "            f.write('\\n')\n",
    "    \n",
    "    def charIndConversion(self):\n",
    "        char2ind = {'A':0, 'C':1, 'G':2, 'T':3}\n",
    "        ind2char = {0:'A', 1:'C', 2:'G', 3:'T'}\n",
    "        return char2ind, ind2char\n",
    "    \n",
    "    def singleSmallParsimony(self, n, adjC, adjP, adj, nodes, char2ind, ind2char, charInd):\n",
    "        s = [[np.inf]*4 for _ in range(len(adjC))]\n",
    "        backtrack = [[(-1, -1) for _ in range(4)] for __ in range(len(adjC))]\n",
    "        processed = [0 for _ in range(len(adjC))]\n",
    "        ripe = set()\n",
    "        for i in range(n):\n",
    "            s[i][char2ind[nodes[i][charInd]]] = 0\n",
    "            processed[i] = 1\n",
    "            if len(adjP[i]) > 0:\n",
    "                ripe.add(adjP[i][0])\n",
    "        \n",
    "        while len(ripe) > 0:\n",
    "            v = ripe.pop()\n",
    "            for k in range(4):\n",
    "                l = [s[adjC[v][0]][i] + (0 if k == i else 1) for i in range(4)]\n",
    "                r = [s[adjC[v][1]][i] + (0 if k == i else 1) for i in range(4)]\n",
    "                largmin = np.argmin(l)\n",
    "                rargmin = np.argmin(r)\n",
    "                backtrack[v][k] = (largmin, rargmin)\n",
    "                s[v][k] = l[largmin] + r[rargmin]\n",
    "            processed[v] = 1\n",
    "            if len(adjP[v]) > 0 and all([processed[u] for u in adjC[adjP[v][0]]]):\n",
    "                ripe.add(adjP[v][0])\n",
    "\n",
    "        ind = np.argmin(s[v])\n",
    "        nodes[v] += ind2char[ind]\n",
    "        smin = s[v][ind]\n",
    "\n",
    "        q = queue.Queue()\n",
    "        q.put((v, ind))\n",
    "        while not q.empty():\n",
    "            v, k = q.get()\n",
    "            if len(adjC[v]) > 0:\n",
    "                u, w = adjC[v]\n",
    "                l, r = backtrack[v][k]\n",
    "                \n",
    "                if k != l:\n",
    "                    adj[v][u] += 1\n",
    "                    adj[u][v] += 1\n",
    "                if k != r:\n",
    "                    adj[v][w] += 1\n",
    "                    adj[w][v] += 1\n",
    "\n",
    "                if len(adjC[u]) > 0:\n",
    "                    nodes[u] += ind2char[l]\n",
    "                    q.put((u, l))\n",
    "                if len(adjC[w]) > 0:\n",
    "                    nodes[w] += ind2char[r]\n",
    "                    q.put((w, r))\n",
    "      \n",
    "        return smin\n",
    "    \n",
    "    def runSmallParsimony(self, n, adj, nodes, lastEdge): # for unrooted binary tree\n",
    "        def dist(v, w):\n",
    "            d = 0\n",
    "            l = len(v)\n",
    "            for i in range(l):\n",
    "                if v[i] != w[i]:\n",
    "                    d += 1\n",
    "            return d\n",
    "\n",
    "        char2ind, ind2char = self.charIndConversion()\n",
    "        root = len(adj)\n",
    "        del adj[lastEdge[0]][lastEdge[1]]\n",
    "        del adj[lastEdge[1]][lastEdge[0]]\n",
    "        adj.append(dict())\n",
    "        adj[root][lastEdge[0]] = 0\n",
    "        adj[lastEdge[0]][root] = 0\n",
    "        adj[root][lastEdge[1]] = 0\n",
    "        adj[lastEdge[1]][root] = 0\n",
    "        adjC = [[] for _ in range(len(adj))]\n",
    "        adjP = [[] for _ in range(len(adj))]\n",
    "        q = queue.Queue()\n",
    "        q.put(root)\n",
    "        visited = [False for _ in range(len(adj))]\n",
    "        visited[root] = True\n",
    "        while not q.empty():\n",
    "            curr = q.get()\n",
    "            for v in adj[curr].keys():\n",
    "                if not visited[v]:\n",
    "                    adjP[v].append(curr)\n",
    "                    visited[v] = True\n",
    "                    q.put(v)\n",
    "        for u, d in enumerate(adjP):\n",
    "            for v in d:\n",
    "                adjC[v].append(u)\n",
    "        s = 0\n",
    "        for i in range(len(nodes[0])):\n",
    "            s += self.singleSmallParsimony(n, adjC, adjP, adj, nodes, char2ind, ind2char, i)\n",
    "        d = dist(nodes[lastEdge[0]], nodes[lastEdge[1]])\n",
    "        del adj[root]\n",
    "        del adj[lastEdge[0]][root]\n",
    "        del adj[lastEdge[1]][root]\n",
    "        adj[lastEdge[0]][lastEdge[1]] = d\n",
    "        adj[lastEdge[1]][lastEdge[0]] = d\n",
    "        return s, adj, nodes\n",
    "    \n",
    "    def findNearestNeighbors(self, edge, adj):\n",
    "        adj1 = deepcopy(adj)\n",
    "        adj2 = deepcopy(adj)\n",
    "\n",
    "        del adj1[edge[0]][edge[1]]\n",
    "        del adj1[edge[1]][edge[0]]\n",
    "        e0 = list(adj1[edge[0]].keys())\n",
    "        e1 = list(adj1[edge[1]].keys())\n",
    "        adj1[edge[0]][e1[0]] = 0\n",
    "        adj1[edge[1]][e0[0]] = 0\n",
    "        adj1[e1[0]][edge[0]] = 0\n",
    "        adj1[e0[0]][edge[1]] = 0\n",
    "        del adj1[e1[0]][edge[1]]\n",
    "        del adj1[e0[0]][edge[0]]\n",
    "        del adj1[edge[0]][e0[0]]\n",
    "        del adj1[edge[1]][e1[0]]\n",
    "        adj1[edge[0]][edge[1]] = 0\n",
    "        adj1[edge[1]][edge[0]] = 0\n",
    "\n",
    "        adj2[edge[0]][e1[1]] = 0\n",
    "        adj2[edge[1]][e0[0]] = 0\n",
    "        adj2[e1[1]][edge[0]] = 0\n",
    "        adj2[e0[0]][edge[1]] = 0\n",
    "        del adj2[e1[1]][edge[1]]\n",
    "        del adj2[e0[0]][edge[0]]\n",
    "        del adj2[edge[0]][e0[0]]\n",
    "        del adj2[edge[1]][e1[1]]\n",
    "        return adj1, adj2\n",
    "\n",
    "    def runNearestNeighborInterchange(self, n, adj, nodes, lastEdge):\n",
    "        trees = []\n",
    "        score = np.inf\n",
    "        newScore, newAdj, newNodes = self.runSmallParsimony(n, adj, deepcopy(nodes), lastEdge)\n",
    "        while newScore < score:\n",
    "            score = newScore\n",
    "            adj = newAdj\n",
    "            visited = set()\n",
    "            for v in range(n, len(adj)):\n",
    "                for u in adj[v].keys():\n",
    "                    if u >= n and not (v, u) in visited:\n",
    "                        adj1, adj2 = self.findNearestNeighbors([v, u], adj)\n",
    "                        for i, a in enumerate(adj1):\n",
    "                            adj1[i] = dict.fromkeys(a, 0)\n",
    "                        for i, a in enumerate(adj2):\n",
    "                            adj2[i] = dict.fromkeys(a, 0)\n",
    "                        neighborScore, neighborAdj, neighborNodes = self.runSmallParsimony(n, adj1, deepcopy(nodes), [v, u])\n",
    "                        if neighborScore < newScore:\n",
    "                            newScore = neighborScore\n",
    "                            newAdj = neighborAdj\n",
    "                            newNodes = neighborNodes\n",
    "                        neighborScore, neighborAdj, neighborNodes = self.runSmallParsimony(n, adj2, deepcopy(nodes), [v, u])\n",
    "                        if neighborScore < newScore:\n",
    "                            newScore = neighborScore\n",
    "                            newAdj = neighborAdj\n",
    "                            newNodes = neighborNodes                \n",
    "                        visited.add((v, u))\n",
    "                        visited.add((u, v))\n",
    "            if newScore < score:\n",
    "                trees.append((newScore, newAdj, newNodes))\n",
    "        return trees"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "We represent the masses in a spectrum as a sequence Spectrum of integers  s1,…,sms1,…,sm in increasing order, where s1s1 is \n",
    "zero and smsm is the total mass of the (unknown) peptide. We define a labeled graph Graph(Spectrum) by forming a node for each \n",
    "element of Spectrum, then connecting nodes sisi and sjsj by a directed edge labeled by an amino acid aa if sj−sisj−si is equal \n",
    "to the mass of aa. As we assumed when sequencing antibiotics, we do not distinguish between amino acids having the same integer \n",
    "masses (i.e., the pairs K/Q and I/L).\n",
    "Construct the graph of a spectrum.\n",
    "     Given: A space-delimited list of integers Spectrum.\n",
    "     Return: Graph(Spectrum).\n",
    "Note: Throughout this chapter, all dataset problems implicitly use the standard integer-valued mass table for the regular twenty \n",
    "amino acids. Examples sometimes use the toy amino acid alphabet {X, Z} whose masses are 4 and 5, respectively.\n",
    "Sample Input:\n",
    "57 71 154 185 301 332 415 429 486\n",
    "Sample Output:\n",
    "0->57:G\n",
    "0->71:A\n",
    "57->154:P\n",
    "57->185:K\n",
    "71->185:N\n",
    "154->301:F\n",
    "185->332:F\n",
    "301->415:N\n",
    "301->429:K\n",
    "332->429:P\n",
    "415->486:A\n",
    "429->486:G\n",
    "'''\n",
    "\n",
    "massTable = {'A': 71, 'C': 103, 'E': 129, 'D': 115, 'G': 57, 'F': 147, 'I': 113, 'H': 137, 'K': 128, 'M': 131, 'L': 113, 'N': 114, 'Q': 128, 'P': 97, 'S': 87, 'R': 156, 'T': 101, 'W': 186, 'V': 99, 'Y': 163}\n",
    "\n",
    "def spectrum_graph(Spectrum):\n",
    "    Spectrum = sorted(Spectrum)\n",
    "    for i in range(len(Spectrum)):\n",
    "        for j in range(i+1,len(Spectrum)):\n",
    "            if Spectrum[j]-Spectrum[i] in massTable.values():\n",
    "                aa = [ key for key,val in massTable.items() if val == Spectrum[j]-Spectrum[i]]\n",
    "                print(str(Spectrum[i])+'->'+str(Spectrum[j]) + ':' + aa[0])\n",
    "                \n",
    "Spectrum = [57,71,154,185,301,332,415,429,486]\n",
    "Spectrum = [0] + Spectrum\n",
    "spectrum_graph(Spectrum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "integer_mass_table.txt --\n",
    "\n",
    "G 57\n",
    "A 71\n",
    "S 87\n",
    "P 97\n",
    "V 99\n",
    "T 101\n",
    "C 103\n",
    "I 113\n",
    "L 113\n",
    "N 114\n",
    "D 115\n",
    "K 128\n",
    "Q 128\n",
    "E 129\n",
    "M 131\n",
    "H 137\n",
    "F 147\n",
    "R 156\n",
    "Y 163\n",
    "W 186\n",
    "\n",
    "'''\n",
    "\n",
    "'''\n",
    "Given an amino acid string Peptide, its ideal spectrum, denoted IdealSpectrum(Peptide), is the collection of integer masses of \n",
    "all its prefixes and suffixes. Note that an ideal spectrum may have repeated masses; for example, IdealSpectrum(GPG) = {0, 57, \n",
    "57, 154, 154, 211}. We say that an amino acid string Peptide explains a collection of integers Spectrum if \n",
    "IdealSpectrum(Peptide) = Spectrum.\n",
    "Decoding an Ideal Spectrum Problem: Reconstruct a peptide from its ideal spectrum.\n",
    "     Input: A collection of integers Spectrum.\n",
    "     Output: An amino acid string Peptide that explains Spectrum.\n",
    "Sample Input:\n",
    "57 71 154 185 301 332 415 429 486\n",
    "Sample Output:\n",
    "GPFNA\n",
    "'''\n",
    "\n",
    "class Node(object):\n",
    "    def __init__(self, value):\n",
    "        self.value = value\n",
    "        self.children = {}\n",
    "\n",
    "    def add_child(self, child, acid):\n",
    "        self.children[child] = acid\n",
    "\n",
    "\n",
    "def interpreter(conn):\n",
    "    tspec = conn.readline().strip().split(' ')\n",
    "    tspecint = [int(x) for x in tspec]\n",
    "    tspecint.insert(0, 0)\n",
    "    return tspecint\n",
    "\n",
    "\n",
    "def graph_build(spec, conn):\n",
    "    graph = {i: Node(i) for i in spec}\n",
    "    for i in spec:\n",
    "        for j in spec:\n",
    "            if i < j:\n",
    "                diff = j-i\n",
    "                if diff in mass_to_amino:\n",
    "                    conn.write(str(i)+'->'+str(j)+':'+mass_to_amino[diff]+'\\n')\n",
    "                    graph[i].add_child(j, mass_to_amino[diff])\n",
    "    return graph\n",
    "\n",
    "\n",
    "def mass(seq):\n",
    "    m = 0\n",
    "    for lett in seq:\n",
    "        m += amino_to_mass[lett]\n",
    "    return m\n",
    "\n",
    "def ideal_spec(seq):\n",
    "    spec = list()\n",
    "    spec.append(0)\n",
    "    spec.append(mass(seq))\n",
    "    for i in range(1, len(seq)):\n",
    "        spec.append(mass(seq[:i]))\n",
    "        spec.append(mass(seq[i:]))\n",
    "    spec = list(set(spec))\n",
    "    spec.sort()\n",
    "    return spec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "trial.txt --\n",
    "\n",
    "57 71 154 185 301 332 415 429 486\n",
    "\n",
    "'''\n",
    "with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/integer_mass_table.txt') as e:\n",
    "        mass_to_amino = dict()\n",
    "        amino_to_mass = dict()\n",
    "        for item in e:\n",
    "            temp = item.strip().split(' ')\n",
    "            try:\n",
    "                mass_to_amino[int(temp[1])] = temp[0]\n",
    "                amino_to_mass[temp[0]] = int(temp[1])\n",
    "            except IndexError:\n",
    "                pass\n",
    "\n",
    "with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/trial.txt', 'r') as f:\n",
    "    spectrum = interpreter(f)\n",
    "    final = max(spectrum)\n",
    "\n",
    "with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/out.txt', 'w') as g:\n",
    "    spec_graph = graph_build(spectrum, g)\n",
    "\n",
    "possibles = [(spec_graph[0], '')]\n",
    "out_pep = []\n",
    "while possibles:\n",
    "    curr = possibles.pop()\n",
    "    if not curr[0].children:\n",
    "        if curr[0].value == final:\n",
    "            out_pep.append(curr[1])\n",
    "    else:\n",
    "        for child, acid in curr[0].children.items():\n",
    "            possibles.append((spec_graph[child], curr[1]+acid))\n",
    "\n",
    "for item in out_pep:\n",
    "    spec_out = ideal_spec(item)\n",
    "    if spec_out == spectrum:\n",
    "        print(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Given an amino acid string Peptide = a1 . . . an of length n, we will represent its prefix masses using a binary peptide \n",
    "vector Peptide' with mass(Peptide) coordinates. This vector contains a 1 at each of the n prefix coordinates\n",
    "mass(a1), mass(a1 a2), . . . , mass(a1 a2 . . . an ) , and it contains a 0 in each of the remaining noise coordinates. The \n",
    "toy peptide XZZXX, whose prefix masses are 4, 9, 14, 18, and 22, corresponds to the peptide vector \n",
    "(0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,1) of length 22.\n",
    "1. Converting a Peptide into a Peptide Vector Problem: Convert a peptide into a peptide vector.\n",
    "     Input: An amino acid string Peptide.\n",
    "     Output: The peptide vector Peptide'.\n",
    "Sample Input:\n",
    "XZZXX\n",
    "Sample Output:\n",
    "0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 1 0 0 0 1\n",
    "2. Converting a Peptide Vector into a Peptide Problem: Convert a peptide vector into a peptide.\n",
    "     Input: A binary vector P.\n",
    "     Output: A peptide whose peptide vector is equal to P (if such a peptide exists).\n",
    "Sample Input:\n",
    "0 0 0 1 0 0 0 0 1 0 0 0 0 1 0 0 0 1 0 0 0 1\n",
    "Sample Output:\n",
    "XZZXX\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1. Converting a Peptide into a Peptide Vector Problem\n",
    "\n",
    "massTable = {'A': 71, 'C': 103, 'E': 129, 'D': 115, 'G': 57, 'F': 147, 'I': 113, 'H': 137, 'K': 128, 'M': 131, 'L': 113, 'N': 114, 'Q': 128, 'P': 97, 'S': 87, 'R': 156, 'T': 101, 'W': 186, 'V': 99, 'Y': 163,'X':4,'Z':5}\n",
    "\n",
    "def get_mass(peptide):\n",
    "    if len(peptide) == 0:\n",
    "        return 0\n",
    "    return sum([massTable[pep] for pep in peptide])\n",
    "\n",
    "def peptide_vector(peptide):\n",
    "    mass = []\n",
    "    for i in range(1,len(peptide)+1):\n",
    "        mass.append(get_mass(peptide[:i]))\n",
    "    pepVec = [0] * max(mass)\n",
    "    for mi in mass:\n",
    "        pepVec[mi-1] = 1\n",
    "    return pepVec\n",
    "\n",
    "peptide = 'XZZXX'\n",
    "ret = peptide_vector(peptide)\n",
    "print(' '.join(map(str,ret)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#2. Converting a Peptide Vector into a Peptide Problem\n",
    "\n",
    "massTable = {'A': 71, 'C': 103, 'E': 129, 'D': 115, 'G': 57, 'F': 147, 'I': 113, 'H': 137, 'K': 128, 'M': 131, 'L': 113, 'N': 114, 'Q': 128, 'P': 97, 'S': 87, 'R': 156, 'T': 101, 'W': 186, 'V': 99, 'Y': 163,'X':4,'Z':5}\n",
    "\n",
    "def get_mass(peptide):\n",
    "    if len(peptide) == 0:\n",
    "        return 0\n",
    "    return sum([massTable[pep] for pep in peptide])\n",
    "\n",
    "def spectrum_graph(Spectrum):\n",
    "    Spectrum = sorted(Spectrum)\n",
    "    specGraph = defaultdict(list)\n",
    "    weighGraph = defaultdict(list)\n",
    "    for i in range(len(Spectrum)):\n",
    "        for j in range(i+1,len(Spectrum)):\n",
    "            if Spectrum[j]-Spectrum[i] in massTable.values():\n",
    "                aa = [ key for key,val in massTable.items() if val == Spectrum[j]-Spectrum[i]]\n",
    "                specGraph[Spectrum[i]].append(Spectrum[j])\n",
    "                weighGraph[Spectrum[i]].append(aa[0])\n",
    "    specGraph[max(Spectrum)] = []\n",
    "    weighGraph[max(Spectrum)] = []\n",
    "    return specGraph,weighGraph\n",
    "\n",
    "def depthFirstPaths(graph,s):\n",
    "    nodes = graph.keys()\n",
    "    edgeTo = dict((key,None) for key in graph.keys())\n",
    "    marked = dict((key,False) for key in graph.keys())\n",
    "    postorder = []\n",
    "    edgeTo, marked,postorder = dfs(graph,s,edgeTo,marked,postorder)\n",
    "    return edgeTo, marked,postorder\n",
    "\n",
    "def pathTo(graph,weights,s,v,edgeTo):\n",
    "    stack = []\n",
    "    weightstack = []\n",
    "    w = edgeTo[v]\n",
    "    stack.append(v)\n",
    "    weightstack.append(weights[w][graph[w].index(v)])\n",
    "    while w != s:\n",
    "        stack.append(w)\n",
    "        v = w\n",
    "        w = edgeTo[v]\n",
    "        weightstack.append(weights[w][graph[w].index(v)])\n",
    "    stack.append(s)\n",
    "    return stack[::-1],''.join(weightstack[::-1])\n",
    "\n",
    "\n",
    "def from_peptide_vector(vector):\n",
    "    mass = [i+1 for i in range(len(vector)) if vector[i]==1]\n",
    "    mass = [0] + mass\n",
    "    return mass\n",
    "\n",
    "def ideal_spectrum(peptide):\n",
    "    idealPrefixSpec = []\n",
    "    for i in range(1,len(peptide)+1):\n",
    "        idealPrefixSpec.append(get_mass(peptide[:i]))\n",
    "    return idealPrefixSpec\n",
    "\n",
    "def decode_ideal_spec(spec):\n",
    "    specGraph,weighGraph = spectrum_graph(spec)\n",
    "    source = min(spec)\n",
    "    sink = max(spec)\n",
    "    edgeTo, marked,postorder = depthFirstPaths(specGraph,source)\n",
    "    paths,peptides = pathTo(specGraph,weighGraph,source,sink,edgeTo)\n",
    "    peptide = []\n",
    "    for i in range(len(paths)-1):\n",
    "        peptide.append(weighGraph[paths[i]][specGraph[paths[i]].index(paths[i+1])])\n",
    "    return ''.join(peptide)\n",
    "\n",
    "pepvector = [0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1]\n",
    "spec = from_peptide_vector(pepvector)\n",
    "print(decode_ideal_spec(spec))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Given a spectral vector Spectrum', our goal is to find a peptide Peptide maximizing Score(Peptide, Spectrum'). Since the mass \n",
    "of a peptide and the parent mass of the spectrum that it generates should be the same, a peptide vector should have the same \n",
    "length as the spectral vector under consideration. We will therefore define the score between a peptide vector and a spectral \n",
    "vector of different length as −∞.\n",
    "Peptide Sequencing Problem: Given a spectral vector, find a peptide with maximum score against this spectrum.\n",
    "     Input: A spectral vector Spectrum'.\n",
    "     Output: An amino acid string Peptide that maximizes Score(Peptide', Spectrum') among all possible amino acid strings.\n",
    "Given a spectral vector Spectrum' = (s1, . . . , sm), we will construct a DAG on m + 1 nodes, labeled with the integers from \n",
    "0 (source) to m (sink), and then connect node i to node j by a directed edge if j − i is equal to the mass of an amino acid. \n",
    "We will further assign weight si to node i (for 1 ≤ i ≤ m) and assign weight zero to node 0.\n",
    "Any path connecting source to sink in this DAG corresponds to an amino acid string Peptide, and the total weight of nodes on \n",
    "this path is equal to Score(Peptide', Spectrum'). We have therefore reduced the Peptide Sequencing Problem to the problem of \n",
    "finding a maximum-weight path from source to sink in a node-weighted DAG.\n",
    "Solve the Peptide Sequencing Problem.\n",
    "     Given: A space-delimited spectral vector Spectrum'.\n",
    "     Return: An amino acid string with maximum score against Spectrum'. For masses\n",
    "     with more than one amino acid, any choice may be used.\n",
    "Note: When a spectral vector Spectrum' = s1 ... sm is given, it does not have a zero-th element; in your implementations, you \n",
    "should assume that s0 is equal to zero.\n",
    "Input\n",
    "20 2 -14 -4 -10 -4 5 16 20 -12 -1 -9 11 -11 12 3 -1 0 3 21 3 -2 10 11 -11 15 17 2 4 8 -19 28 28 29 1 21 27 -15 25 -15 10 10 26 -9 -13 7 -6 9 27 -3 2 -12 0 20 26 -14 15 29 10 30 -17 25 21 -6 26 25 24 6 9 29 -13 27 16 -5 27 25 -3 -20 3 12 27 5 29 3 -13 9 6 12 14 -14 -17 -8 -13 17 6 20 0 -20 -7 -4 -12 18 7 11 3 -8 23 0 -6 27 6 -20 6 1 15 -14 -20 -3 22 -13 6 10 10 -18 -6 -1 -14 8 16 -4 -12 -12 -11 30 11 -4 -11 10 -15 23 -19 27 -17 -3 -3 13 12 -19 5 22 -8 19 21 -6 21 -20 -6 4 -20 -5 17 0 2 13 24 22 6 20 -3 17 -12 24 -18 5 -12 -9 -17 17 -3 29 9 -12 1 -17 -4 8 2 17 -20 28 -15 -11 26 16 19 4 23 -18 22 -15 1 16 14 30 3 12 -5 27 14 -5 29 5 13 16 -3 4 -9 -20 11 11 12 13 2 13 -2 -15 3 4 -20 28 -8 4 28 5 28 16 -3 3 19 17 6 -5 6 -15 6 15 -8 29 22 -3 -5 -5 16 24 14 26 -19 13 -17 22 16 -20 -7 14 -11 -10 -19 30 14 28 20 14 12 -11 -19 -4 -19 -20 22 2 -5 9 -10 -4 -12 15 8 22 27 5 9 18 -13 27 27 -12 8 28 29 -12 -9 29 -15 -7 7 18 -6 -4 -15 -11 23 25 27 -8 3 22 29 7 16 -8 11 -17 -9 5 22 4 -17 7 -18 2 0 -8 2 0 3 -6 23 -3 -17 21 9 -6 7 3 20 9 24 -11 21 1 25 30 -17 -11 12 -13 8 7 -12 -17 0 -1 19 20 -11 13 -6 -14 6 -4 27 6 -18 -4 21 20 -11 21 16 -15 2 -8 8 24 24 -15 -12 -5 -20 -15 27 29 25 -19 -20 14 -16 7 18 -15 25 11 -12 -5 1 -4 12 14 -3 -11 14 -14 25 4 21 -2 -1 -17 23 -3 -9 19 26 8 -18 2 3 20 19 -15 -20 3 26 -3 11 -14 8 25 -16 8 8 2 23 29 -10 28 -16 13 -1 11 22 -20 8 18 2 -16 -17 18 30 0 -19 23 -6 -10 11 24 16 27 19 -16 -17 -9 -8 -5 1 5 0 15 9 24 10 18 10 -7 -9 12 10 21 27 -10 4 -13 -17 -1 3 -17 0 -7 -20 20 29 -13 -11 -4 -12 11 3 25 26 20 27 14 29 17 19 -2 8 17 -8 -8 14 -15 -5 0 23 0 23 22 24 13 1 25 -17 -16 7 18 4 24 -9 10 -12 17 8 28 3 29 -2 26 -8 -3 22 23 -19 1 18 -8 -4 21 29 29 6 -5 5 24 -9 -12 -9 -10 3 5 0 -7 23 22 13 11 16 -20 28 -8 -17 -3 13 13 24 8 13 14 6 -5 -3 20 24 -19 14 -9 -6 16 3 28 -20 -18 11 -18 15 7 29 5 11 -14 21 6 16 -4 16 -10 -17 7 10 2 25 -16 1 16 -19 20 5 1 -3 5 11 21 21 -1 6 -11 -7 28 -8 8 -10 3 24 -9 -4 21 -6 1 -1 11 -13 -12 20 20 -13 -14 -20 -19 26 15 -16 -5 -9 10 4 5 -10 23 14 22 9 -16 16 21 9 15 3 30 6 8 22 25 24 22 1 -16 -3 19 6 -4 -7 -6 25 -13 13 15 21 10 30 -12 19 -1 -2 -19 14 29 -16 12 -17 -8 -12 5 8 25 18 22 7 14 13 25 -20\n",
    "Output\n",
    "GGPGGPGGAGG\n",
    "'''\n",
    "\n",
    "class PeptideSequencing:\n",
    "    def __init__(self):\n",
    "        massDict, aaDict = self.AminoAcidMassDict()\n",
    "        spectralVector = self.readFromFile()\n",
    "        peptide = self.findPeptide(spectralVector, massDict)\n",
    "        print(peptide)\n",
    "        f = open('result.txt', 'w')\n",
    "        f.write(peptide)\n",
    "        f.close()\n",
    "        \n",
    "    def AminoAcidMassDict(self):\n",
    "        massTable = '''\n",
    "G 57\n",
    "A 71\n",
    "S 87\n",
    "P 97\n",
    "V 99\n",
    "T 101\n",
    "C 103\n",
    "I 113\n",
    "L 113\n",
    "N 114\n",
    "D 115\n",
    "K 128\n",
    "Q 128\n",
    "E 129\n",
    "M 131\n",
    "H 137\n",
    "F 147\n",
    "R 156\n",
    "Y 163\n",
    "W 186'''\n",
    "        mass = massTable.split()\n",
    "        return {int(mass[i+1]):mass[i] for i in range(0, len(mass), 2)}, {mass[i]:int(mass[i+1]) for i in range(0, len(mass), 2)}\n",
    "    \n",
    "    def _input(self):\n",
    "        data = sys.stdin.read().strip().split()\n",
    "        spectralVector = list(map(int, data))\n",
    "        spectralVector.insert(0, 0)\n",
    "        return spectralVector\n",
    "    \n",
    "    def readFromFile(self):\n",
    "        f = open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/peptide_sequencing.txt', 'r')\n",
    "        for line in f:\n",
    "            data = line.strip().split()\n",
    "        spectralVector = list(map(int, data))\n",
    "        spectralVector.insert(0, 0)\n",
    "        return spectralVector\n",
    "    \n",
    "    def findPeptide(self, spectralVector, massDict):\n",
    "        l = len(spectralVector)\n",
    "        adj = [[] for _ in range(l)]\n",
    "        for i in range(l):\n",
    "            for j in range(i, l):\n",
    "                if j-i in massDict:\n",
    "                    adj[i].append(j)\n",
    "        \n",
    "        # Bellman-Ford algorithm\n",
    "        dist = [-np.inf] * l\n",
    "        parent = [None] * l\n",
    "        dist[0] = 0\n",
    "        updated = True\n",
    "        for i in range(l-1):\n",
    "            if not updated:\n",
    "                break\n",
    "            updated = False\n",
    "            for u in range(l):\n",
    "                for v in adj[u]:\n",
    "                    if dist[u] + spectralVector[v] > dist[v]:\n",
    "                        dist[v] = dist[u] + spectralVector[v]\n",
    "                        parent[v] = u\n",
    "                        updated = True\n",
    "        u = l-1\n",
    "        path = [u]\n",
    "        while 0 != u:\n",
    "            u = parent[u]\n",
    "            path.insert(0, u)\n",
    "\n",
    "        peptide = ''.join([massDict[path[i+1]-path[i]] for i in range(len(path)-1)])\n",
    "        return peptide\n",
    "    \n",
    "PeptideSequencing()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Despite many attempts, researchers have still not devised a scoring function that reliably assigns the highest score to the \n",
    "biologically correct peptide, i.e., the peptide that generated the spectrum. Fortunately, although the correct peptide often \n",
    "does not achieve the highest score among all peptides, it typically does score highest among all peptides limited to the \n",
    "species’s proteome. As a result, we can transition from peptide sequencing to peptide identification by limiting our search to \n",
    "peptides present in the proteome, which we concatenate into a single amino acid string Proteome.\n",
    "Peptide Identification Problem: Find a peptide from a proteome with maximum score against a spectrum.\n",
    "     Input: A spectral vector Spectrum and an amino acid string Proteome.\n",
    "     Output: An amino acid string Peptide that maximizes Score(Peptide, Spectrum) among all substrings of Proteome.\n",
    "Sample Input:\n",
    "0 0 0 4 -2 -3 -1 -7 6 5 3 2 1 9 3 -8 0 3 1 2 1 8\n",
    "XZZXZXXXZXZZXZXXZ\n",
    "Sample Output:\n",
    "ZXZXX\n",
    "\n",
    "'''\n",
    "massTable = {'A': 71, 'C': 103, 'E': 129, 'D': 115, 'G': 57, 'F': 147, \\\n",
    "            'I': 113, 'H': 137, 'K': 128, 'M': 131, 'L': 113, 'N': 114, \\\n",
    "            'Q': 128, 'P': 97, 'S': 87, 'R': 156, 'T': 101, 'W': 186, \\\n",
    "            'V': 99, 'Y': 163,'X':4,'Z':5}\n",
    "\n",
    "def get_mass(peptide):\n",
    "    if len(peptide) == 0:\n",
    "        return 0\n",
    "    return sum( massTable[pep] for pep in peptide )\n",
    "\n",
    "def peptide_vector(peptide):\n",
    "    mass = []\n",
    "    for i in range(1,len(peptide)+1):\n",
    "        mass.append(get_mass(peptide[:i]))\n",
    "    pepVec = [0] * max(mass)\n",
    "    for mi in mass:\n",
    "        pepVec[mi-1] = 1\n",
    "    return pepVec\n",
    "\n",
    "def peptide_identification(proteome,specVec):\n",
    "    massSpec = len(specVec)\n",
    "    peptides = []\n",
    "    for i in range(len(proteome)):\n",
    "        for j in range(i+1,len(proteome)):\n",
    "            masspep = get_mass(proteome[i:j])\n",
    "            if masspep == massSpec:\n",
    "                pepVec = peptide_vector(proteome[i:j])\n",
    "                if len(pepVec) == len(specVec):\n",
    "                    peptides.append(proteome[i:j])\n",
    "            elif masspep > massSpec:\n",
    "                #if the prefix is already larger, then dont need to consider the rest\n",
    "                break\n",
    "    if len(peptides) != 0:\n",
    "        maxScore = float('-inf')\n",
    "        for peptide in peptides:\n",
    "            pepVec = peptide_vector(peptide)\n",
    "            score = sum([pepVec[i]*specVec[i] for i in range(len(pepVec))])\n",
    "            if score > maxScore:\n",
    "                maxScore = score\n",
    "                pepHit = peptide\n",
    "    else:\n",
    "        pepHit = []\n",
    "    return pepHit\n",
    "\n",
    "specVec = list(map(int,'0 0 0 4 -2 -3 -1 -7 6 5 3 2 1 9 3 -8 0 3 1 2 1 8'.split()))\n",
    "proteome = 'XZZXZXXXZXZZXZXXZ'\n",
    "\n",
    "peptide_identification(proteome,specVec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Like peptide sequencing algorithms, peptide identification algorithms may return an erroneous peptide, particularly if the \n",
    "score of the highest-scoring peptide found in the proteome is much lower than the score of the highest-scoring peptide over \n",
    "all peptides. For this reason, biologists usually establish a score threshold and only pay attention to a solution of the \n",
    "Peptide Identification Problem if its score is at least equal to the threshold.\n",
    "Given a set of spectral vectors SpectralVectors, an amino acid string Proteome, and a score threshold threshold, we will solve \n",
    "the Peptide Identification Problem for each vector Spectrum' in SpectralVectors and identify a peptide Peptide having maximum \n",
    "score for this spectral vector over all peptides in Proteome (ties are broken arbitrarily). If Score(Peptide, Spectrum) is \n",
    "greater than or equal to threshold, then we conclude that Peptide is present in the sample and call the pair \n",
    "(Peptide, Spectrum') a Peptide- Spectrum Match (PSM). The resulting collection of PSMs for SpectralVectors is denoted \n",
    "PSMthreshold(Proteome, SpectralVectors).\n",
    "PSM Search Problem: Identify all Peptide-Spectrum Matches scoring above a threshold for a set of spectra and a proteome.\n",
    "     Input: A set of spectral vectors SpectralVectors, an amino acid string Proteome, and an integer threshold.\n",
    "     Output: The set PSMthreshold(Proteome, SpectralVectors).\n",
    "Sample Input:\n",
    "-1 5 -4 5 3 -1 -4 5 -1 0 0 4 -1 0 1 4 4 4\n",
    "-4 2 -2 -4 4 -5 -1 4 -1 2 5 -3 -1 3 2 -3\n",
    "XXXZXZXXZXZXXXZXXZX\n",
    "5 \n",
    "Sample Output:\n",
    "XZXZ\n",
    "PSMSearch(SpectralVectors, Proteome, threshold).\n",
    "    PSMSet ← an empty set\n",
    "    for each vector Spectrum' in SpectralVectors\n",
    "          Peptide ← PeptideIdentification(Spectrum', Proteome)\n",
    "          if Score(Peptide, Spectrum) ≥ threshold\n",
    "              add the PSM (Peptide, Spectrum') to PSMSet\n",
    "    return PSMSet\n",
    "    \n",
    "'''\n",
    "massTable = {'A': 71, 'C': 103, 'E': 129, 'D': 115, 'G': 57, 'F': 147, \\\n",
    "            'I': 113, 'H': 137, 'K': 128, 'M': 131, 'L': 113, 'N': 114, \\\n",
    "            'Q': 128, 'P': 97, 'S': 87, 'R': 156, 'T': 101, 'W': 186, \\\n",
    "            'V': 99, 'Y': 163,'X':4,'Z':5}\n",
    "\n",
    "def get_mass(peptide):\n",
    "    if len(peptide) == 0:\n",
    "        return 0\n",
    "    return sum( massTable[pep] for pep in peptide )\n",
    "\n",
    "def peptide_vector(peptide):\n",
    "    mass = []\n",
    "    for i in range(1,len(peptide)+1):\n",
    "        mass.append(get_mass(peptide[:i]))\n",
    "    pepVec = [0] * max(mass)\n",
    "    for mi in mass:\n",
    "        pepVec[mi-1] = 1\n",
    "    return pepVec\n",
    "\n",
    "def peptide_identification(proteome,specVec):\n",
    "    massSpec = len(specVec)\n",
    "    peptides = []\n",
    "    for i in range(len(proteome)):\n",
    "        for j in range(i+1,len(proteome)):\n",
    "            masspep = get_mass(proteome[i:j])\n",
    "            if masspep == massSpec:\n",
    "                pepVec = peptide_vector(proteome[i:j])\n",
    "                if len(pepVec) == len(specVec):\n",
    "                    peptides.append(proteome[i:j])\n",
    "            elif masspep > massSpec:\n",
    "                #if the prefix is already larger, then dont need to consider the rest\n",
    "                break\n",
    "    if len(peptides) != 0:\n",
    "        maxScore = float('-inf')\n",
    "        for peptide in peptides:\n",
    "            pepVec = peptide_vector(peptide)\n",
    "            score = sum([pepVec[i]*specVec[i] for i in range(len(pepVec))])\n",
    "            if score > maxScore:\n",
    "                maxScore = score\n",
    "                pepHit = peptide\n",
    "    else:\n",
    "        pepHit = []\n",
    "    return pepHit\n",
    "\n",
    "def PSMSearch(SpectralVectors, Proteome, threshold):\n",
    "    PSMset = []\n",
    "    for specVec in SpectralVectors:\n",
    "        specVec = list(map(int,specVec.split()))\n",
    "        peptide = peptide_identification(Proteome,specVec)\n",
    "        print(peptide)\n",
    "        if len(peptide) != 0:\n",
    "            pepVec = peptide_vector(peptide)\n",
    "            score = sum([pepVec[i]*specVec[i] for i in range(len(pepVec))])\n",
    "            if score >= threshold:\n",
    "                PSMset.append(peptide)\n",
    "    return set(PSMset)\n",
    "\n",
    "specVecs = ['-1 5 -4 5 3 -1 -4 5 -1 0 0 4 -1 0 1 4 4 4'\\\n",
    ",'0 -4 2 -2 -4 4 -5 -1 4 -1 2 5 -3 -1 3 2 -3']\n",
    "proteome = 'XXXZXZXXZXZXXXZXXZX'\n",
    "threshold = 5\n",
    "\n",
    "peptides = PSMSearch(specVecs, proteome, threshold)\n",
    "print('\\n'.join(peptides))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Use this for PSM search\n",
    "class PSMSearch:\n",
    "    def __init__(self):\n",
    "        massDict, aaDict = self.AminoAcidMassDict()\n",
    "        spectralVectorSet, threshold, proteome = self.readFromFile()\n",
    "        PSMSet = self.searchPSM(spectralVectorSet, threshold, proteome, aaDict)\n",
    "        for peptide in PSMSet:\n",
    "            print(peptide)\n",
    "        f = open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/result.txt', 'w')\n",
    "        for peptide in PSMSet:\n",
    "            f.write(peptide+'\\n')\n",
    "        f.close()\n",
    "\n",
    "    def AminoAcidMassDict(self):\n",
    "        massTable = '''\n",
    "G 57\n",
    "A 71\n",
    "S 87\n",
    "P 97\n",
    "V 99\n",
    "T 101\n",
    "C 103\n",
    "I 113\n",
    "L 113\n",
    "N 114\n",
    "D 115\n",
    "K 128\n",
    "Q 128\n",
    "E 129\n",
    "M 131\n",
    "H 137\n",
    "F 147\n",
    "R 156\n",
    "Y 163\n",
    "W 186'''\n",
    "        mass = massTable.split()\n",
    "        return {int(mass[i+1]):mass[i] for i in range(0, len(mass), 2)}, {mass[i]:int(mass[i+1]) for i in range(0, len(mass), 2)}\n",
    "\n",
    "    def readFromFile(self):\n",
    "        f = open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/trial.txt', 'r')\n",
    "        data = []\n",
    "        for line in f:\n",
    "            data.append(line.strip().split())\n",
    "        threshold = int(data[-1][0])\n",
    "        proteome = data[-2][0]\n",
    "        spectralVectorSet = [[0]+list(map(int, d)) for d in data[:-2]]\n",
    "        return spectralVectorSet, threshold, proteome \n",
    "\n",
    "    def findPeptide(self, spectralVector, proteome, aaDict):\n",
    "        l = len(spectralVector) - 1\n",
    "        n = len(proteome)\n",
    "        score = -np.inf\n",
    "        peptide = ''\n",
    "        for i in range(n-1):\n",
    "            for j in range(i, n):\n",
    "                if sum([aaDict[aa] for aa in proteome[i:j+1]]) > l:\n",
    "                    break\n",
    "                if sum([aaDict[aa] for aa in proteome[i:j+1]]) != l:\n",
    "                    continue\n",
    "                prefixMasses = [sum([aaDict[aa] for aa in proteome[i:i+k+1]]) for k in range(j-i+1)]\n",
    "                currScore = sum([spectralVector[m] for m in prefixMasses])\n",
    "                if currScore > score:\n",
    "                    score = currScore\n",
    "                    peptide = proteome[i:j+1]\n",
    "        return score, peptide\n",
    "\n",
    "    def searchPSM(self, spectralVectorSet, threshold, proteome, aaDict):\n",
    "        PSMSet = []\n",
    "        for spectralVector in spectralVectorSet:\n",
    "            score, peptide = self.findPeptide(spectralVector, proteome, aaDict)\n",
    "            if score >= threshold:\n",
    "                if not peptide in PSMSet:\n",
    "                    PSMSet.append(peptide)\n",
    "        return PSMSet\n",
    "\n",
    "PSMSearch()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Define Pr(i, t) as the sum of probabilities of all peptides with mass i for which Score(Peptide, Spectrum'i ) is equal to t. \n",
    "The set of peptides contributing to Pr(i, t) can be split into 20 subsets depending on their final amino acid. Each peptide \n",
    "Peptide ending in a specific amino acid a results in a shorter peptide Peptidea if we remove a; Peptidea has mass i − |a| and \n",
    "score t − si. Since the probability of Peptide is 20 times smaller than the probability of Peptidea, the contribution of Peptide \n",
    "to Pr(i, t) is 20 times smaller than contribution of Peptidea to Pr(i − |a|, t − si ). Therefore, Pr(i, t) can be computed as\n",
    "Pr(i,t)=∑all amino acids a120⋅Pr(i−|a|,t−si),\n",
    "which differs from the recurrence for computing Size(i, t) only in the presence of the factor 1/20. We can now compute the \n",
    "probability of a spectral dictionary as \n",
    "Pr(Dictionarythreshold(Spectrum'))=∑t≥thresholdPr(m,t).\n",
    "Solve the Probability of Spectral Dictionary Problem.\n",
    "     Given: A spectral vector Spectrum', an integer threshold, and an integer max_score.\n",
    "     Return: The probability of the dictionary Dictionarythreshold(Spectrum').\n",
    "Note: Use the provided max_score for the height of your table.\n",
    "Input - trial.txt --\n",
    "-10 11 3 10 11 12 -6 -5 4 4 -2 9 6 -8 9 -6 -1 10 -6 14 4 13 1 -6 5 -7 13 0 -1 12 -2 11 7 -10 9 13 14 -7 7 -9 -6 4 14 2 -9 1 12 13 15 6 15 13 -6 -10 -10 -8 -8 -7 -10 -7 -6 -4 6 9 -6 7 11 -1 -8 1 9 -5 6 7 -3 -10 -9 -1 4 7 7 -6 14 -6 12 15 7 8 11 -5 8 -8 12 -3 -1 -7 -6 9 13 12 -3 7 7 6 3 1 2 4 10 11 -10 -3 14 9 6 8 -9 1 5 -6 -8 5 -7 6 -6 -7 4 1 -3 7 5 10 11 12 0 -10 12 13 11 3 9 8 -10 9 -8 0 15 4 1 1 -4 12 2 4 0 15 -10 4 -10 -10 6 -5 -5 0 10 -5 8 1 14 6 -3 12 9 -7 -4 -9 -9 7 2 6 4 -10 -9 8 -4 -5 0 7 -4 -3 5 12 -10 3 -6 -10 6 10 -6 3 -5 15 4 14 -1 10 -9 13 11 -7 -5 -3 14 15 6 -3 -8 -5 0 12 0 12 2 8 -1 6 2 4 -6 3 11 -4 -10 1 -5 0 14 -5 -6 -1 15 13 12 -10 6 4 0 14 -1 5 15 13 4 -6 13 12 7 14 6 15 10 -9 1 -8 10 9 6 6 2 9 -2 5 11 -4 -6 -10 -7 10 9 8 -6 1 -8 2 -1 -1 -4 -2 0 9 11 -6 9 11 5 5 14 7 -10 14 -4 7 4 14 14 14 8 2 5 14 -4 13 7 10 14 -7 -6 11 -7 -2 -6 -3 1 -7 7 10 15 -6 -2 0 14 1 9 -7 5 -3 -5 5 -5 0 -4 1 3 11 9 -4 -3 -4 0 1 -4 15 -8 -3 0 0 11 -9 11 5 -9 1 -1 -7 -3 8 -9 11 5 4 4 -7 11 -1 -4 -5 7 -7 7 3 6 13 -1 11 -3 13 11 4 3 2 3 0 12 -6 3 12 -10 -8 -9 12 -2 12 5 -3 5 11 5 1 -2 3 5 1 11 6 -6 -2 0 -7 15 14 15 -10 0 6 13 9 10 -2 10 2 8 6 -6 5 -2 1 13 8 14 1 -4 11 11 -8 0 8 5 5 9 -1 -7 3 15 -7 -8 -3 11 9 0 10 2 1 13 4 0 -6 15 15 -1 10 3 1 2\n",
    "30\n",
    "200\n",
    "Output\n",
    "0.00132187890625\n",
    "'''\n",
    "\n",
    "class SpectralDictProb:\n",
    "    def __init__(self):\n",
    "        massList = self.AminoAcidMassList()\n",
    "        sVector, threshold, maxScore = self.readFromFile()\n",
    "        p = self.dictProb(sVector, threshold, maxScore, massList)\n",
    "        print(p)\n",
    "        f = open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/result.txt', 'w')\n",
    "        f.write(str(p))\n",
    "        f.close()\n",
    "        \n",
    "    def AminoAcidMassList(self):\n",
    "        massTable = '''\n",
    "G 57\n",
    "A 71\n",
    "S 87\n",
    "P 97\n",
    "V 99\n",
    "T 101\n",
    "C 103\n",
    "I 113\n",
    "L 113\n",
    "N 114\n",
    "D 115\n",
    "K 128\n",
    "Q 128\n",
    "E 129\n",
    "M 131\n",
    "H 137\n",
    "F 147\n",
    "R 156\n",
    "Y 163\n",
    "W 186\n",
    "'''\n",
    "        mass = massTable.split()\n",
    "        return [int(mass[i+1]) for i in range(0, len(mass), 2)]\n",
    "\n",
    "    def readFromFile(self):\n",
    "        f = open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/dataset_11866_11.txt', 'r')\n",
    "        data = []\n",
    "        for line in f:\n",
    "            data.append(line.strip().split())\n",
    "        sVector = [0] + list(map(int, data[0]))\n",
    "        threshold = int(data[1][0])\n",
    "        maxScore = int(data[2][0])\n",
    "        return sVector, threshold, maxScore\n",
    "    \n",
    "    def dictProb(self, sVector, threshold, maxScore, massList):\n",
    "        prob = dict()\n",
    "        prob[(0, 0)] = 1\n",
    "        p = sum([self.getProb(len(sVector)-1, t, sVector, massList, prob) for t in range(threshold, maxScore+1)])\n",
    "        return p\n",
    "    \n",
    "    def getProb(self, i, t, sVector, massList, prob):\n",
    "        if (i, t) in prob:\n",
    "            return prob[(i, t)]\n",
    "        if i < 0 or t < 0:\n",
    "            prob[(i, t)] = 0\n",
    "            return 0\n",
    "        p = sum([self.getProb(i-m, t-sVector[i], sVector, massList, prob)/20 for m in massList])\n",
    "        prob[(i, t)] = p\n",
    "        return p\n",
    "    \n",
    "SpectralDictProb()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#or u can use this as well for spectral dict probability\n",
    "massTable = {'A': 71, 'C': 103, 'E': 129, 'D': 115, 'G': 57, 'F': 147, \\\n",
    "            'I': 113, 'H': 137, 'K': 128, 'M': 131, 'L': 113, 'N': 114, \\\n",
    "            'Q': 128, 'P': 97, 'S': 87, 'R': 156, 'T': 101, 'W': 186, \\\n",
    "            'V': 99, 'Y': 163}\n",
    "\n",
    "#massTable = {'X':4,'Z':5}\n",
    "\n",
    "def get_size_spectral_dictionary(spectrum,threshold,maxscore):\n",
    "    size = []\n",
    "    for t in range(maxscore+1):\n",
    "        size.append([0]*(len(spectrum)))\n",
    "    size[0][0] = 1\n",
    "\n",
    "    for i in range(1,len(spectrum)):\n",
    "        for t in range(maxscore+1):\n",
    "            total = 0\n",
    "            for aa in massTable.keys():\n",
    "                m = massTable[aa]\n",
    "                if i-m>=0 and t-spectrum[i]<=maxscore and t-spectrum[i]>=0:\n",
    "                    total = total + size[t-spectrum[i]][i-m]\n",
    "            size[t][i] = float(total)/20 # divided by 20 means \n",
    "    ret = 0\n",
    "    for i in range(len(spectrum)-1,len(spectrum)):\n",
    "        for t in range(maxscore+1):\n",
    "            if t >= threshold and t<=maxscore:\n",
    "                ret += size[t][i]\n",
    "    return ret\n",
    "\n",
    "with open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/trial.txt') as f:\n",
    "    lines = f.read().splitlines()\n",
    "spectrum = lines[0]\n",
    "threshold = int(lines[1])\n",
    "maxscore = int(lines[2])\n",
    "\n",
    "spectrum = list(map(int,[0] + spectrum.split(' ')))\n",
    "specSize = get_size_spectral_dictionary(spectrum,threshold,maxscore)\n",
    "print(specSize)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Size of Spectral Dictionary Problem: Find the size of a spectral dictionary for a given spectrum and score threshold.\n",
    "     Input: A spectral vector Spectrum' and an integer threshold.\n",
    "     Output: The number of peptides in Dictionarythreshold(Spectrum').\n",
    "We will use dynamic programming to solve the Size of Spectral Dictionary Problem. Given a spectral vector Spectrum' = \n",
    "(s1, . . . , sm), we define its i-prefix (for i between 1 and m) as Spectrum'i = (s1, . . . , si ) and introduce a variable \n",
    "Size(i, t) as the number of peptides Peptide of mass i such that Score(Peptide, Spectrum'i) is equal to t.\n",
    "The key to establishing a recurrence relation for computing Size(i, t) is to realize that the set of peptides contributing \n",
    "to Size(i, t) can be split into 20 subsets depending on their final amino acid a. Each peptide ending in a specific amino \n",
    "acid a results in a shorter peptide with mass i − |a| and score t − si   if we remove a from the peptide (here, |a| denotes \n",
    "the mass of a). Thus,\n",
    "Size(i,t)=∑all amino acids aSize(i−|a|,t−si).\n",
    "Since there is a single “empty” peptide of length zero, we initialize Size(0, 0) = 1. We also define Size(0, t) = 0 for all \n",
    "possible scores t, and set Size(i, t) = 0 for negative values of i. Using the above recurrence, we can compute the size of a \n",
    "spectral dictionary of Spectrum' = (s1, . . . , sm) as\n",
    "|Dictionarythreshold(Spectrum)|=∑t≥thresholdSize(m,t).\n",
    "Solve the Size of Spectral Dictionary Problem.\n",
    "     Given: A spectral vector Spectrum', an integer threshold, and an integer max_score.\n",
    "     Return: The size of the dictionary Dictionarythreshold(Spectrum').\n",
    "Note: Use the 20 amino acid alphabet as well as the provided max_score for the height of your table. Your answer should be the \n",
    "number of peptides whose score is at least T and at most max_score.\n",
    "\n",
    "Input -- trial.txt\n",
    "14 -4 -3 -3 5 9 0 14 2 1 -4 6 -1 13 2 -5 13 -8 -8 3 0 -10 14 4 14 14 8 -8 1 3 -10 -2 2 -9 3 6 13 -10 6 -8 12 2 8 -1 -5 -6 -6 10 3 -3 12 -4 14 3 11 14 15 12 -7 -5 -2 11 13 -9 15 -8 -10 5 -8 5 6 -9 -2 7 -6 -1 -2 12 12 -3 9 0 3 0 5 -6 3 3 -7 6 0 -6 5 8 -7 5 3 13 13 -2 -9 0 2 13 13 12 7 -2 -10 -5 -7 7 13 11 14 -4 -9 15 -10 5 -7 -6 -7 -6 11 5 9 8 -4 7 1 -9 12 2 8 12 -6 0 2 -5 10 11 14 15 -1 3 -3 3 -3 12 15 4 -2 14 13 8 -10 2 -3 0 -6 8 3 10 0 9 10 13 15 6 9 -10 -9 1 -3 -10 8 1 -10 2 -1 14 -3 15 -1 0 1 6 -7 5 12 6 -9 2 1 -2 14 -5 1 -8 -6 11 -5 2 -3 -8 7 -6 -10 8 6 13 -8 -5 -10 12 -5 -9 8 9 0 10 15 -1 4 2 -8 9 1 -9 -6 -8 -1 -1 5 10 -4 7 3 11 4 12 6 6 13 -3 12 -3 1 7 11 6 13 8 3 -6 5 11 4 -1 15 10 -8 -7 0 4 7 5 -4 8 -3 -4 -8 9 -2 -3 13 1 12 4 -1 13 -1 -5 -5 7 7 -7 -5 6 6 -2 -5 7 10 14 11 12 -9 6 -3 4 15 -8 11 -3 -7 5 -4 7 9 15 -9 8 13 6 -2 -3 9 6 5 14 10 -7 -9 -8 10 2 -3 -1 2 3 12 13 6 -2 8 -5 5 -3 -8 10 3 0 12 -7 10 6 15 8 7 -2 8 14 -2 13 -1 8 15 -7 -7 -7 7 -3 -2 5 -4 -3 15 11 -4 9 11 13 15 8 4 -6 7 12 14 6 -10 -5 -9 4 -9 13 -3 0 12 3 12 -5 11 1 15 -8 5 3 -5 7 15 -2 -9 0 0 1 1 -1 -4 -1 5 12 12 -5 8 5 14 12 5 -9 2 -10 -9 4 -2 6 5 -3 -7 7 5 -8 -10 8 -1 7 3 6 -6 14 -8 6 -5 -8 -10 14 -2 12 4 5 -2 9 -4 1 -5 -3 -6 -8 -9 -10 10 4 9 11 -7 6 -4 5 13 -8 -7 -3 10 14 4 10 6 4 0 13 -3 11 -9 2 -8 6 -8 4 -1\n",
    "37\n",
    "200\n",
    "Output\n",
    "330\n",
    "'''\n",
    "\n",
    "class SpectralDictSize:\n",
    "    def __init__(self):\n",
    "        massList = self.AminoAcidMassList()\n",
    "        sVector, threshold, maxScore = self.readFromFile()\n",
    "        s = self.dictSize(sVector, threshold, maxScore, massList)\n",
    "        print(s)\n",
    "        \n",
    "    def AminoAcidMassList(self):\n",
    "        massTable = '''\n",
    "G 57\n",
    "A 71\n",
    "S 87\n",
    "P 97\n",
    "V 99\n",
    "T 101\n",
    "C 103\n",
    "I 113\n",
    "L 113\n",
    "N 114\n",
    "D 115\n",
    "K 128\n",
    "Q 128\n",
    "E 129\n",
    "M 131\n",
    "H 137\n",
    "F 147\n",
    "R 156\n",
    "Y 163\n",
    "W 186'''\n",
    "        mass = massTable.split()\n",
    "        return [int(mass[i+1]) for i in range(0, len(mass), 2)]\n",
    "\n",
    "    def readFromFile(self):\n",
    "        f = open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/trial.txt', 'r')\n",
    "        data = []\n",
    "        for line in f:\n",
    "            data.append(line.strip().split())\n",
    "        sVector = [0] + list(map(int, data[0]))\n",
    "        threshold = int(data[1][0])\n",
    "        maxScore = int(data[2][0])\n",
    "        return sVector, threshold, maxScore\n",
    "    \n",
    "    def dictSize(self, sVector, threshold, maxScore, massList):\n",
    "        size = dict()\n",
    "        size[(0, 0)] = 1\n",
    "        s = sum([self.getSize(len(sVector)-1, t, sVector, massList, size) for t in range(threshold, maxScore+1)])\n",
    "        return s\n",
    "    \n",
    "    def getSize(self, i, t, sVector, massList, size):\n",
    "        if (i, t) in size:\n",
    "            return size[(i, t)]\n",
    "        if i < 0 or t < 0:\n",
    "            size[(i, t)] = 0\n",
    "            return 0\n",
    "        s = sum([self.getSize(i-m, t-sVector[i], sVector, massList, size) for m in massList])\n",
    "        size[(i, t)] = s\n",
    "        return s\n",
    "    \n",
    "SpectralDictSize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import queue\n",
    "import numpy as np\n",
    "from copy import deepcopy\n",
    "\n",
    "'''\n",
    "We will use the term block indel to refer to the addition or removal of a block of consecutive zeroes from a binary vector. \n",
    "Thus, applying k modifications to an amino ￼acid string Peptide corresponds to applying k block indels to its peptide vector \n",
    "Peptide'. We define Variantsk(Peptide) as the set of all modified variants of Peptide with up to k modifications.\n",
    "Given a peptide Peptide and a spectral vector Spectrum', our goal is to find a modified peptide from Variantsk(Peptide) with \n",
    "maximum score against Spectrum'.\n",
    "Spectral Alignment Problem: Given a peptide and a spectral vector, find a modified variant of this peptide that maximizes the \n",
    "peptide-spectrum score, among all variants of the peptide with up to k modifications.\n",
    "     Input: An amino acid string Peptide, a spectral vector Spectrum', and an integer k.\n",
    "     Output: A peptide of maximum score against Spectrum' among all peptides in Variantsk(Peptide).\n",
    "     \n",
    "Input -- trial.txt\n",
    "LVWSTE\n",
    "-3 1 12 1 5 12 -4 10 0 9 -4 -9 -6 14 -10 8 11 -3 3 2 11 12 -10 -4 15 -5 2 1 4 -10 -10 7 9 10 8 -9 15 13 -9 7 7 1 13 15 0 -3 -2 7 6 6 7 15 -5 10 7 -4 3 11 14 10 -10 2 -8 12 0 0 -6 15 1 14 3 12 4 11 -4 -7 5 3 8 -10 3 7 11 8 5 15 7 13 3 1 -1 -1 9 5 7 3 9 10 3 -9 10 6 13 8 5 2 -5 14 3 14 12 -7 11 -3 15 11 -8 10 -4 14 -4 5 -2 8 -3 2 -7 12 -10 0 -3 14 0 6 14 0 14 7 -5 3 -3 -1 -3 7 8 -4 2 -6 10 5 13 -9 -3 10 1 4 0 11 12 5 -10 -6 1 -9 1 13 -3 7 -5 6 0 5 3 -9 3 -4 1 -6 -4 11 -10 8 -2 15 6 5 -7 12 3 11 5 15 -8 6 -4 8 10 -7 12 -3 -4 10 11 -4 -3 9 0 -5 1 -6 11 7 -3 -7 -9 13 10 -7 -6 14 15 -10 14 15 -8 5 1 10 -5 2 -7 4 14 10 7 -1 1 12 -7 7 -8 -1 12 3 -9 4 -7 6 8 10 8 5 6 12 4 4 4 12 -4 6 7 -10 7 -7 0 -1 -2 6 -10 -5 -4 5 -3 -1 1 2 10 -9 -6 14 9 -3 -8 -2 -4 -1 0 -10 6 10 -6 -3 4 -1 -7 0 6 3 -3 5 4 -4 0 3 0 -1 -1 -4 -10 2 -3 -9 -10 5 7 10 5 2 -8 2 -6 13 6 -7 5 5 -5 15 8 11 12 4 8 5 11 3 -7 14 3 8 0 2 -4 6 -7 1 8 13 -1 14 3 -9 12 -3 12 14 0 -2 11 11 13 3 15 -9 -5 2 1 -10 7 -8 -10 -6 -7 1 0 7 9 5 -9 5 0 -3 -2 13 14 -10 4 3 5 13 -3 -4 -5 11 14 5 1 -10 -1 4 -6 13 -6 1 -5 2 -2 3 12 0 -7 15 10 -10 11 7 0 9 -7 -5 9 2 -8 -9 -5 -6 13 7 11 -3 10 1 -2 0 6 -2 -4 5 5 -4 15 11 4 -7 10 -1 9 -2 5 13 -10 14 7 6 4 -8 1 12 12 2 5 13 7 6 -4 -2 7 4 -8 -7 8 14 -8 14 -2 -5 -9 -5 11 10 7 3 1 11 1 14 4 -9 14 11 12 -1 -3 -7 10 13 -7 4 6 8 2 1 -3 -3 -9 11 7 -1 5 -5 -6 -9 2 11 8 13 1 -6 7 5 3 2 2 6 12 -9 12 3 -3 -1 -10 -4 -4 10 10 14 12 -2 7 14 9 3 10 -5 5 -7 4 9 1 13 10 13 13 -8 6 4 10 -4 15 -9 -1 14 9 -10 3 15 0 -3 2 4 -5 1 8 3 15 12 3 0 9 5 1 3 5 14 -6 13 0 13 -7 4 -9 -2 -9 -3 8 8 13 13 8 4 -2 0 5 -9 -2 0 11 7 0 -2 -6 -6 11 2 -3 7 1 5 -10 7 -10 0 12 12 -2 7 -3 2 -6 13 8 -4 11 -10 -9 4 9 -8 14 -3 7 -2 -4 8 4 -3 -1 -2 -5 12 -9 8 15 10 10 13 10 6 4 -8 7 -9 -1 7 -2 14 -10 2 6 -9 6 9 0 6 15 10 2 6 12 10 9 -2 5 -7 4 -3 9 -6 6 1 6 14 13 14 3 4 9 -8 0 10 7 -6 3 11 7 9 -8 8 5 -8 8 4 10 0\n",
    "3\n",
    "Output\n",
    "L(-61)VW(-9)STE(+69)\n",
    "'''\n",
    "\n",
    "class SpectralAlignment:\n",
    "    def __init__(self):\n",
    "        massDict, aaDict = self.AminoAcidMassDict()\n",
    "        peptide, sVector, k = self.readFromFile()\n",
    "        mPeptide = self.constructAlignment(peptide, sVector, k, aaDict)\n",
    "        print(mPeptide)\n",
    "        f = open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/result.txt', 'w')\n",
    "        f.write(mPeptide)\n",
    "        f.close()\n",
    "    \n",
    "    def AminoAcidMassDict(self):\n",
    "        massTable = '''\n",
    "G 57\n",
    "A 71\n",
    "S 87\n",
    "P 97\n",
    "V 99\n",
    "T 101\n",
    "C 103\n",
    "I 113\n",
    "L 113\n",
    "N 114\n",
    "D 115\n",
    "K 128\n",
    "Q 128\n",
    "E 129\n",
    "M 131\n",
    "H 137\n",
    "F 147\n",
    "R 156\n",
    "Y 163\n",
    "W 186'''\n",
    "        mass = massTable.split()\n",
    "        return {int(mass[i+1]):mass[i] for i in range(0, len(mass), 2)}, {mass[i]:int(mass[i+1]) for i in range(0, len(mass), 2)}\n",
    "    \n",
    "    def readFromFile(self):\n",
    "        f = open('/Users/patsnap/Desktop/Neo4J_and_other_codes/Bioinformatics/trial.txt', 'r')\n",
    "        data = []\n",
    "        for line in f:\n",
    "            data.append(line.strip().split())\n",
    "        peptide = data[0][0]\n",
    "        sVector = [0] + list(map(int, data[1]))\n",
    "        k = int(data[2][0])\n",
    "        return peptide, sVector, k\n",
    "    \n",
    "    def getScore(self, node, score):\n",
    "        if node in score:\n",
    "            return score[node]\n",
    "        else:\n",
    "            score[node] = -np.inf\n",
    "            return -np.inf\n",
    "\n",
    "    def constructAlignment(self, peptide, sVector, k, aaDict):\n",
    "        prefixMasses = [0] + [sum([aaDict[aa] for aa in peptide[:i+1]]) for i in range(len(peptide))]\n",
    "        diff = {prefixMasses[i+1]:(prefixMasses[i+1]-prefixMasses[i]) for i in range(len(peptide))}\n",
    "        score = dict()\n",
    "        score[(0,0,0)] = 0\n",
    "        backtrack = dict()\n",
    "        for i in prefixMasses[1:]:\n",
    "            if i < len(sVector):\n",
    "                score[(i,i,0)] = sVector[i]+score[(i-diff[i],i-diff[i],0)]\n",
    "                backtrack[(i,i,0)] = (i-diff[i],i-diff[i],0)\n",
    "            else:\n",
    "                break\n",
    "        for t in range(1, k+1):\n",
    "            for i in prefixMasses[1:]:\n",
    "                for j in range(1, len(sVector)):\n",
    "                    prevList = [(i-diff[i],j-diff[i],t)] + [(i-diff[i],j1,t-1) for j1 in range(j)]\n",
    "                    prevIndex = np.argmax([sVector[j] + self.getScore(node, score) for node in prevList])\n",
    "                    score[(i, j, t)] = sVector[j] + self.getScore(prevList[prevIndex], score)\n",
    "                    backtrack[(i, j, t)] = prevList[prevIndex]\n",
    "        lastNodes = [(prefixMasses[-1], len(sVector)-1, t) for t in range(k+1)]\n",
    "        t = np.argmax([self.getScore(node, score) for node in lastNodes])\n",
    "        prevNode = lastNodes[t]\n",
    "        mPeptide = ''\n",
    "        pList = [peptide[i] for i in range(len(peptide))]\n",
    "        while (0,0,0) != prevNode:\n",
    "            node = backtrack[prevNode]\n",
    "            if node[2] == prevNode[2]:\n",
    "                mPeptide = pList.pop() + mPeptide\n",
    "                prevNode = node\n",
    "            else:\n",
    "                indel = prevNode[1]-node[1]-diff[prevNode[0]]\n",
    "                if indel > 0:\n",
    "                    mPeptide = '(+'+str(indel)+')'+mPeptide\n",
    "                else:\n",
    "                    mPeptide = '('+str(indel)+')'+mPeptide \n",
    "                mPeptide = pList.pop() + mPeptide\n",
    "                prevNode = node\n",
    "        return mPeptide\n",
    "    \n",
    "SpectralAlignment()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
